<!DOCTYPE html>
<html class="writer-html5" lang="en" >
<head>
  <meta charset="utf-8" /><meta name="generator" content="Docutils 0.17.1: http://docutils.sourceforge.net/" />

  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  <title>Using AIRR Community Reference Sets with IgBLAST &mdash; receptor-utils  documentation</title>
      <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
      <link rel="stylesheet" href="_static/css/theme.css" type="text/css" />
  <!--[if lt IE 9]>
    <script src="_static/js/html5shiv.min.js"></script>
  <![endif]-->
  
        <script data-url_root="./" id="documentation_options" src="_static/documentation_options.js"></script>
        <script src="_static/jquery.js"></script>
        <script src="_static/underscore.js"></script>
        <script src="_static/_sphinx_javascript_frameworks_compat.js"></script>
        <script src="_static/doctools.js"></script>
        <script src="_static/sphinx_highlight.js"></script>
    <script src="_static/js/theme.js"></script>
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="next" title="Using AIRR Community Reference Sets with MiXCR" href="airrc_sets_with_mixcr.html" />
    <link rel="prev" title="Introduction" href="introduction.html" /> 
</head>

<body class="wy-body-for-nav"> 
  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >
            <a href="index.html" class="icon icon-home"> receptor-utils
          </a>
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>
        </div><div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <p class="caption" role="heading"><span class="caption-text">Contents:</span></p>
<ul class="current">
<li class="toctree-l1"><a class="reference internal" href="introduction.html">Introduction</a></li>
<li class="toctree-l1 current"><a class="current reference internal" href="#">Using AIRR Community Reference Sets with IgBLAST</a><ul>
<li class="toctree-l2"><a class="reference internal" href="#prerequisites">Prerequisites</a></li>
<li class="toctree-l2"><a class="reference internal" href="#download-the-germline-set-for-human-ighv-genes-and-additional-igblast-files">1. Download the germline set for human IGHV genes and additional IgBLAST files</a></li>
<li class="toctree-l2"><a class="reference internal" href="#log-in-to-the-immcantation-container-so-that-we-can-use-its-installed-tools">2. Log in to the Immcantation container so that we can use its installed tools</a></li>
<li class="toctree-l2"><a class="reference internal" href="#build-igblast-databases-for-the-germline-set">3. Build IgBLAST databases for the germline set</a></li>
<li class="toctree-l2"><a class="reference internal" href="#annotate-the-sample-sequences-using-igblast">4. Annotate the sample sequences using IgBLAST</a></li>
<li class="toctree-l2"><a class="reference internal" href="#convert-the-igblast-output-to-change-o-format">5. Convert the IgBLAST output to Change-O format</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="airrc_sets_with_mixcr.html">Using AIRR Community Reference Sets with MiXCR</a></li>
<li class="toctree-l1"><a class="reference internal" href="custom_igblast.html">Using custom databases with IgBlast</a></li>
<li class="toctree-l1"><a class="reference internal" href="scripts.html">Command-Line Utilities</a></li>
<li class="toctree-l1"><a class="reference internal" href="modules.html">receptor_utils</a></li>
</ul>

        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap"><nav class="wy-nav-top" aria-label="Mobile navigation menu" >
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="index.html">receptor-utils</a>
      </nav>

      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="Page navigation">
  <ul class="wy-breadcrumbs">
      <li><a href="index.html" class="icon icon-home"></a></li>
      <li class="breadcrumb-item active">Using AIRR Community Reference Sets with IgBLAST</li>
      <li class="wy-breadcrumbs-aside">
            <a href="_sources/airrc_sets_with_igblast.rst.txt" rel="nofollow"> View page source</a>
      </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
             
  <section id="using-airr-community-reference-sets-with-igblast">
<span id="airrc-sets-with-igblast"></span><h1>Using AIRR Community Reference Sets with IgBLAST<a class="headerlink" href="#using-airr-community-reference-sets-with-igblast" title="Permalink to this heading"></a></h1>
<p>The <a class="reference internal" href="download_germline_set.html#download-germline-set"><span class="std std-ref">download_germline_set</span></a> utility is designed to make it as easy as possible to use AIRR Community germline reference sets from the Open Germline Receptor Database (<a class="reference external" href="https://ogrdb.airr-community.org">OGRDB</a>) with IgBLAST.
The utility can download the germline reference set in a format that is compatible with IgBLAST, and create the necessary auxiliary files for IgBLAST to use the germline reference set. In this section we will describe how to
annotate a small example dataset using the germline reference set for human IGHV genes, and the accompanying auxiliary files.</p>
<p>For simplicity, we will use the Immcantation Docker/Singularity container. IgBLAST is installed in the container, together with other Immcantation tools.
If you are not already familiar with the container, you can read about it in the <a class="reference external" href="https://immcantation.readthedocs.io/en/stable/">Immcantation documentation</a>. Alternatively, you can install IgBLAST on your local machine and skip the ‘log in’ step below.
To install locally, use the instructions provided on the <a class="reference external" href="https://ncbi.github.io/IgBLAST/">IgBLAST website</a>. Please make sure that you use a recent version of IgBLAST, as the procedure for using custom databases has been significantly simplified from version 1.20 onwards.</p>
<p>We will annotate a sample set of sequences provided on the <a class="reference external" href="http://clip.med.yale.edu/immcantation/examples/AIRR_Example.tar.gz">Immcantation website</a>.
The sequences have been preprocessed and quality-filtered from Illumina paired-end reads and are present in a FASTA file called <code class="docutils literal notranslate"><span class="pre">HD13M.fasta</span></code>. The process required to create such a file from sequencing reads, will depend on the sequencing protocol.
You can consult the Presto documentation on the Immcantation website, or other sources, to determine a suitable approach for you sequencing data.</p>
<section id="prerequisites">
<h2>Prerequisites<a class="headerlink" href="#prerequisites" title="Permalink to this heading"></a></h2>
<p>Before you start, you will need to have the following installed on your machine:</p>
<blockquote>
<div><ul class="simple">
<li><p>Docker or Singularity</p></li>
<li><p>Python 3.9 or above</p></li>
<li><p>The receptor-utils package (see <a class="reference internal" href="introduction.html#introduction-label"><span class="std std-ref">Introduction</span></a> for installation instructions)</p></li>
<li><p>The file <code class="docutils literal notranslate"><span class="pre">HD13M.fasta</span></code>, extracted from the tarball which you can download <a class="reference external" href="http://clip.med.yale.edu/immcantation/examples/AIRR_Example.tar.gz">here</a>.</p></li>
</ul>
</div></blockquote>
<p>The steps we shall follow are as follows:</p>
<ol class="arabic simple">
<li><p>Download the germline set for human IGHV genes and additional IgBLAST files.</p></li>
<li><p>Log in to the Immcantation container so that we can use its installed tools.</p></li>
<li><p>Build IgBLAST databases for the germline set.</p></li>
<li><p>Annotate the sample sequences using IgBLAST.</p></li>
<li><p>Convert the IgBLAST output to Change-O format.</p></li>
</ol>
</section>
<section id="download-the-germline-set-for-human-ighv-genes-and-additional-igblast-files">
<h2>1. Download the germline set for human IGHV genes and additional IgBLAST files<a class="headerlink" href="#download-the-germline-set-for-human-ighv-genes-and-additional-igblast-files" title="Permalink to this heading"></a></h2>
<p>In a suitable directory to use for this test, use the <a class="reference internal" href="download_germline_set.html#download-germline-set"><span class="std std-ref">download_germline_set</span></a> utility to download the germline set for human IGHV genes in a format that is compatible with IgBLAST:</p>
<div class="highlight-none notranslate"><div class="highlight"><pre><span></span>$ download_germline_set &quot;Homo sapiens&quot; IGH -f MULTI-IGBLAST
https://ogrdb.airr-community.org/api_v2/germline/species
Homo sapiens: 9606
9606.IGH_VDJ
FASTA files saved to Homo_sapiens_IGH_V.fasta, Homo_sapiens_IGH_D.fasta, Homo_sapiens_IGH_J.fasta, Homo_sapiens_IGH_V_gapped.fasta
IgBLAST ndm file saved to Homo_sapiens_IGH.ndm
IgBLAST aux file saved to Homo_sapiens_IGH.aux
</pre></div>
</div>
<p>Finally, extract the file <code class="docutils literal notranslate"><span class="pre">HD13M.fasta`</span></code> from the downloaded tarball and copy it to the directory.</p>
</section>
<section id="log-in-to-the-immcantation-container-so-that-we-can-use-its-installed-tools">
<h2>2. Log in to the Immcantation container so that we can use its installed tools<a class="headerlink" href="#log-in-to-the-immcantation-container-so-that-we-can-use-its-installed-tools" title="Permalink to this heading"></a></h2>
<p>Log in to the container, mounting the current local directory as /data.</p>
<p>From Linux using Docker:</p>
<div class="highlight-none notranslate"><div class="highlight"><pre><span></span>docker run -it -v $(pwd):/data:z immcantation/suite:4.5.0 bash
</pre></div>
</div>
<p>From Windows using Docker:</p>
<div class="highlight-none notranslate"><div class="highlight"><pre><span></span>docker run -it -v %cd%:/data:z immcantation/suite:4.5.0 bash
</pre></div>
</div>
<p>For further options, e.g. use with Singularity, please refer to the Immcantation documentation.</p>
<p>Once in the container, cd to /data and check that the reference set files are present:</p>
<div class="highlight-none notranslate"><div class="highlight"><pre><span></span>[root@b426e4d7c0ae /]# cd /data
[root@b426e4d7c0ae data]# ls
HD13M.fasta           Homo_sapiens_IGH.ndm      Homo_sapiens_IGH_J.fasta  Homo_sapiens_IGH_V_gapped.fasta
Homo_sapiens_IGH.aux  Homo_sapiens_IGH_D.fasta  Homo_sapiens_IGH_V.fasta
[root@b426e4d7c0ae data]#
</pre></div>
</div>
</section>
<section id="build-igblast-databases-for-the-germline-set">
<h2>3. Build IgBLAST databases for the germline set<a class="headerlink" href="#build-igblast-databases-for-the-germline-set" title="Permalink to this heading"></a></h2>
<p>Use IgBLAST’s makeblastdb tool to build the germline databases:</p>
<div class="highlight-none notranslate"><div class="highlight"><pre><span></span>makeblastdb -parse_seqids -dbtype nucl -in Homo_sapiens_IGH_V.fasta -out Homo_sapiens_IGH_V
makeblastdb -parse_seqids -dbtype nucl -in Homo_sapiens_IGH_D.fasta -out Homo_sapiens_IGH_D
makeblastdb -parse_seqids -dbtype nucl -in Homo_sapiens_IGH_J.fasta -out Homo_sapiens_IGH_J
</pre></div>
</div>
<p>After these commands have run, you will see many more files in the directory, for example</p>
<div class="highlight-none notranslate"><div class="highlight"><pre><span></span>[root@b426e4d7c0ae data]# ls Homo_sapiens_IGH_V.*
Homo_sapiens_IGH_V.fasta  Homo_sapiens_IGH_V.nhr  Homo_sapiens_IGH_V.njs  Homo_sapiens_IGH_V.nos  Homo_sapiens_IGH_V.nsq  Homo_sapiens_IGH_V.nto
Homo_sapiens_IGH_V.ndb    Homo_sapiens_IGH_V.nin  Homo_sapiens_IGH_V.nog  Homo_sapiens_IGH_V.not  Homo_sapiens_IGH_V.ntf
[root@b426e4d7c0ae data]#
</pre></div>
</div>
</section>
<section id="annotate-the-sample-sequences-using-igblast">
<h2>4. Annotate the sample sequences using IgBLAST<a class="headerlink" href="#annotate-the-sample-sequences-using-igblast" title="Permalink to this heading"></a></h2>
<p>Annotate the sequences in <code class="docutils literal notranslate"><span class="pre">HD13M.fasta`</span></code> with IgBLAST. We use the verbose output format <code class="docutils literal notranslate"><span class="pre">7</span> <span class="pre">std</span> <span class="pre">qseq</span> <span class="pre">sseq</span> <span class="pre">btop</span></code>. The IGDATA environment variable needs
to be set for IgBLAST to run, but in fact the files in that directory are not used, as we override them with command-line options in order
to use the files downloaded from OGRDB:</p>
<div class="highlight-none notranslate"><div class="highlight"><pre><span></span>export IGDATA=/usr/local/share/igblast
igblastn \
    -germline_db_V Homo_sapiens_IGH_V \
    -germline_db_D Homo_sapiens_IGH_D \
    -germline_db_J Homo_sapiens_IGH_J \
    -auxiliary_data Homo_sapiens_IGH.aux \
    -custom_internal_data Homo_sapiens_IGH.ndm \
    -domain_system imgt \
    -outfmt &#39;7 std qseq sseq btop&#39; \
    -num_threads 20 \
    -query HD13M.fasta \
    -out HD13M.fmt7
</pre></div>
</div>
<p>Once IgBLAST has run, you can examine the output file <cite>HD13M.fmt7`</cite> witn <cite>more</cite> or another suitable tool, to confirm it contains sequence annotations.</p>
</section>
<section id="convert-the-igblast-output-to-change-o-format">
<h2>5. Convert the IgBLAST output to Change-O format<a class="headerlink" href="#convert-the-igblast-output-to-change-o-format" title="Permalink to this heading"></a></h2>
<p>Run the Immcantation tool MakeDb to create a TSV database of the IgBLAST output. Note that we use the –failed and –log options to capture details of any
sequences that fail annotation or import. Please refer to <a class="reference internal" href="custom_igblast.html#custom-igblast-label"><span class="std std-ref">Using custom databases with IgBlast</span></a> for some guidance on how to check for errors.</p>
<div class="highlight-none notranslate"><div class="highlight"><pre><span></span>MakeDb.py igblast -i HD13M.fmt7 -s HD13M.fasta --failed --log HD13M.log \
    -r Homo_sapiens_IGH_V_gapped.fasta Homo_sapiens_IGH_D.fasta Homo_sapiens_IGH_J.fasta \
    --extended
</pre></div>
</div>
<p>The output should be similar to this:</p>
<div class="highlight-none notranslate"><div class="highlight"><pre><span></span>OUTPUT&gt; HD13M_db-pass.tsv
PASS&gt; 7887
FAIL&gt; 1402
END&gt; MakeDb
</pre></div>
</div>
<p>You can review <code class="docutils literal notranslate"><span class="pre">HD13M_db-pass.tsv</span></code> to see the resulting output.</p>
</section>
</section>


           </div>
          </div>
          <footer><div class="rst-footer-buttons" role="navigation" aria-label="Footer">
        <a href="introduction.html" class="btn btn-neutral float-left" title="Introduction" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left" aria-hidden="true"></span> Previous</a>
        <a href="airrc_sets_with_mixcr.html" class="btn btn-neutral float-right" title="Using AIRR Community Reference Sets with MiXCR" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right" aria-hidden="true"></span></a>
    </div>

  <hr/>

  <div role="contentinfo">
    <p>&#169; Copyright 2022, William Lees.</p>
  </div>

  Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    provided by <a href="https://readthedocs.org">Read the Docs</a>.
   

</footer>
        </div>
      </div>
    </section>
  </div>
  <script>
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script> 

</body>
</html>