<!DOCTYPE html>
<html class="writer-html5" lang="en" >
<head>
  <meta charset="utf-8" /><meta name="generator" content="Docutils 0.17.1: http://docutils.sourceforge.net/" />

  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  <title>make_igblast_ndm &mdash; receptor-utils  documentation</title>
      <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
      <link rel="stylesheet" href="_static/css/theme.css" type="text/css" />
  <!--[if lt IE 9]>
    <script src="_static/js/html5shiv.min.js"></script>
  <![endif]-->
  
        <script data-url_root="./" id="documentation_options" src="_static/documentation_options.js"></script>
        <script src="_static/jquery.js"></script>
        <script src="_static/underscore.js"></script>
        <script src="_static/_sphinx_javascript_frameworks_compat.js"></script>
        <script src="_static/doctools.js"></script>
        <script src="_static/sphinx_highlight.js"></script>
    <script src="_static/js/theme.js"></script>
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="next" title="annotate_j" href="annotate_j.html" />
    <link rel="prev" title="download_germline_set" href="download_germline_set.html" /> 
</head>

<body class="wy-body-for-nav"> 
  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >
            <a href="index.html" class="icon icon-home"> receptor-utils
          </a>
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>
        </div><div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <p class="caption" role="heading"><span class="caption-text">Contents:</span></p>
<ul class="current">
<li class="toctree-l1"><a class="reference internal" href="introduction.html">Introduction</a></li>
<li class="toctree-l1"><a class="reference internal" href="airrc_sets_with_igblast.html">Using AIRR Community Reference Sets with IgBLAST</a></li>
<li class="toctree-l1"><a class="reference internal" href="airrc_sets_with_mixcr.html">Using AIRR Community Reference Sets with MiXCR</a></li>
<li class="toctree-l1"><a class="reference internal" href="custom_igblast.html">Using custom databases with IgBlast</a></li>
<li class="toctree-l1 current"><a class="reference internal" href="scripts.html">Command-Line Utilities</a><ul class="current">
<li class="toctree-l2"><a class="reference internal" href="scripts.html#naming-and-aligning-v-d-and-j-sequences">Naming and Aligning V, D and J Sequences</a></li>
<li class="toctree-l2 current"><a class="reference internal" href="scripts.html#working-with-igblast">Working with IgBLAST</a><ul class="current">
<li class="toctree-l3"><a class="reference internal" href="download_germline_set.html">download_germline_set</a></li>
<li class="toctree-l3 current"><a class="current reference internal" href="#">make_igblast_ndm</a><ul>
<li class="toctree-l4"><a class="reference internal" href="#positional-arguments">Positional Arguments</a></li>
<li class="toctree-l4"><a class="reference internal" href="#named-arguments">Named Arguments</a></li>
<li class="toctree-l4"><a class="reference internal" href="#notes">Notes</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="annotate_j.html">annotate_j</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="scripts.html#convenience-tools">Convenience Tools</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="modules.html">receptor_utils</a></li>
</ul>

        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap"><nav class="wy-nav-top" aria-label="Mobile navigation menu" >
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="index.html">receptor-utils</a>
      </nav>

      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="Page navigation">
  <ul class="wy-breadcrumbs">
      <li><a href="index.html" class="icon icon-home"></a></li>
          <li class="breadcrumb-item"><a href="scripts.html">Command-Line Utilities</a></li>
      <li class="breadcrumb-item active">make_igblast_ndm</li>
      <li class="wy-breadcrumbs-aside">
            <a href="_sources/make_igblast_ndm.rst.txt" rel="nofollow"> View page source</a>
      </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
             
  <section id="make-igblast-ndm">
<span id="make-igblast-ndm-label"></span><h1>make_igblast_ndm<a class="headerlink" href="#make-igblast-ndm" title="Permalink to this heading"></a></h1>
<p><p>Make IMGT-delineated igblast ndm file from an IMGT-gapped variable-region reference set or AIRR-C JSON format file.</p>
</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">usage</span><span class="p">:</span> <span class="n">make_igblast_ndm</span> <span class="p">[</span><span class="o">-</span><span class="n">h</span><span class="p">]</span> <span class="p">[</span><span class="o">--</span><span class="n">cdr_coords</span> <span class="n">CDR_COORDS</span><span class="p">]</span> <span class="n">ref_file</span> <span class="n">chain</span> <span class="n">ndm_file</span>
</pre></div>
</div>
<section id="positional-arguments">
<h2>Positional Arguments<a class="headerlink" href="#positional-arguments" title="Permalink to this heading"></a></h2>
<dl class="option-list">
<dt><kbd>ref_file</kbd></dt>
<dd><p>gapped reference set (.fasta, .json)</p>
</dd>
<dt><kbd>chain</kbd></dt>
<dd><p>chain as required by igblast (eg VH)</p>
</dd>
<dt><kbd>ndm_file</kbd></dt>
<dd><p>output file (ndm)</p>
</dd>
</dl>
</section>
<section id="named-arguments">
<h2>Named Arguments<a class="headerlink" href="#named-arguments" title="Permalink to this heading"></a></h2>
<dl class="option-list">
<dt><kbd>--cdr_coords, -c</kbd></dt>
<dd><p>comma-separated list of 1-based co-ordinates for CDR1, CDR2 and CDR3-start (default 79,114,166,195,313)</p>
</dd>
</dl>
</section>
<section id="notes">
<h2>Notes<a class="headerlink" href="#notes" title="Permalink to this heading"></a></h2>
<p>If the reference set is available in AIRR-C JSON format, it is recommended to use this format as it will ensure correct placement of the CDRs using coordinates for each sequence from the file.
All germline sets on <a class="reference external" href="https://ogrdb.airr-community.org">OGRDB</a> are available in this format.</p>
<p>If the fasta format is used, V sequences must be IMGT gapped. By default, make_igblast_ndm assumes that the gapped alignment follows the  <a class="reference external" href="https://www.imgt.org/IMGTScientificChart/Numbering/IMGTnumbering.html">IMGT numbering scheme</a>
with no inserted codons. Please note that some reference sets available from IMGT, such as those for rhesus macaque, do contain insertions in the alignment. In this case, you can use the <code class="docutils literal notranslate"><span class="pre">--cdr_coords</span></code> option to
specify coordinates of the CDRs. If in doubt, please analyse one sequence from the gapped set in IMGT V-QUEST and review the codon alignment: if there are insertions, denoted by a codon number followed by A, B, C, etc.,
you will need to specify the CDR coordinates using the <code class="docutils literal notranslate"><span class="pre">--cdr_coords</span></code> option. An alternative is to review the sequences in the reference set and remove the insertions. This may be an option, depending upon
the application, the number of germline sequences that have the insertion, and their usage in your samples. We offer a specific tool, <a class="reference internal" href="fix_macaque_gaps.html#fix-macaque-gaps-label"><span class="std std-ref">fix_macaque_gaps</span></a> to conduct this realignment for IMGT rhesus macaque sets.</p>
<p>Please note that downstream analysis tools may assume that IMGT-gapped sequences follow the default alignment with no inserted codons. If you use a tool that requires a gapped set of V sequences and there is no
provision to specify CDR coordinates, this is likely to be the case.</p>
</section>
</section>


           </div>
          </div>
          <footer><div class="rst-footer-buttons" role="navigation" aria-label="Footer">
        <a href="download_germline_set.html" class="btn btn-neutral float-left" title="download_germline_set" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left" aria-hidden="true"></span> Previous</a>
        <a href="annotate_j.html" class="btn btn-neutral float-right" title="annotate_j" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right" aria-hidden="true"></span></a>
    </div>

  <hr/>

  <div role="contentinfo">
    <p>&#169; Copyright 2022, William Lees.</p>
  </div>

  Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    provided by <a href="https://readthedocs.org">Read the Docs</a>.
   

</footer>
        </div>
      </div>
    </section>
  </div>
  <script>
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script> 

</body>
</html>