<!DOCTYPE html>
<html class="writer-html5" lang="en" >
<head>
  <meta charset="utf-8" /><meta name="generator" content="Docutils 0.17.1: http://docutils.sourceforge.net/" />

  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  <title>receptor_utils package &mdash; receptor-utils  documentation</title>
      <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
      <link rel="stylesheet" href="_static/css/theme.css" type="text/css" />
  <!--[if lt IE 9]>
    <script src="_static/js/html5shiv.min.js"></script>
  <![endif]-->
  
        <script data-url_root="./" id="documentation_options" src="_static/documentation_options.js"></script>
        <script src="_static/jquery.js"></script>
        <script src="_static/underscore.js"></script>
        <script src="_static/_sphinx_javascript_frameworks_compat.js"></script>
        <script src="_static/doctools.js"></script>
        <script src="_static/sphinx_highlight.js"></script>
    <script src="_static/js/theme.js"></script>
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="prev" title="API" href="modules.html" /> 
</head>

<body class="wy-body-for-nav"> 
  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >
            <a href="index.html" class="icon icon-home"> receptor-utils
          </a>
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>
        </div><div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <p class="caption" role="heading"><span class="caption-text">Contents:</span></p>
<ul class="current">
<li class="toctree-l1"><a class="reference internal" href="introduction.html">Introduction</a></li>
<li class="toctree-l1"><a class="reference internal" href="custom_igblast.html">Using custom databases with IgBlast</a></li>
<li class="toctree-l1"><a class="reference internal" href="scripts.html">Command-Line Utilities</a></li>
<li class="toctree-l1 current"><a class="reference internal" href="modules.html">API</a><ul class="current">
<li class="toctree-l2 current"><a class="current reference internal" href="#">receptor_utils package</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#submodules">Submodules</a></li>
<li class="toctree-l3"><a class="reference internal" href="#module-receptor_utils.novel_allele_name">receptor_utils.novel_allele_name module</a><ul>
<li class="toctree-l4"><a class="reference internal" href="#receptor_utils.novel_allele_name.build_ambiguous_ref"><code class="docutils literal notranslate"><span class="pre">build_ambiguous_ref()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#receptor_utils.novel_allele_name.find_indels"><code class="docutils literal notranslate"><span class="pre">find_indels()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#receptor_utils.novel_allele_name.name_novel"><code class="docutils literal notranslate"><span class="pre">name_novel()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#receptor_utils.novel_allele_name.run_tests"><code class="docutils literal notranslate"><span class="pre">run_tests()</span></code></a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="#module-receptor_utils.number_v">receptor_utils.number_v module</a><ul>
<li class="toctree-l4"><a class="reference internal" href="#receptor_utils.number_v.check_conserved_residues"><code class="docutils literal notranslate"><span class="pre">check_conserved_residues()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#receptor_utils.number_v.distribute_cdr"><code class="docutils literal notranslate"><span class="pre">distribute_cdr()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#receptor_utils.number_v.gap_align"><code class="docutils literal notranslate"><span class="pre">gap_align()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#receptor_utils.number_v.gap_align_aa"><code class="docutils literal notranslate"><span class="pre">gap_align_aa()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#receptor_utils.number_v.gap_align_aa_from_nt"><code class="docutils literal notranslate"><span class="pre">gap_align_aa_from_nt()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#receptor_utils.number_v.gap_nt_from_aa"><code class="docutils literal notranslate"><span class="pre">gap_nt_from_aa()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#receptor_utils.number_v.gap_sequence"><code class="docutils literal notranslate"><span class="pre">gap_sequence()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#receptor_utils.number_v.insert_space"><code class="docutils literal notranslate"><span class="pre">insert_space()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#receptor_utils.number_v.match_score"><code class="docutils literal notranslate"><span class="pre">match_score()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#receptor_utils.number_v.nt_diff"><code class="docutils literal notranslate"><span class="pre">nt_diff()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#receptor_utils.number_v.number_from_trp"><code class="docutils literal notranslate"><span class="pre">number_from_trp()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#receptor_utils.number_v.number_ighv"><code class="docutils literal notranslate"><span class="pre">number_ighv()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#receptor_utils.number_v.pretty_gapped"><code class="docutils literal notranslate"><span class="pre">pretty_gapped()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#receptor_utils.number_v.run_tests"><code class="docutils literal notranslate"><span class="pre">run_tests()</span></code></a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="#module-receptor_utils.simple_bio_seq">receptor_utils.simple_bio_seq module</a><ul>
<li class="toctree-l4"><a class="reference internal" href="#receptor_utils.simple_bio_seq.chunks"><code class="docutils literal notranslate"><span class="pre">chunks()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#receptor_utils.simple_bio_seq.closest_ref"><code class="docutils literal notranslate"><span class="pre">closest_ref()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#receptor_utils.simple_bio_seq.dumb_consensus"><code class="docutils literal notranslate"><span class="pre">dumb_consensus()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#receptor_utils.simple_bio_seq.nt_diff"><code class="docutils literal notranslate"><span class="pre">nt_diff()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#receptor_utils.simple_bio_seq.read_csv"><code class="docutils literal notranslate"><span class="pre">read_csv()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#receptor_utils.simple_bio_seq.read_fasta"><code class="docutils literal notranslate"><span class="pre">read_fasta()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#receptor_utils.simple_bio_seq.read_imgt_fasta"><code class="docutils literal notranslate"><span class="pre">read_imgt_fasta()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#receptor_utils.simple_bio_seq.read_single_fasta"><code class="docutils literal notranslate"><span class="pre">read_single_fasta()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#receptor_utils.simple_bio_seq.reverse_complement"><code class="docutils literal notranslate"><span class="pre">reverse_complement()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#receptor_utils.simple_bio_seq.sample_fasta"><code class="docutils literal notranslate"><span class="pre">sample_fasta()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#receptor_utils.simple_bio_seq.scored_consensus"><code class="docutils literal notranslate"><span class="pre">scored_consensus()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#receptor_utils.simple_bio_seq.scored_gap_consensus"><code class="docutils literal notranslate"><span class="pre">scored_gap_consensus()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#receptor_utils.simple_bio_seq.toSeqRecords"><code class="docutils literal notranslate"><span class="pre">toSeqRecords()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#receptor_utils.simple_bio_seq.translate"><code class="docutils literal notranslate"><span class="pre">translate()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#receptor_utils.simple_bio_seq.write_csv"><code class="docutils literal notranslate"><span class="pre">write_csv()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#receptor_utils.simple_bio_seq.write_fasta"><code class="docutils literal notranslate"><span class="pre">write_fasta()</span></code></a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="#module-receptor_utils">Module contents</a></li>
</ul>
</li>
</ul>
</li>
</ul>

        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap"><nav class="wy-nav-top" aria-label="Mobile navigation menu" >
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="index.html">receptor-utils</a>
      </nav>

      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="Page navigation">
  <ul class="wy-breadcrumbs">
      <li><a href="index.html" class="icon icon-home"></a></li>
          <li class="breadcrumb-item"><a href="modules.html">API</a></li>
      <li class="breadcrumb-item active">receptor_utils package</li>
      <li class="wy-breadcrumbs-aside">
            <a href="_sources/receptor_utils.rst.txt" rel="nofollow"> View page source</a>
      </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
             
  <section id="receptor-utils-package">
<h1>receptor_utils package<a class="headerlink" href="#receptor-utils-package" title="Permalink to this heading"></a></h1>
<section id="submodules">
<h2>Submodules<a class="headerlink" href="#submodules" title="Permalink to this heading"></a></h2>
</section>
<section id="module-receptor_utils.novel_allele_name">
<span id="receptor-utils-novel-allele-name-module"></span><h2>receptor_utils.novel_allele_name module<a class="headerlink" href="#module-receptor_utils.novel_allele_name" title="Permalink to this heading"></a></h2>
<dl class="py function">
<dt class="sig sig-object py" id="receptor_utils.novel_allele_name.build_ambiguous_ref">
<span class="sig-prename descclassname"><span class="pre">receptor_utils.novel_allele_name.</span></span><span class="sig-name descname"><span class="pre">build_ambiguous_ref</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">full_ref</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">start</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/receptor_utils/novel_allele_name.html#build_ambiguous_ref"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#receptor_utils.novel_allele_name.build_ambiguous_ref" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="receptor_utils.novel_allele_name.find_indels">
<span class="sig-prename descclassname"><span class="pre">receptor_utils.novel_allele_name.</span></span><span class="sig-name descname"><span class="pre">find_indels</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">closest_ref_seq</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">novel_seq</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/receptor_utils/novel_allele_name.html#find_indels"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#receptor_utils.novel_allele_name.find_indels" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="receptor_utils.novel_allele_name.name_novel">
<span class="sig-prename descclassname"><span class="pre">receptor_utils.novel_allele_name.</span></span><span class="sig-name descname"><span class="pre">name_novel</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">novel_seq</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ref_set</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">dict</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">v_gene</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">bool</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">True</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/receptor_utils/novel_allele_name.html#name_novel"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#receptor_utils.novel_allele_name.name_novel" title="Permalink to this definition"></a></dt>
<dd><p>Make a name for the novel allele, given its gapped or ungapped sequence.
The name conforms to the description <a class="reference external" href="https://wordpress.vdjbase.org/index.php/vdjbase_help/airr-seq-data-allele-names/">here</a> .
The sequence must be full-length at the 5 prime end, or gapped</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>novel_seq</strong> (<em>str</em>) – the full-length sequence to name, which may be either gapped or ungapped in the case of a V-gene</p></li>
<li><p><strong>ref_set</strong> (<em>dict</em>) – dict of reference genes (gapped in the case of v-genes), in the format returned by read_fasta</p></li>
<li><p><strong>v_gene</strong> (<em>bool</em>) – True only if we are naming a v_gene</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>tuple consisting of three strings: novel_name, novel_seq, notes. novel_seq is gapped in the case of a V-gene</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p>tuple</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="receptor_utils.novel_allele_name.run_tests">
<span class="sig-prename descclassname"><span class="pre">receptor_utils.novel_allele_name.</span></span><span class="sig-name descname"><span class="pre">run_tests</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/receptor_utils/novel_allele_name.html#run_tests"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#receptor_utils.novel_allele_name.run_tests" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

</section>
<section id="module-receptor_utils.number_v">
<span id="receptor-utils-number-v-module"></span><h2>receptor_utils.number_v module<a class="headerlink" href="#module-receptor_utils.number_v" title="Permalink to this heading"></a></h2>
<dl class="py function">
<dt class="sig sig-object py" id="receptor_utils.number_v.check_conserved_residues">
<span class="sig-prename descclassname"><span class="pre">receptor_utils.number_v.</span></span><span class="sig-name descname"><span class="pre">check_conserved_residues</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">aa</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/receptor_utils/number_v.html#check_conserved_residues"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#receptor_utils.number_v.check_conserved_residues" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="receptor_utils.number_v.distribute_cdr">
<span class="sig-prename descclassname"><span class="pre">receptor_utils.number_v.</span></span><span class="sig-name descname"><span class="pre">distribute_cdr</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">cdr</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">length</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/receptor_utils/number_v.html#distribute_cdr"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#receptor_utils.number_v.distribute_cdr" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="receptor_utils.number_v.gap_align">
<span class="sig-prename descclassname"><span class="pre">receptor_utils.number_v.</span></span><span class="sig-name descname"><span class="pre">gap_align</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">seq</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ref</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/receptor_utils/number_v.html#gap_align"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#receptor_utils.number_v.gap_align" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="receptor_utils.number_v.gap_align_aa">
<span class="sig-prename descclassname"><span class="pre">receptor_utils.number_v.</span></span><span class="sig-name descname"><span class="pre">gap_align_aa</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">seq</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ref</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/receptor_utils/number_v.html#gap_align_aa"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#receptor_utils.number_v.gap_align_aa" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="receptor_utils.number_v.gap_align_aa_from_nt">
<span class="sig-prename descclassname"><span class="pre">receptor_utils.number_v.</span></span><span class="sig-name descname"><span class="pre">gap_align_aa_from_nt</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">aa_seq</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">nt_gapped</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/receptor_utils/number_v.html#gap_align_aa_from_nt"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#receptor_utils.number_v.gap_align_aa_from_nt" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="receptor_utils.number_v.gap_nt_from_aa">
<span class="sig-prename descclassname"><span class="pre">receptor_utils.number_v.</span></span><span class="sig-name descname"><span class="pre">gap_nt_from_aa</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">nucleotide_seq</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">peptide_seq</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/receptor_utils/number_v.html#gap_nt_from_aa"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#receptor_utils.number_v.gap_nt_from_aa" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="receptor_utils.number_v.gap_sequence">
<span class="sig-prename descclassname"><span class="pre">receptor_utils.number_v.</span></span><span class="sig-name descname"><span class="pre">gap_sequence</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">seq</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">gapped_ref</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ungapped_ref</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/receptor_utils/number_v.html#gap_sequence"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#receptor_utils.number_v.gap_sequence" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="receptor_utils.number_v.insert_space">
<span class="sig-prename descclassname"><span class="pre">receptor_utils.number_v.</span></span><span class="sig-name descname"><span class="pre">insert_space</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">seq</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">pos</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/receptor_utils/number_v.html#insert_space"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#receptor_utils.number_v.insert_space" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="receptor_utils.number_v.match_score">
<span class="sig-prename descclassname"><span class="pre">receptor_utils.number_v.</span></span><span class="sig-name descname"><span class="pre">match_score</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">target</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">match_list</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">thresh</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/receptor_utils/number_v.html#match_score"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#receptor_utils.number_v.match_score" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="receptor_utils.number_v.nt_diff">
<span class="sig-prename descclassname"><span class="pre">receptor_utils.number_v.</span></span><span class="sig-name descname"><span class="pre">nt_diff</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">s1</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">s2</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/receptor_utils/number_v.html#nt_diff"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#receptor_utils.number_v.nt_diff" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="receptor_utils.number_v.number_from_trp">
<span class="sig-prename descclassname"><span class="pre">receptor_utils.number_v.</span></span><span class="sig-name descname"><span class="pre">number_from_trp</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">seq</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">next_p</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">gapped</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">trp_pos</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/receptor_utils/number_v.html#number_from_trp"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#receptor_utils.number_v.number_from_trp" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="receptor_utils.number_v.number_ighv">
<span class="sig-prename descclassname"><span class="pre">receptor_utils.number_v.</span></span><span class="sig-name descname"><span class="pre">number_ighv</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">seq</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/receptor_utils/number_v.html#number_ighv"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#receptor_utils.number_v.number_ighv" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="receptor_utils.number_v.pretty_gapped">
<span class="sig-prename descclassname"><span class="pre">receptor_utils.number_v.</span></span><span class="sig-name descname"><span class="pre">pretty_gapped</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">seq</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/receptor_utils/number_v.html#pretty_gapped"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#receptor_utils.number_v.pretty_gapped" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="receptor_utils.number_v.run_tests">
<span class="sig-prename descclassname"><span class="pre">receptor_utils.number_v.</span></span><span class="sig-name descname"><span class="pre">run_tests</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/receptor_utils/number_v.html#run_tests"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#receptor_utils.number_v.run_tests" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

</section>
<section id="module-receptor_utils.simple_bio_seq">
<span id="receptor-utils-simple-bio-seq-module"></span><h2>receptor_utils.simple_bio_seq module<a class="headerlink" href="#module-receptor_utils.simple_bio_seq" title="Permalink to this heading"></a></h2>
<p>Wrapper around BioSeq functions for simple applications
principles:
- store sequences as strings, use dicts for collectiopns
- all sequences are coerced to upper case on input
- iterators are coreced into lists for ease of debugging
basically just make things simple for cases where we don’t need to do more</p>
<dl class="py function">
<dt class="sig sig-object py" id="receptor_utils.simple_bio_seq.chunks">
<span class="sig-prename descclassname"><span class="pre">receptor_utils.simple_bio_seq.</span></span><span class="sig-name descname"><span class="pre">chunks</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">l</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">n</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">int</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/receptor_utils/simple_bio_seq.html#chunks"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#receptor_utils.simple_bio_seq.chunks" title="Permalink to this definition"></a></dt>
<dd><p>Yield successive n-sized chunks from l.</p>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="receptor_utils.simple_bio_seq.closest_ref">
<span class="sig-prename descclassname"><span class="pre">receptor_utils.simple_bio_seq.</span></span><span class="sig-name descname"><span class="pre">closest_ref</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">seq</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ref</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">dict</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/receptor_utils/simple_bio_seq.html#closest_ref"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#receptor_utils.simple_bio_seq.closest_ref" title="Permalink to this definition"></a></dt>
<dd><p>Given an input sequence, find the closest entry or entries in an IMGT or other reference set as determined by nt_diff.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>seq</strong> (<em>str</em>) – the input sequence</p></li>
<li><p><strong>ref</strong> (<em>dict</em>) – the reference set</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>the name of the closest entry, entries, if &gt;1 sequences in the reference set were equally close</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p>list</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="receptor_utils.simple_bio_seq.dumb_consensus">
<span class="sig-prename descclassname"><span class="pre">receptor_utils.simple_bio_seq.</span></span><span class="sig-name descname"><span class="pre">dumb_consensus</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">seqs</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">dict</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">threshold</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">float</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">0.7</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/receptor_utils/simple_bio_seq.html#dumb_consensus"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#receptor_utils.simple_bio_seq.dumb_consensus" title="Permalink to this definition"></a></dt>
<dd><p>Return a dumb consensus on a dict of sequences, using Bio.Align.AlignInfo.SummaryInfo.dumb_consensus. All sequences should be the same length.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>seqs</strong> (<em>dict</em>) – the sequences to analyse</p></li>
<li><p><strong>threshold</strong> (<em>float</em>) – the threshold value that is required to add a particular atom</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>the consensus sequence</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p>str</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="receptor_utils.simple_bio_seq.nt_diff">
<span class="sig-prename descclassname"><span class="pre">receptor_utils.simple_bio_seq.</span></span><span class="sig-name descname"><span class="pre">nt_diff</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">s1</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">s2</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/receptor_utils/simple_bio_seq.html#nt_diff"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#receptor_utils.simple_bio_seq.nt_diff" title="Permalink to this definition"></a></dt>
<dd><p>Returns a count of the positions in which the input sequences differ. If the sequences are of different lengths, the count only covers
positions in the shorter sequence.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>s1</strong> (<em>str</em>) – the first sequence</p></li>
<li><p><strong>s2</strong> (<em>str</em>) – the second sequence</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>The number of positions at which the sequences differ</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p>int</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="receptor_utils.simple_bio_seq.read_csv">
<span class="sig-prename descclassname"><span class="pre">receptor_utils.simple_bio_seq.</span></span><span class="sig-name descname"><span class="pre">read_csv</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">file</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">delimiter</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">Optional</span><span class="p"><span class="pre">[</span></span><span class="pre">str</span><span class="p"><span class="pre">]</span></span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/receptor_utils/simple_bio_seq.html#read_csv"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#receptor_utils.simple_bio_seq.read_csv" title="Permalink to this definition"></a></dt>
<dd><p>Read a delimited file into a list of dicts (as produced by DictReader)</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>file</strong> (<em>str</em>) – filename of the file</p></li>
<li><p><strong>delimiter</strong> (<em>str</em>) – the delimiter (‘,’ by default)</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>the list of dicts</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p>list</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="receptor_utils.simple_bio_seq.read_fasta">
<span class="sig-prename descclassname"><span class="pre">receptor_utils.simple_bio_seq.</span></span><span class="sig-name descname"><span class="pre">read_fasta</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">infile</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/receptor_utils/simple_bio_seq.html#read_fasta"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#receptor_utils.simple_bio_seq.read_fasta" title="Permalink to this definition"></a></dt>
<dd><p>Read a FASTA file into a dict</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>infile</strong> (<em>str</em>) – Pathname of the file</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>A dictionary indexed by FASTA ID, containing the sequences (in upper case)</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p>dict</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="receptor_utils.simple_bio_seq.read_imgt_fasta">
<span class="sig-prename descclassname"><span class="pre">receptor_utils.simple_bio_seq.</span></span><span class="sig-name descname"><span class="pre">read_imgt_fasta</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">infile</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">species</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">chains</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">('IGHV',</span> <span class="pre">'IGHD',</span> <span class="pre">'IGHJ',</span> <span class="pre">'CH')</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">functional_only</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">bool</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">include_orphon</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">bool</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/receptor_utils/simple_bio_seq.html#read_imgt_fasta"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#receptor_utils.simple_bio_seq.read_imgt_fasta" title="Permalink to this definition"></a></dt>
<dd><p>read V,D,J regions from one or more species from an IMGT reference file</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>infile</strong> (<em>str</em>) – The IMGT reference file</p></li>
<li><p><strong>species</strong> (<em>str</em>) – Species as specified in the file, with spaces replaced by underscore</p></li>
<li><p><strong>chains</strong> (<em>list</em>) – List of chains to read (e.g. [‘IGHV’, ‘IGHJ’])</p></li>
<li><p><strong>functional_only</strong> – If True, returns only sequences marked as functional (F)</p></li>
<li><p><strong>functional_only</strong> – If True, includes orphons</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>A dict containing the serquences, indexed by name</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p>dict</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="receptor_utils.simple_bio_seq.read_single_fasta">
<span class="sig-prename descclassname"><span class="pre">receptor_utils.simple_bio_seq.</span></span><span class="sig-name descname"><span class="pre">read_single_fasta</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">infile</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/receptor_utils/simple_bio_seq.html#read_single_fasta"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#receptor_utils.simple_bio_seq.read_single_fasta" title="Permalink to this definition"></a></dt>
<dd><p>Read a single sequence from a FASTA file</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>infile</strong> (<em>str</em>) – Pathname of the file</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>The first (or only) sequence in the file</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p>str</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="receptor_utils.simple_bio_seq.reverse_complement">
<span class="sig-prename descclassname"><span class="pre">receptor_utils.simple_bio_seq.</span></span><span class="sig-name descname"><span class="pre">reverse_complement</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">seq</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/receptor_utils/simple_bio_seq.html#reverse_complement"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#receptor_utils.simple_bio_seq.reverse_complement" title="Permalink to this definition"></a></dt>
<dd><p>Return the reverse complement of a nucelotide sequence</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>seq</strong> (<em>str</em>) – the nucleotide sequence</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>the reverse complement</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p>str</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="receptor_utils.simple_bio_seq.sample_fasta">
<span class="sig-prename descclassname"><span class="pre">receptor_utils.simple_bio_seq.</span></span><span class="sig-name descname"><span class="pre">sample_fasta</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">seqs</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">dict</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">number</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">int</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/receptor_utils/simple_bio_seq.html#sample_fasta"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#receptor_utils.simple_bio_seq.sample_fasta" title="Permalink to this definition"></a></dt>
<dd><p>Return a random sample of sequences stored in a dict. Sequences are not resampled.
Returns a ValueError if the dict does not contain enough sequences</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>seqs</strong> (<em>dict</em>) – The sequences to sample</p></li>
<li><p><strong>number</strong> (<em>int</em>) – The number of sequences to return</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>The sampled sequences</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p>dict</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="receptor_utils.simple_bio_seq.scored_consensus">
<span class="sig-prename descclassname"><span class="pre">receptor_utils.simple_bio_seq.</span></span><span class="sig-name descname"><span class="pre">scored_consensus</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">seqs</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">dict</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">threshold</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">float</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">0.7</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/receptor_utils/simple_bio_seq.html#scored_consensus"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#receptor_utils.simple_bio_seq.scored_consensus" title="Permalink to this definition"></a></dt>
<dd><p>Return a dumb consensus on a dict of sequences, using Bio.Align.AlignInfo.SummaryInfo.gap_consensus. All sequences should be the same length.
The output is modified to provide, as well as the consensus string, the mimimum score achieved at any position</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>seqs</strong> (<em>dict</em>) – the sequences to analyse</p></li>
<li><p><strong>threshold</strong> (<em>float</em>) – the threshold value that is required to add a particular atom</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>(consensus sequence, min_score)</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p>str</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="receptor_utils.simple_bio_seq.scored_gap_consensus">
<span class="sig-prename descclassname"><span class="pre">receptor_utils.simple_bio_seq.</span></span><span class="sig-name descname"><span class="pre">scored_gap_consensus</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">alignment</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">threshold</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0.7</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ambiguous</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'X'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">require_multiple</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/receptor_utils/simple_bio_seq.html#scored_gap_consensus"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#receptor_utils.simple_bio_seq.scored_gap_consensus" title="Permalink to this definition"></a></dt>
<dd><p>Adapted from BioPython.Bio.ALign. This function is called by receptor_utils.scored_consensus.</p>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="receptor_utils.simple_bio_seq.toSeqRecords">
<span class="sig-prename descclassname"><span class="pre">receptor_utils.simple_bio_seq.</span></span><span class="sig-name descname"><span class="pre">toSeqRecords</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">seqs</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">dict</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/receptor_utils/simple_bio_seq.html#toSeqRecords"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#receptor_utils.simple_bio_seq.toSeqRecords" title="Permalink to this definition"></a></dt>
<dd><p>Convert a dict of sequences to a list of BioPython SeqRecords</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>seqs</strong> (<em>dict</em>) – the sequences to convert</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>the BioPython SeqRecords</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p>list</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="receptor_utils.simple_bio_seq.translate">
<span class="sig-prename descclassname"><span class="pre">receptor_utils.simple_bio_seq.</span></span><span class="sig-name descname"><span class="pre">translate</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">seq</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">truncate</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">bool</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ignore_partial_codon</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">bool</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">True</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/receptor_utils/simple_bio_seq.html#translate"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#receptor_utils.simple_bio_seq.translate" title="Permalink to this definition"></a></dt>
<dd><p>Translate a nucleotide sequence to amino acid</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>seq</strong> (<em>str</em>) – the sequence to translate</p></li>
<li><p><strong>truncate</strong> (<em>bool</em>) – If True, truncate the sequence so that it terminates on a codon boundary. Otherwise pad with N if necessary</p></li>
<li><p><strong>ignore_partial_codon</strong> (<em>bool</em>) – If True, if any position in a codon contains - or ., set the entire codon to — ensuring it gets translated as -</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>Amino acid string</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p>str</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="receptor_utils.simple_bio_seq.write_csv">
<span class="sig-prename descclassname"><span class="pre">receptor_utils.simple_bio_seq.</span></span><span class="sig-name descname"><span class="pre">write_csv</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">file</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">rows</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">list</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">delimiter</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">Optional</span><span class="p"><span class="pre">[</span></span><span class="pre">str</span><span class="p"><span class="pre">]</span></span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/receptor_utils/simple_bio_seq.html#write_csv"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#receptor_utils.simple_bio_seq.write_csv" title="Permalink to this definition"></a></dt>
<dd><p>Write a list of dicts to a delimited file. The header row is determined from the keys of the first item</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>file</strong> (<em>str</em>) – filename of the delimited file to create</p></li>
<li><p><strong>rows</strong> (<em>list</em>) – the rows to write</p></li>
<li><p><strong>delimiter</strong> (<em>str</em>) – the delimiter (‘,’ by default)</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>None</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p>None</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="receptor_utils.simple_bio_seq.write_fasta">
<span class="sig-prename descclassname"><span class="pre">receptor_utils.simple_bio_seq.</span></span><span class="sig-name descname"><span class="pre">write_fasta</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">outfile</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">seqs</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">dict</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/receptor_utils/simple_bio_seq.html#write_fasta"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#receptor_utils.simple_bio_seq.write_fasta" title="Permalink to this definition"></a></dt>
<dd><p>Write a dict into a FASTA file. The dict should be indexed by sequence name</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>outfile</strong> (<em>str</em>) – Pathname of the file to be written</p></li>
<li><p><strong>seqs</strong> (<em>dict</em>) – The sequences to write</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>the number of records written</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p>int</p>
</dd>
</dl>
</dd></dl>

</section>
<section id="module-receptor_utils">
<span id="module-contents"></span><h2>Module contents<a class="headerlink" href="#module-receptor_utils" title="Permalink to this heading"></a></h2>
</section>
</section>


           </div>
          </div>
          <footer><div class="rst-footer-buttons" role="navigation" aria-label="Footer">
        <a href="modules.html" class="btn btn-neutral float-left" title="API" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left" aria-hidden="true"></span> Previous</a>
    </div>

  <hr/>

  <div role="contentinfo">
    <p>&#169; Copyright 2022, William Lees.</p>
  </div>

  Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    provided by <a href="https://readthedocs.org">Read the Docs</a>.
   

</footer>
        </div>
      </div>
    </section>
  </div>
  <script>
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script> 

</body>
</html>