Search.setIndex({"docnames": ["airrc_sets_with_igblast", "airrc_sets_with_mixcr", "annotate_j", "at_coords", "create_alignment", "custom_igblast", "download_germline_set", "extract_refs", "fix_macaque_gaps", "gap_sequences", "identical_seqs", "index", "introduction", "make_igblast_ndm", "merge_fasta", "modules", "name_allele", "receptor_utils", "rev_comp", "scripts"], "filenames": ["airrc_sets_with_igblast.rst", "airrc_sets_with_mixcr.rst", "annotate_j.rst", "at_coords.rst", "create_alignment.rst", "custom_igblast.rst", "download_germline_set.rst", "extract_refs.rst", "fix_macaque_gaps.rst", "gap_sequences.rst", "identical_seqs.rst", "index.rst", "introduction.rst", "make_igblast_ndm.rst", "merge_fasta.rst", "modules.rst", "name_allele.rst", "receptor_utils.rst", "rev_comp.rst", "scripts.rst"], "titles": ["Using AIRR Community Reference Sets with IgBLAST", "Using AIRR Community Reference Sets with MiXCR", "annotate_j", "at_coords", "create_alignment", "Using custom databases with IgBlast", "download_germline_set", "extract_refs", "fix_macaque_gaps", "gap_sequences", "identical_seqs", "Receptor-utils: code and utilities for working with IG/TR receptor sequences", "Introduction", "make_igblast_ndm", "merge_fasta", "receptor_utils", "name_allele", "receptor_utils package", "rev_comp", "Command-Line Utilities"], "terms": {"The": [0, 1, 4, 5, 6, 7, 9, 12, 14, 17], "download_germline_set": [0, 1, 12, 19], "util": [0, 1, 6, 7, 8, 12, 14], "i": [0, 1, 3, 4, 5, 6, 8, 9, 12, 13, 17], "design": 0, "make": [0, 5, 13, 17], "easi": 0, "possibl": [0, 4, 5, 6], "from": [0, 1, 3, 4, 5, 6, 7, 8, 12, 13, 14, 16, 17], "open": [0, 5, 6, 12], "receptor": [0, 1, 6, 12], "ogrdb": [0, 1, 6, 12, 13], "compat": 0, "creat": [0, 1, 4, 5, 6, 12, 16, 17], "necessari": [0, 5, 16, 17], "auxiliari": [0, 12], "In": [0, 1, 5, 9, 13], "thi": [0, 1, 4, 5, 6, 7, 8, 12, 13, 14, 17], "section": [0, 5, 7, 12], "describ": [0, 5, 12], "how": [0, 5], "small": [0, 8, 9], "exampl": [0, 12], "dataset": 0, "accompani": 0, "For": [0, 1, 4, 5, 6, 12], "simplic": [0, 1], "docker": [0, 1], "singular": [0, 1], "togeth": 0, "other": [0, 1, 4, 5, 8, 9, 12, 17], "If": [0, 1, 5, 6, 7, 13, 17], "you": [0, 1, 5, 8, 13], "ar": [0, 1, 4, 5, 6, 9, 12, 13, 14, 17], "alreadi": 0, "familiar": [0, 1], "read": [0, 1, 5, 12, 17], "about": [0, 1], "document": [0, 1, 12], "altern": [0, 1, 4, 13], "your": [0, 1, 5, 13], "local": [0, 1, 5], "machin": [0, 1], "skip": [0, 1], "step": [0, 1, 5], "below": 0, "To": 0, "instruct": [0, 1], "provid": [0, 1, 5, 6, 12, 17], "websit": [0, 1], "pleas": [0, 6, 13], "sure": 0, "recent": 0, "version": [0, 5, 6], "procedur": [0, 5], "custom": [0, 4, 6, 11], "ha": [0, 1, 12], "been": [0, 1, 5], "significantli": [0, 5], "simplifi": [0, 12], "20": [0, 4, 5], "onward": 0, "have": [0, 1, 5, 8, 13], "preprocess": [0, 1], "qualiti": [0, 1, 5], "filter": [0, 1, 4], "illumina": [0, 1], "pair": [0, 1], "end": [0, 1, 3, 4, 9, 17], "present": [0, 1, 8, 14], "fasta": [0, 1, 2, 3, 4, 6, 7, 8, 9, 10, 12, 13, 14, 16, 17], "call": [0, 1, 5, 17], "hd13m": [0, 1], "process": [0, 1, 16], "requir": [0, 1, 5, 12, 13, 17], "depend": [0, 1, 5, 13], "protocol": [0, 1], "consult": [0, 1], "presto": 0, "sourc": [0, 1, 17], "determin": [0, 1, 6, 17], "suitabl": [0, 1, 16], "approach": [0, 1, 5], "data": [0, 1, 5, 12], "befor": [0, 1, 4, 8, 16], "start": [0, 1, 3, 4, 5, 13, 17], "need": [0, 1, 5, 6, 12, 13, 17], "follow": [0, 1, 4, 5, 13], "python": [0, 1], "9": [0, 1], "abov": [0, 1, 5, 6, 12], "packag": [0, 1, 12, 15], "see": [0, 1, 5, 6], "introduct": [0, 1, 11], "extract": [0, 1, 3, 7], "tarbal": [0, 1], "which": [0, 1, 4, 5, 6, 12, 17], "here": [0, 1, 7, 12, 17], "shall": [0, 1], "directori": [0, 1], "test": [0, 1, 5], "homo": [0, 1, 6, 7], "sapien": [0, 1, 6, 7], "igh": [0, 1, 6, 7, 8, 16], "f": [0, 1, 4, 6, 7, 17], "multi": [0, 1, 6], "http": [0, 1, 6, 7], "org": [0, 1, 6, 7], "api_v2": [0, 1, 6], "speci": [0, 1, 4, 6, 7, 8, 12, 17], "9606": [0, 1], "igh_vdj": [0, 1], "save": [0, 1], "homo_sapiens_igh_v": [0, 1], "homo_sapiens_igh_d": [0, 1], "homo_sapiens_igh_j": [0, 1], "homo_sapiens_igh_v_gap": [0, 1], "ndm": [0, 5, 6, 12, 13], "homo_sapiens_igh": 0, "aux": [0, 2, 5, 6], "final": [0, 1, 6], "copi": [0, 1], "mount": [0, 1], "current": [0, 1, 4, 7], "linux": [0, 1], "run": [0, 1], "v": [0, 1, 2, 4, 5, 6, 13, 16, 17], "pwd": [0, 1, 6], "z": 0, "suit": [0, 12], "0": [0, 17], "bash": [0, 1], "window": [0, 1], "cd": [0, 1], "further": [0, 6], "option": [0, 4, 5, 13, 17], "e": [0, 4, 5, 6, 7, 16, 17], "g": [0, 4, 6, 7, 16, 17], "onc": [0, 1, 5, 12], "check": [0, 1, 5, 8], "root": 0, "b426e4d7c0a": 0, "l": [0, 1, 7, 17], "": [0, 1, 8], "makeblastdb": 0, "parse_seqid": 0, "dbtype": 0, "nucl": 0, "out": 0, "after": [0, 1], "command": [0, 1, 5, 6, 11], "mani": [0, 5], "more": [0, 3, 5, 12, 17], "nhr": 0, "nj": 0, "nos": 0, "nsq": 0, "nto": 0, "ndb": 0, "nin": 0, "nog": 0, "ntf": 0, "verbos": [0, 2], "7": [0, 5, 17], "std": [0, 5], "qseq": [0, 5], "sseq": [0, 5], "btop": [0, 5], "igdata": 0, "environ": 0, "variabl": [0, 13], "fact": 0, "overrid": 0, "them": [0, 5, 8], "line": [0, 4, 5, 11, 12], "order": 0, "export": [0, 1], "usr": 0, "share": 0, "igblastn": 0, "germline_db_v": 0, "germline_db_d": 0, "germline_db_j": 0, "auxiliary_data": [0, 5], "custom_internal_data": [0, 5], "domain_system": 0, "imgt": [0, 4, 5, 7, 8, 10, 12, 13, 16, 17], "outfmt": [0, 5], "num_thread": 0, "queri": 0, "fmt7": 0, "examin": [0, 5], "witn": 0, "anoth": [0, 5, 6], "confirm": [0, 8], "makedb": [0, 5], "tsv": [0, 1], "note": [0, 6, 17], "fail": [0, 5], "captur": 0, "detail": [0, 5, 6, 12], "ani": [0, 5, 6, 9, 12, 14, 17], "import": [0, 5], "some": [0, 1, 5, 13], "guidanc": 0, "error": [0, 5, 6], "py": [0, 5], "r": [0, 3, 7, 16], "extend": [0, 5, 9], "should": [0, 4, 5, 16, 17], "similar": [0, 5], "hd13m_db": 0, "pass": [0, 5], "7887": 0, "1402": 0, "review": [0, 13], "result": [0, 1, 3, 5], "case": [1, 5, 12, 13, 17], "just": [1, 5, 17], "walkthrough": 1, "immcant": 1, "file": [1, 3, 4, 5, 6, 7, 8, 9, 10, 12, 13, 14, 17], "addit": [5, 6, 8, 12], "igblast": [2, 6, 11, 13], "format": [1, 2, 4, 5, 12, 13, 17], "work": [1, 12], "ghcr": 1, "io": 1, "milaboratori": 1, "latest": [1, 6, 7], "4": [], "licens": 1, "match": [1, 5, 14], "kei": [1, 17], "mi_licens": 1, "licenc": 1, "buildlibrari": 1, "debug": [1, 17], "featur": [1, 8], "vregion": 1, "j": [1, 2, 4, 5, 6, 17], "d": [1, 4, 5, 6, 17], "chain": [1, 8, 13, 17], "taxon": 1, "id": [1, 12, 14, 17], "json": [1, 2, 6, 13], "gz": 1, "mai": [1, 4, 5, 8, 13, 17], "warn": [1, 5, 9, 12], "dure": 1, "stop": [1, 5], "codon": [1, 4, 5, 8, 12, 13, 17], "were": [1, 5, 17], "found": [1, 5, 6], "expect": [1, 5, 8], "pseudogen": [1, 5], "includ": [1, 4, 5, 6, 12, 17], "c": [1, 2, 4, 6, 13], "No": 1, "action": 1, "As": [1, 4], "plain": 1, "gener": 1, "pacbio": 1, "templat": [1, 9, 12], "analyz": 1, "librari": 1, "assembl": 1, "clonotyp": 1, "fr1": 1, "cdr1": [1, 4, 13], "fr2": 1, "cdr2": [1, 4, 13], "fr3": 1, "cdr3": [1, 4, 13], "fr4": 1, "view": 1, "clones_igh": 1, "annot": [2, 5, 6, 12], "set": [2, 4, 5, 6, 8, 9, 11, 12, 13, 16, 17], "sequenc": [2, 3, 4, 5, 6, 8, 9, 10, 13, 14, 16, 17, 18], "find": [2, 5, 12, 17], "frame": [2, 5], "align": [2, 4, 5, 8, 13, 17], "1": [2, 3, 4, 5, 13, 14, 17], "base": [2, 3, 4, 13], "index": [2, 11, 17], "first": [2, 14, 17], "nucleotid": [2, 4, 12, 17, 18], "conserv": [2, 8, 12], "phe": 2, "trp": 2, "usag": [2, 3, 4, 6, 7, 8, 9, 10, 13, 14, 16, 18], "h": [2, 3, 4, 6, 7, 8, 9, 10, 13, 14, 16, 18], "seq_fil": 2, "out_fil": [2, 9], "airr": [2, 6, 11, 13], "output": [2, 4, 5, 9, 13, 14, 17], "us": [2, 4, 6, 8, 9, 11, 13, 16, 17], "default": [2, 3, 4, 6, 7, 13, 16, 17], "fals": [2, 3, 7, 16, 17], "specifi": [3, 4, 5, 6, 7, 13, 16, 17], "co": [3, 8, 12, 13], "ordin": [3, 8, 12, 13], "n": [3, 6, 17], "xr": 3, "input_fil": [3, 4, 10], "contain": [3, 4, 5, 6, 8, 9, 12, 13, 14, 16, 17], "refer": [3, 5, 6, 7, 9, 10, 11, 12, 13, 16, 17], "coord": 3, "label": 3, "must": [3, 12, 13, 17], "than": [3, 5, 12, 14], "one": [3, 5, 12, 13, 17], "revers": [3, 12, 16, 17, 18], "complement": [3, 12, 16, 17, 18], "instal": 5, "can": [4, 5, 6, 7, 9, 12, 13], "configur": [4, 5, 6], "support": [4, 5, 12], "organ": 5, "either": [5, 17], "an": [4, 5, 6, 7, 13, 16, 17], "tool": [4, 5, 6, 8, 9, 13], "modifi": [5, 17], "simplif": 5, "upgrad": 5, "least": [5, 9], "encourag": 5, "full": [5, 6, 17], "two": [5, 12, 14], "germlin": [5, 6, 12, 13], "gene": [4, 5, 7, 16, 17], "orient": 5, "posit": [5, 17], "cdr": [5, 6, 8, 12, 13, 17], "wish": [5, 8], "add": [5, 17], "allel": [4, 5, 12, 16, 17], "exist": 5, "all": [5, 6, 13, 14, 17], "blast": 5, "link": 5, "new": 5, "paramet": [5, 17], "make_igblast_ndm": [5, 6, 12, 19], "gap": [4, 5, 6, 7, 8, 9, 10, 12, 13, 16, 17], "gap_sequ": [5, 12, 15, 17, 19], "partial": 5, "close": [5, 17], "relat": 5, "annotate_j": [5, 6, 12, 19], "It": [5, 12], "veri": 5, "carefulli": 5, "when": 5, "issu": [5, 9], "particularli": 5, "substanti": 5, "proport": 5, "submit": 5, "my": [5, 12], "own": 5, "repertoir": [5, 8], "convert": [5, 17], "changeo": 5, "record": [5, 14, 17], "log": 5, "count": [5, 17], "typic": 5, "might": 5, "under": [5, 12], "10": 5, "possibli": 5, "fewer": 5, "A": [5, 8, 13, 17], "higher": 5, "sign": 5, "db_fail": 5, "discount": 5, "what": 5, "do": [5, 13, 14, 17], "thei": [5, 9, 12], "repres": 5, "few": 5, "percent": 5, "obvious": 5, "short": 5, "length": [5, 17], "again": 5, "take": [4, 5], "sampl": [5, 13, 17], "look": 5, "though": 5, "analys": [5, 13, 17], "quest": [5, 13], "interest": 5, "doe": [5, 17], "non": [5, 9], "function": [5, 9, 12, 17], "number": [4, 5, 8, 13, 17], "strongli": 5, "suspect": 5, "problem": [5, 8, 12], "report": [5, 12], "analysi": [5, 13], "mi": 5, "junction": 5, "assign": 5, "mean": [5, 6], "inde": 5, "part": 5, "miss": [5, 12], "There": 5, "simpli": 5, "too": 5, "cover": [5, 12, 17], "entir": [5, 17], "like": [5, 13], "incorrectli": 5, "empti": 5, "310": 5, "occur": 5, "corrupt": 5, "indic": 5, "circumst": [5, 9], "recognis": 5, "being": 5, "canon": 5, "someth": [5, 8], "along": 5, "ighv1": [4, 5], "69": 5, "02": 5, "latitud": 5, "allow": 5, "name": [5, 17], "depart": 5, "caus": [5, 8, 12], "particular": [5, 17], "without": [4, 5, 6, 8], "suffix": 5, "identifi": [5, 7, 12], "resolv": 5, "excel": 5, "reason": 5, "famili": 5, "arrang": 5, "investig": 5, "except": 5, "continu": 5, "exercis": 5, "caution": 5, "consciou": 5, "hope": 5, "doesn": 5, "t": [5, 17], "put": 5, "off": 5, "But": 5, "serv": 5, "systemat": 5, "lookout": 5, "well": [5, 17], "flag": 5, "also": [5, 6, 7], "understand": 5, "underli": 5, "download": [6, 7, 12], "databas": [6, 11], "airrc": 6, "singl": [6, 16, 17], "fg": 6, "fu": 6, "u": 6, "url": 6, "p": 6, "prefix": 6, "locu": [6, 7, 16], "igk": [6, 7, 8], "igl": [6, 8], "tra": 6, "trb": [6, 7, 16], "trd": 6, "attempt": 6, "none": [6, 17], "specif": [4, 6, 8, 12, 13], "otherwis": [6, 17], "choic": [4, 6], "commun": [6, 11], "filenam": [6, 17], "species_locu": 6, "_": 6, "substitut": 6, "space": [6, 17], "multipl": 6, "separ": [4, 6, 12, 13], "plu": 6, "By": [4, 6, 13], "advantag": 6, "inform": 6, "delin": [6, 12, 13], "delieat": 6, "principl": [6, 17], "load": 6, "recommend": [6, 13], "ungap": [6, 9, 17], "respect": 6, "_gap": 6, "replac": [6, 17], "four": 6, "These": [6, 8, 12], "_v": 6, "_d": 6, "_j": 6, "_v_gap": 6, "multi_f": 6, "human": [4, 6], "mous": 6, "c57bl": 6, "6": 6, "mu": 6, "musculu": 6, "try": 6, "help": 6, "event": 6, "10090": 6, "avail": [6, 13], "balb": 6, "ighv": [6, 8, 17], "cast": 6, "eij": 6, "lew": 6, "msm": 6, "msj": 6, "nod": 6, "shiltj": 6, "phj": 6, "airrc_igh": 6, "nomin": 7, "ref": [7, 17], "species_nam": 7, "quot": 7, "macaca": 7, "mulatta": 7, "www": 7, "db": 7, "imgtgenedb": 7, "referencesequ": 7, "nt": 7, "withgap": 7, "orf": 7, "inframep": 7, "trg": 7, "functional_onli": [7, 17], "rhesu": [8, 13], "macaqu": [8, 12, 13], "iglv": 8, "igkv": 8, "insert": [8, 12, 13], "rel": [8, 12], "most": [8, 9, 12], "downstream": [8, 12, 13], "chang": [8, 12], "o": 8, "residu": [8, 12], "restor": 8, "standard": [8, 12], "remov": [8, 12, 13], "list": [4, 8, 10, 12, 13, 17], "those": [8, 9, 13], "person": 8, "observ": 8, "produc": [4, 8, 17], "unwant": 8, "ig": [4, 8, 12], "infil": [8, 17], "outfil": [8, 14, 17], "customari": 8, "closest": [9, 12, 17], "onli": [4, 9, 16, 17], "complet": 9, "5": [4, 9, 17], "second": [9, 14, 17], "cystein": 9, "neither": 9, "appar": 9, "given": [9, 12, 17], "fraction": 9, "total": 9, "ignor": 9, "inferred_fil": 9, "ref_fil": [9, 13], "infer": 9, "ident": 10, "sub": 10, "identical_sequ": 10, "mixcr": 11, "receptor_util": 11, "modul": [11, 12, 15], "search": 11, "page": 11, "code": 12, "tr": [4, 12], "pip": 12, "biopython": [12, 17], "name_allel": [12, 19], "nearest": 12, "taken": 12, "suppli": 12, "incorpor": 12, "snp": [12, 16], "scheme": [12, 13], "accord": 12, "emploi": 12, "each": [12, 13], "fix_macaque_gap": [12, 13, 19], "revert": 12, "next": 12, "extract_ref": [12, 16, 19], "identical_seq": [12, 19], "where": [12, 17], "same": [4, 12, 17], "differ": [4, 12, 17], "rev_comp": [12, 16, 19], "at_coord": [12, 19], "within": 12, "merge_fasta": [12, 19], "merg": [12, 14], "duplic": 12, "simpl": [12, 17], "wrapper": [12, 17], "around": [12, 17], "commonli": 12, "write": [12, 17], "comma": [4, 12, 13], "tab": 12, "manag": 12, "string": [12, 17], "dict": [12, 17], "compar": [12, 14, 16], "expens": 12, "flexibl": 12, "rare": 12, "region": [4, 12, 13, 17], "cdr_coord": 13, "ndm_file": 13, "eg": 13, "vh": 13, "79": 13, "114": 13, "166": 13, "195": 13, "313": 13, "ensur": [13, 17], "correct": [4, 13], "placement": 13, "coordin": [4, 13], "assum": 13, "doubt": 13, "denot": 13, "b": 13, "etc": 13, "upon": 13, "applic": [13, 17], "we": [13, 17], "offer": 13, "conduct": 13, "realign": 13, "provis": 13, "file1": 14, "file2": 14, "themselv": 14, "rather": 14, "keep": 14, "2": [1, 14], "file_1": 14, "file_2": 14, "submodul": 15, "novel_allele_nam": 15, "aligned_diff": [15, 17], "build_ambiguous_ref": [15, 17], "closest_aligned_ref": [15, 17], "find_indel": [15, 17], "name_novel": [15, 17], "run_test": [15, 17], "number_v": 15, "check_conserved_residu": [15, 17], "distribute_cdr": [15, 17], "gap_align": [15, 17], "gap_align_aa": [15, 17], "gap_align_aa_from_nt": [15, 17], "gap_nt_from_aa": [15, 17], "insert_spac": [15, 17], "match_scor": [15, 17], "nt_diff": [15, 17], "number_from_trp": [15, 17], "number_ighv": [15, 17], "pretty_gap": [15, 17], "simple_bio_seq": 15, "chunk": [15, 17], "closest_ref": [15, 17], "dumb_consensu": [15, 17], "read_csv": [15, 17], "read_fasta": [15, 17], "read_imgt_fasta": [15, 17], "read_single_fasta": [15, 17], "reverse_compl": [15, 17], "sample_fasta": [15, 17], "scored_consensu": [15, 17], "scored_gap_consensu": [15, 17], "toseqrecord": [15, 17], "translat": [4, 12, 15, 17], "write_csv": [15, 17], "write_fasta": [15, 17], "content": 15, "construct": 16, "novel": [16, 17], "gene_nam": 16, "ref_set": [16, 17], "input": [4, 16, 17, 18], "consid": 16, "novel_seq": 17, "str": 17, "ref_seq": 17, "full_ref": 17, "seq": 17, "entri": 17, "global": 17, "return": [17, 18], "equal": 17, "type": [4, 12, 17], "closest_ref_seq": 17, "v_gene": 17, "bool": 17, "true": 17, "its": 17, "conform": 17, "descript": 17, "prime": 17, "tupl": 17, "consist": [4, 17], "three": 17, "novel_nam": 17, "aa": 17, "aa_seq": 17, "nt_gap": 17, "nucleotide_seq": 17, "peptide_seq": 17, "gapped_ref": 17, "ungapped_ref": 17, "po": 17, "target": 17, "match_list": 17, "thresh": 17, "s1": 17, "s2": 17, "next_p": 17, "trp_po": 17, "bioseq": 17, "store": 17, "collectiopn": 17, "coerc": 17, "upper": 17, "iter": 17, "corec": 17, "eas": 17, "basic": 17, "thing": 17, "don": 17, "int": 17, "yield": 17, "success": 17, "size": 17, "threshold": 17, "float": 17, "dumb": 17, "consensu": 17, "bio": 17, "aligninfo": 17, "summaryinfo": 17, "valu": 17, "atom": 17, "shorter": [4, 17], "delimit": 17, "dictread": 17, "pathnam": 17, "dictionari": 17, "ighd": 17, "ighj": 17, "ch": 17, "include_orphon": 17, "underscor": 17, "mark": [4, 17], "orphon": 17, "serquenc": 17, "nucelotid": 17, "random": 17, "resampl": 17, "valueerror": 17, "enough": 17, "gap_consensu": 17, "mimimum": 17, "score": 17, "achiev": 17, "min_scor": 17, "ambigu": 17, "x": 17, "require_multipl": 17, "adapt": 17, "seqrecord": 17, "truncat": 17, "ignore_partial_codon": 17, "amino": [4, 12, 17], "acid": [4, 12, 17], "so": 17, "termin": 17, "boundari": [4, 17], "pad": 17, "get": 17, "row": 17, "scan_al": 17, "header": 17, "item": 17, "scan": 17, "written": 17, "displai": [4, 12], "readabl": 4, "show": [4, 12], "both": 4, "automat": 4, "span": [], "due": [], "wrap": 4, "appropri": [], "across": [], "sort": 4, "alphabet": 4, "codon_wrap": 4, "v_coord": 4, "output_fil": 4, "sequence_typ": 4, "implement": 4, "w": 4, "per": 4, "27": 4, "38": 4, "56": 4, "65": 4, "105": 4, "substr": 4, "18": 4, "txt": 4, "d_sequenc": 4, "d_align": 4, "25": 4, "30": 4, "54": 4, "60": 4, "100": 4, "15": 4, "pattern": 4, "create_align": [12, 19], "silent": 4, "nonsil": 4, "mutat": 4, "becaus": 4, "occupi": 4, "user": 4, "would": 4}, "objects": {"": [[17, 0, 0, "-", "receptor_utils"]], "receptor_utils": [[17, 0, 0, "-", "novel_allele_name"], [17, 0, 0, "-", "number_v"], [17, 0, 0, "-", "simple_bio_seq"]], "receptor_utils.novel_allele_name": [[17, 1, 1, "", "aligned_diff"], [17, 1, 1, "", "build_ambiguous_ref"], [17, 1, 1, "", "closest_aligned_ref"], [17, 1, 1, "", "find_indels"], [17, 1, 1, "", "name_novel"], [17, 1, 1, "", "run_tests"]], "receptor_utils.number_v": [[17, 1, 1, "", "check_conserved_residues"], [17, 1, 1, "", "distribute_cdr"], [17, 1, 1, "", "gap_align"], [17, 1, 1, "", "gap_align_aa"], [17, 1, 1, "", "gap_align_aa_from_nt"], [17, 1, 1, "", "gap_nt_from_aa"], [17, 1, 1, "", "gap_sequence"], [17, 1, 1, "", "insert_space"], [17, 1, 1, "", "match_score"], [17, 1, 1, "", "nt_diff"], [17, 1, 1, "", "number_from_trp"], [17, 1, 1, "", "number_ighv"], [17, 1, 1, "", "pretty_gapped"], [17, 1, 1, "", "run_tests"]], "receptor_utils.simple_bio_seq": [[17, 1, 1, "", "chunks"], [17, 1, 1, "", "closest_ref"], [17, 1, 1, "", "dumb_consensus"], [17, 1, 1, "", "nt_diff"], [17, 1, 1, "", "read_csv"], [17, 1, 1, "", "read_fasta"], [17, 1, 1, "", "read_imgt_fasta"], [17, 1, 1, "", "read_single_fasta"], [17, 1, 1, "", "reverse_complement"], [17, 1, 1, "", "sample_fasta"], [17, 1, 1, "", "scored_consensus"], [17, 1, 1, "", "scored_gap_consensus"], [17, 1, 1, "", "toSeqRecords"], [17, 1, 1, "", "translate"], [17, 1, 1, "", "write_csv"], [17, 1, 1, "", "write_fasta"]]}, "objtypes": {"0": "py:module", "1": "py:function"}, "objnames": {"0": ["py", "module", "Python module"], "1": ["py", "function", "Python function"]}, "titleterms": {"us": [0, 1, 5, 12], "airr": [0, 1], "commun": [0, 1], "refer": [0, 1], "set": [0, 1], "igblast": [0, 5, 12, 19], "prerequisit": [0, 1], "1": [0, 1], "download": [0, 1], "germlin": [0, 1], "human": [0, 1], "ighv": [0, 1], "gene": [0, 1], "addit": 0, "file": 0, "2": 0, "log": [0, 1], "immcant": 0, "contain": [0, 1], "so": [0, 1], "we": [0, 1], "can": [0, 1], "its": [0, 1], "instal": [0, 1, 12], "tool": [0, 1, 12, 19], "3": [0, 1], "build": [0, 1], "databas": [0, 1, 5, 12], "4": [0, 1], "annot": [0, 1], "sampl": [0, 1], "sequenc": [0, 1, 11, 12, 19], "5": 0, "convert": 0, "output": 0, "chang": 0, "o": 0, "format": [0, 6], "mixcr": 1, "annotate_j": 2, "posit": [2, 3, 4, 6, 7, 8, 9, 10, 13, 14, 16, 18], "argument": [2, 3, 4, 6, 7, 8, 9, 10, 13, 14, 16, 18], "name": [2, 3, 4, 6, 7, 12, 13, 16, 19], "at_coord": 3, "custom": [5, 12], "download_germline_set": 6, "option": 6, "exampl": [4, 6], "extract_ref": 7, "fix_macaque_gap": 8, "gap_sequ": 9, "identical_seq": 10, "receptor": 11, "util": [11, 19], "code": 11, "work": [11, 19], "ig": 11, "tr": 11, "content": [11, 17], "indic": 11, "tabl": 11, "introduct": 12, "align": [12, 19], "v": [12, 19], "d": [12, 19], "j": [12, 19], "conveni": [12, 19], "simple_bio_seq": [12, 17], "api": 12, "make_igblast_ndm": 13, "note": 13, "merge_fasta": 14, "receptor_util": [15, 17], "name_allel": 16, "packag": 17, "submodul": 17, "novel_allele_nam": 17, "modul": 17, "number_v": 17, "rev_comp": 18, "command": 19, "line": 19, "create_align": 4, "cdr": 4, "delin": 4}, "envversion": {"sphinx.domains.c": 2, "sphinx.domains.changeset": 1, "sphinx.domains.citation": 1, "sphinx.domains.cpp": 8, "sphinx.domains.index": 1, "sphinx.domains.javascript": 2, "sphinx.domains.math": 2, "sphinx.domains.python": 3, "sphinx.domains.rst": 2, "sphinx.domains.std": 2, "sphinx.ext.viewcode": 1, "sphinx": 57}, "alltitles": {"Using AIRR Community Reference Sets with IgBLAST": [[0, "using-airr-community-reference-sets-with-igblast"]], "Prerequisites": [[0, "prerequisites"], [1, "prerequisites"]], "1. Download the germline set for human IGHV genes and additional IgBLAST files": [[0, "download-the-germline-set-for-human-ighv-genes-and-additional-igblast-files"]], "2. Log in to the Immcantation container so that we can use its installed tools": [[0, "log-in-to-the-immcantation-container-so-that-we-can-use-its-installed-tools"]], "3. Build IgBLAST databases for the germline set": [[0, "build-igblast-databases-for-the-germline-set"]], "4. Annotate the sample sequences using IgBLAST": [[0, "annotate-the-sample-sequences-using-igblast"]], "5. Convert the IgBLAST output to Change-O format": [[0, "convert-the-igblast-output-to-change-o-format"]], "Using AIRR Community Reference Sets with MiXCR": [[1, "using-airr-community-reference-sets-with-mixcr"]], "1. Download the germline set for human IGHV genes": [[1, "download-the-germline-set-for-human-ighv-genes"]], "1. Log in to the MiXCR container so that we can use its installed tools": [[1, "log-in-to-the-mixcr-container-so-that-we-can-use-its-installed-tools"]], "3. Build MiXCR database for the germline set": [[1, "build-mixcr-database-for-the-germline-set"]], "4. Annotate the sample sequences using MiXCR": [[1, "annotate-the-sample-sequences-using-mixcr"]], "annotate_j": [[2, "annotate-j"]], "Positional Arguments": [[2, "positional-arguments"], [3, "positional-arguments"], [4, "positional-arguments"], [6, "positional-arguments"], [7, "positional-arguments"], [8, "positional-arguments"], [9, "positional-arguments"], [10, "positional-arguments"], [13, "positional-arguments"], [14, "positional-arguments"], [16, "positional-arguments"], [18, "positional-arguments"]], "Named Arguments": [[2, "named-arguments"], [3, "named-arguments"], [4, "named-arguments"], [6, "named-arguments"], [7, "named-arguments"], [13, "named-arguments"], [16, "named-arguments"]], "at_coords": [[3, "at-coords"]], "create_alignment": [[4, "create-alignment"]], "CDR delineation": [[4, "cdr-delineation"]], "Examples": [[4, "examples"], [6, "examples"]], "Using custom databases with IgBlast": [[5, "using-custom-databases-with-igblast"], [12, "using-custom-databases-with-igblast"]], "download_germline_set": [[6, "download-germline-set"]], "Format Options": [[6, "format-options"]], "extract_refs": [[7, "extract-refs"]], "fix_macaque_gaps": [[8, "fix-macaque-gaps"]], "gap_sequences": [[9, "gap-sequences"]], "identical_seqs": [[10, "identical-seqs-label"]], "Receptor-utils: code and utilities for working with IG/TR receptor sequences": [[11, "receptor-utils-code-and-utilities-for-working-with-ig-tr-receptor-sequences"]], "Contents:": [[11, null]], "Indices and tables": [[11, "indices-and-tables"]], "Introduction": [[12, "introduction"]], "Installation": [[12, "installation"]], "Naming and aligning V, D and J sequences": [[12, "naming-and-aligning-v-d-and-j-sequences"]], "Convenience Tools": [[12, "convenience-tools"], [19, "convenience-tools"]], "simple_bio_seq API": [[12, "simple-bio-seq-api"]], "make_igblast_ndm": [[13, "make-igblast-ndm"]], "Notes": [[13, "notes"]], "merge_fasta": [[14, "merge-fasta"]], "receptor_utils": [[15, "receptor-utils"]], "name_allele": [[16, "name-allele"]], "receptor_utils package": [[17, "receptor-utils-package"]], "Submodules": [[17, "submodules"]], "receptor_utils.novel_allele_name module": [[17, "module-receptor_utils.novel_allele_name"]], "receptor_utils.number_v module": [[17, "module-receptor_utils.number_v"]], "receptor_utils.simple_bio_seq module": [[17, "module-receptor_utils.simple_bio_seq"]], "Module contents": [[17, "module-receptor_utils"]], "rev_comp": [[18, "rev-comp"]], "Command-Line Utilities": [[19, "command-line-utilities"]], "Naming and Aligning V, D and J Sequences": [[19, "naming-and-aligning-v-d-and-j-sequences"]], "Working with IgBLAST": [[19, "working-with-igblast"]]}, "indexentries": {"aligned_diff() (in module receptor_utils.novel_allele_name)": [[17, "receptor_utils.novel_allele_name.aligned_diff"]], "build_ambiguous_ref() (in module receptor_utils.novel_allele_name)": [[17, "receptor_utils.novel_allele_name.build_ambiguous_ref"]], "check_conserved_residues() (in module receptor_utils.number_v)": [[17, "receptor_utils.number_v.check_conserved_residues"]], "chunks() (in module receptor_utils.simple_bio_seq)": [[17, "receptor_utils.simple_bio_seq.chunks"]], "closest_aligned_ref() (in module receptor_utils.novel_allele_name)": [[17, "receptor_utils.novel_allele_name.closest_aligned_ref"]], "closest_ref() (in module receptor_utils.simple_bio_seq)": [[17, "receptor_utils.simple_bio_seq.closest_ref"]], "distribute_cdr() (in module receptor_utils.number_v)": [[17, "receptor_utils.number_v.distribute_cdr"]], "dumb_consensus() (in module receptor_utils.simple_bio_seq)": [[17, "receptor_utils.simple_bio_seq.dumb_consensus"]], "find_indels() (in module receptor_utils.novel_allele_name)": [[17, "receptor_utils.novel_allele_name.find_indels"]], "gap_align() (in module receptor_utils.number_v)": [[17, "receptor_utils.number_v.gap_align"]], "gap_align_aa() (in module receptor_utils.number_v)": [[17, "receptor_utils.number_v.gap_align_aa"]], "gap_align_aa_from_nt() (in module receptor_utils.number_v)": [[17, "receptor_utils.number_v.gap_align_aa_from_nt"]], "gap_nt_from_aa() (in module receptor_utils.number_v)": [[17, "receptor_utils.number_v.gap_nt_from_aa"]], "gap_sequence() (in module receptor_utils.number_v)": [[17, "receptor_utils.number_v.gap_sequence"]], "insert_space() (in module receptor_utils.number_v)": [[17, "receptor_utils.number_v.insert_space"]], "match_score() (in module receptor_utils.number_v)": [[17, "receptor_utils.number_v.match_score"]], "module": [[17, "module-receptor_utils"], [17, "module-receptor_utils.novel_allele_name"], [17, "module-receptor_utils.number_v"], [17, "module-receptor_utils.simple_bio_seq"]], "name_novel() (in module receptor_utils.novel_allele_name)": [[17, "receptor_utils.novel_allele_name.name_novel"]], "nt_diff() (in module receptor_utils.number_v)": [[17, "receptor_utils.number_v.nt_diff"]], "nt_diff() (in module receptor_utils.simple_bio_seq)": [[17, "receptor_utils.simple_bio_seq.nt_diff"]], "number_from_trp() (in module receptor_utils.number_v)": [[17, "receptor_utils.number_v.number_from_trp"]], "number_ighv() (in module receptor_utils.number_v)": [[17, "receptor_utils.number_v.number_ighv"]], "pretty_gapped() (in module receptor_utils.number_v)": [[17, "receptor_utils.number_v.pretty_gapped"]], "read_csv() (in module receptor_utils.simple_bio_seq)": [[17, "receptor_utils.simple_bio_seq.read_csv"]], "read_fasta() (in module receptor_utils.simple_bio_seq)": [[17, "receptor_utils.simple_bio_seq.read_fasta"]], "read_imgt_fasta() (in module receptor_utils.simple_bio_seq)": [[17, "receptor_utils.simple_bio_seq.read_imgt_fasta"]], "read_single_fasta() (in module receptor_utils.simple_bio_seq)": [[17, "receptor_utils.simple_bio_seq.read_single_fasta"]], "receptor_utils": [[17, "module-receptor_utils"]], "receptor_utils.novel_allele_name": [[17, "module-receptor_utils.novel_allele_name"]], "receptor_utils.number_v": [[17, "module-receptor_utils.number_v"]], "receptor_utils.simple_bio_seq": [[17, "module-receptor_utils.simple_bio_seq"]], "reverse_complement() (in module receptor_utils.simple_bio_seq)": [[17, "receptor_utils.simple_bio_seq.reverse_complement"]], "run_tests() (in module receptor_utils.novel_allele_name)": [[17, "receptor_utils.novel_allele_name.run_tests"]], "run_tests() (in module receptor_utils.number_v)": [[17, "receptor_utils.number_v.run_tests"]], "sample_fasta() (in module receptor_utils.simple_bio_seq)": [[17, "receptor_utils.simple_bio_seq.sample_fasta"]], "scored_consensus() (in module receptor_utils.simple_bio_seq)": [[17, "receptor_utils.simple_bio_seq.scored_consensus"]], "scored_gap_consensus() (in module receptor_utils.simple_bio_seq)": [[17, "receptor_utils.simple_bio_seq.scored_gap_consensus"]], "toseqrecords() (in module receptor_utils.simple_bio_seq)": [[17, "receptor_utils.simple_bio_seq.toSeqRecords"]], "translate() (in module receptor_utils.simple_bio_seq)": [[17, "receptor_utils.simple_bio_seq.translate"]], "write_csv() (in module receptor_utils.simple_bio_seq)": [[17, "receptor_utils.simple_bio_seq.write_csv"]], "write_fasta() (in module receptor_utils.simple_bio_seq)": [[17, "receptor_utils.simple_bio_seq.write_fasta"]]}})