Search.setIndex({"docnames": ["annotate_j", "at_coords", "custom_igblast", "extract_refs", "fix_macaque_gaps", "gap_sequences", "identical_seqs", "index", "introduction", "make_igblast_ndm", "merge_fasta", "modules", "name_allele", "receptor_utils", "rev_comp", "scripts"], "filenames": ["annotate_j.rst", "at_coords.rst", "custom_igblast.rst", "extract_refs.rst", "fix_macaque_gaps.rst", "gap_sequences.rst", "identical_seqs.rst", "index.rst", "introduction.rst", "make_igblast_ndm.rst", "merge_fasta.rst", "modules.rst", "name_allele.rst", "receptor_utils.rst", "rev_comp.rst", "scripts.rst"], "titles": ["annotate_j", "at_coords", "Using custom databases with IgBlast", "extract_refs", "fix_macaque_gaps", "gap_sequences", "identical_seqs", "Receptor-utils: code and utilities for working with IG/TR receptor sequences", "Introduction", "make_igblast_ndm", "merge_fasta", "API", "name_allele", "receptor_utils package", "rev_comp", "Command-Line Utilities"], "terms": {"annot": [0, 2], "set": [0, 2, 4, 5, 8, 9, 12, 13], "j": [0, 2, 13], "sequenc": [0, 1, 2, 4, 5, 6, 10, 12, 13, 14], "find": [0, 2, 8, 13], "frame": [0, 2], "align": [0, 2, 4, 13], "1": [0, 1, 2, 10, 13], "base": [0, 1], "index": [0, 7, 13], "first": [0, 10, 13], "nucleotid": [0, 8, 13, 14], "conserv": [0, 4, 8], "phe": 0, "trp": 0, "usag": [0, 1, 3, 4, 5, 6, 9, 10, 12, 14], "h": [0, 1, 3, 4, 5, 6, 9, 10, 12, 14], "seq_fil": 0, "out_fil": [0, 5], "fasta": [0, 1, 3, 4, 5, 6, 8, 9, 10, 12, 13], "output": [0, 2, 5, 9, 10, 13], "aux": [0, 2], "format": [0, 2, 9, 13], "us": [0, 4, 5, 7, 12, 13], "igblast": [0, 7, 9], "extract": [1, 3], "specifi": [1, 2, 3, 12, 13], "co": [1, 4, 8], "ordin": [1, 4, 8], "file": [1, 2, 3, 4, 5, 6, 8, 9, 10, 13], "r": [1, 3, 12], "input_fil": [1, 6], "start": [1, 2, 13], "end": [1, 5, 13], "contain": [1, 2, 4, 5, 8, 10, 12, 13], "refer": [1, 2, 3, 5, 6, 8, 9, 12, 13], "ord": 1, "revers": [1, 8, 12, 13, 14], "complement": [1, 8, 12, 13, 14], "result": [1, 2], "default": [1, 3, 12, 13], "fals": [1, 3, 12, 13], "local": 2, "instal": 2, "can": [2, 3, 5, 8], "configur": 2, "support": [2, 8], "organ": 2, "either": [2, 13], "an": [2, 3, 4, 9, 12, 13], "which": [2, 4, 8, 13], "ar": [2, 5, 8, 10, 13], "provid": [2, 8, 13], "tool": [2, 4, 5], "modifi": [2, 13], "The": [2, 3, 5, 8, 10, 13], "procedur": 2, "i": [2, 4, 5, 8, 13], "significantli": 2, "simplif": 2, "version": 2, "20": 2, "upgrad": 2, "least": [2, 5], "thi": [2, 3, 4, 8, 10, 13], "encourag": 2, "For": [2, 8], "full": [2, 13], "two": [2, 8, 10], "requir": [2, 8, 9, 13], "addit": [2, 4, 8], "germlin": [2, 8], "auxiliary_data": 2, "describ": [2, 8], "detail": [2, 8], "gene": [2, 3, 12, 13], "orient": 2, "ndm": [2, 8, 9], "posit": [2, 13], "cdr": [2, 4, 13], "v": [2, 12, 13], "If": [2, 3, 13], "you": [2, 4], "just": [2, 13], "wish": [2, 4], "add": [2, 13], "allel": [2, 8, 12, 13], "exist": 2, "d": [2, 13], "all": [2, 10, 13], "necessari": [2, 12, 13], "creat": [2, 8, 12, 13], "blast": 2, "link": 2, "abov": [2, 8], "case": [2, 4, 8, 13], "need": [2, 8, 13], "new": 2, "custom_internal_data": 2, "command": [2, 7], "line": [2, 7, 8], "paramet": [2, 13], "make_igblast_ndm": [2, 8, 15], "make": [2, 9, 13], "from": [2, 3, 4, 8, 9, 10, 12, 13], "imgt": [2, 3, 4, 6, 8, 9, 12, 13], "gap": [2, 3, 4, 5, 6, 8, 9, 12, 13], "should": [2, 12, 13], "includ": [2, 8, 13], "them": [2, 4], "gap_sequ": [2, 8, 11, 13, 15], "have": [2, 4], "partial": 2, "match": [2, 10], "your": 2, "close": [2, 13], "relat": 2, "annotate_j": [2, 8, 15], "It": [2, 8], "veri": 2, "import": 2, "test": 2, "carefulli": 2, "when": 2, "issu": [2, 5], "particularli": 2, "mai": [2, 4, 13], "fail": 2, "substanti": 2, "proport": 2, "submit": 2, "my": [2, 8], "own": 2, "approach": 2, "follow": 2, "repertoir": [2, 4], "outfmt": 2, "7": [2, 13], "std": 2, "qseq": 2, "sseq": 2, "btop": 2, "convert": [2, 13], "changeo": 2, "makedb": 2, "option": [2, 13], "record": [2, 10, 13], "log": 2, "examin": 2, "count": [2, 13], "pass": 2, "typic": 2, "might": 2, "expect": [2, 4], "under": [2, 8], "10": 2, "possibli": 2, "mani": 2, "fewer": 2, "than": [2, 8, 10], "depend": 2, "qualiti": 2, "read": [2, 8, 13], "A": [2, 4, 13], "higher": 2, "warn": [2, 5, 8], "sign": 2, "open": 2, "db_fail": 2, "were": [2, 13], "found": 2, "stop": 2, "codon": [2, 4, 8, 13], "discount": 2, "how": 2, "other": [2, 4, 5, 8, 13], "what": 2, "do": [2, 10, 13], "thei": [2, 5, 8], "repres": 2, "more": [2, 8, 13], "few": 2, "percent": 2, "obvious": 2, "short": 2, "e": [2, 3, 12, 13], "length": [2, 13], "again": 2, "take": 2, "sampl": [2, 13], "look": 2, "though": 2, "analys": [2, 13], "quest": 2, "interest": 2, "one": [2, 8, 13], "doe": [2, 13], "non": [2, 5], "function": [2, 5, 8, 13], "number": [2, 4, 13], "strongli": 2, "suspect": 2, "problem": [2, 4, 8], "error": 2, "check": [2, 4], "report": [2, 8], "ani": [2, 4, 5, 8, 10, 13], "analysi": 2, "been": 2, "mi": 2, "see": 2, "junction": 2, "assign": 2, "mean": 2, "inde": 2, "part": 2, "miss": [2, 8], "There": 2, "simpli": 2, "too": 2, "cover": [2, 8, 13], "entir": [2, 13], "like": 2, "incorrectli": 2, "empti": 2, "py": 2, "310": 2, "occur": 2, "pseudogen": 2, "corrupt": 2, "indic": 2, "anoth": 2, "circumst": [2, 5], "recognis": 2, "call": [2, 13], "being": 2, "canon": 2, "someth": [2, 4], "along": 2, "ighv1": 2, "69": 2, "02": 2, "some": 2, "latitud": 2, "allow": 2, "name": [2, 13], "depart": 2, "caus": [2, 4, 8], "In": [2, 5], "particular": [2, 13], "without": [2, 4], "suffix": 2, "onc": [2, 8], "identifi": [2, 3, 8], "step": 2, "resolv": 2, "excel": 2, "similar": 2, "reason": 2, "famili": 2, "arrang": 2, "investig": 2, "except": 2, "continu": 2, "exercis": 2, "caution": 2, "extend": [2, 5], "consciou": 2, "possibl": 2, "hope": 2, "section": [2, 3, 8], "doesn": 2, "t": [2, 13], "put": 2, "off": 2, "But": 2, "serv": 2, "systemat": 2, "lookout": 2, "well": [2, 13], "flag": 2, "also": 2, "understand": 2, "underli": 2, "data": [2, 8], "util": [3, 4, 8, 10], "latest": 3, "download": [3, 8], "here": [3, 8, 13], "nomin": 3, "speci": [3, 4, 8, 13], "ref": [3, 13], "l": [3, 13], "locu": [3, 8, 12], "f": [3, 13], "species_nam": 3, "quot": 3, "g": [3, 12, 13], "homo": 3, "sapien": 3, "macaca": 3, "mulatta": 3, "current": 3, "http": 3, "www": 3, "org": 3, "db": 3, "imgtgenedb": 3, "referencesequ": 3, "nt": 3, "withgap": 3, "orf": 3, "inframep": 3, "igh": [3, 4, 12], "igk": [3, 4], "trb": [3, 12], "trg": 3, "functional_onli": [3, 13], "": 4, "rhesu": 4, "macaqu": [4, 8], "ighv": [4, 13], "iglv": 4, "igkv": 4, "insert": [4, 8], "rel": [4, 8], "most": [4, 5, 8], "am": 4, "awar": 4, "conduct": 4, "exhaust": 4, "search": [4, 7], "These": [4, 8], "downstream": [4, 8], "chang": 4, "o": 4, "featur": 4, "residu": [4, 8], "present": [4, 10], "specif": [4, 8], "restor": 4, "standard": [4, 8], "remov": [4, 8], "small": [4, 5], "list": [4, 6, 8, 13], "those": [4, 5], "person": 4, "observ": 4, "befor": [4, 12], "produc": [4, 13], "unwant": 4, "ig": [4, 8], "infil": [4, 13], "outfil": [4, 10, 13], "chain": [4, 9, 13], "confirm": 4, "customari": 4, "igl": 4, "closest": [5, 8, 13], "templat": [5, 8], "onli": [5, 12, 13], "complet": 5, "5": [5, 13], "second": [5, 10, 13], "cystein": 5, "neither": 5, "appar": 5, "given": [5, 8, 13], "fraction": 5, "total": 5, "ignor": 5, "inferred_fil": 5, "ref_fil": [5, 9], "ungap": [5, 13], "infer": 5, "ident": 6, "sub": 6, "identical_sequ": 6, "introduct": 7, "custom": 7, "databas": 7, "api": 7, "modul": [7, 8, 11], "page": 7, "packag": [8, 11], "code": 8, "work": 8, "tr": 8, "receptor": 8, "pip": 8, "biopython": [8, 13], "separ": 8, "name_allel": [8, 15], "nearest": 8, "taken": 8, "suppli": 8, "incorpor": 8, "snp": [8, 12], "scheme": 8, "accord": 8, "emploi": 8, "must": [8, 13], "each": 8, "fix_macaque_gap": [8, 15], "ha": 8, "revert": 8, "next": 8, "auxiliari": 8, "extract_ref": [8, 12, 15], "identical_seq": [8, 15], "where": [8, 13], "same": [8, 13], "differ": [8, 13], "id": [8, 10, 13], "rev_comp": [8, 12, 15], "at_coord": [8, 15], "within": 8, "merge_fasta": [8, 15], "merg": [8, 10], "duplic": 8, "simpl": [8, 13], "wrapper": [8, 13], "around": [8, 13], "commonli": 8, "suit": 8, "exampl": 8, "write": [8, 13], "comma": 8, "tab": 8, "manag": 8, "string": [8, 13], "dict": [8, 13], "simplifi": 8, "compar": [8, 10, 12], "expens": 8, "flexibl": 8, "rare": 8, "document": 8, "variabl": 9, "region": [9, 13], "ndm_file": 9, "eg": 9, "vh": 9, "file1": 10, "file2": 10, "themselv": 10, "rather": 10, "keep": 10, "2": 10, "file_1": 10, "file_2": 10, "receptor_util": 11, "submodul": 11, "novel_allele_nam": 11, "build_ambiguous_ref": [11, 13], "find_indel": [11, 13], "name_novel": [11, 13], "run_test": [11, 13], "number_ighv": [11, 13], "check_conserved_residu": [11, 13], "distribute_cdr": [11, 13], "gap_align": [11, 13], "gap_align_aa": [11, 13], "gap_align_aa_from_nt": [11, 13], "gap_nt_from_aa": [11, 13], "insert_spac": [11, 13], "match_scor": [11, 13], "nt_diff": [11, 13], "number_from_trp": [11, 13], "pretty_gap": [11, 13], "simple_bio_seq": 11, "chunk": [11, 13], "closest_ref": [11, 13], "dumb_consensu": [11, 13], "read_csv": [11, 13], "read_fasta": [11, 13], "read_imgt_fasta": [11, 13], "read_single_fasta": [11, 13], "reverse_compl": [11, 13], "sample_fasta": [11, 13], "scored_consensu": [11, 13], "scored_gap_consensu": [11, 13], "toseqrecord": [11, 13], "translat": [11, 13], "write_csv": [11, 13], "write_fasta": [11, 13], "content": 11, "construct": 12, "novel": [12, 13], "suitabl": 12, "gene_nam": 12, "ref_set": [12, 13], "singl": [12, 13], "input": [12, 13, 14], "consid": 12, "process": 12, "full_ref": 13, "sourc": 13, "closest_ref_seq": 13, "novel_seq": 13, "str": 13, "v_gene": 13, "bool": 13, "true": 13, "its": 13, "conform": 13, "descript": 13, "prime": 13, "return": [13, 14], "we": 13, "tupl": 13, "consist": 13, "three": 13, "novel_nam": 13, "note": 13, "type": 13, "aa": 13, "seq": 13, "aa_seq": 13, "nt_gap": 13, "nucleotide_seq": 13, "peptide_seq": 13, "gapped_ref": 13, "ungapped_ref": 13, "po": 13, "target": 13, "match_list": 13, "thresh": 13, "s1": 13, "s2": 13, "next_p": 13, "trp_po": 13, "bioseq": 13, "applic": 13, "principl": 13, "store": 13, "collectiopn": 13, "coerc": 13, "upper": 13, "iter": 13, "corec": 13, "eas": 13, "debug": 13, "basic": 13, "thing": 13, "don": 13, "n": 13, "int": 13, "yield": 13, "success": 13, "size": 13, "entri": 13, "determin": 13, "equal": 13, "threshold": 13, "float": 13, "0": 13, "dumb": 13, "consensu": 13, "bio": 13, "aligninfo": 13, "summaryinfo": 13, "valu": 13, "atom": 13, "shorter": 13, "delimit": 13, "none": 13, "dictread": 13, "filenam": 13, "pathnam": 13, "dictionari": 13, "ighd": 13, "ighj": 13, "ch": 13, "space": 13, "replac": 13, "underscor": 13, "mark": 13, "serquenc": 13, "nucelotid": 13, "random": 13, "resampl": 13, "valueerror": 13, "enough": 13, "gap_consensu": 13, "mimimum": 13, "score": 13, "achiev": 13, "min_scor": 13, "ambigu": 13, "x": 13, "require_multipl": 13, "adapt": 13, "seqrecord": 13, "truncat": 13, "ignore_partial_codon": 13, "amino": 13, "acid": 13, "so": 13, "termin": 13, "boundari": 13, "otherwis": 13, "pad": 13, "ensur": 13, "get": 13, "row": 13, "header": 13, "kei": 13, "item": 13, "written": 13, "airr": [0, 9], "c": [0, 9], "json": [0, 9], "delin": 9, "include_orphon": 13, "orphon": 13, "number_v": 11}, "objects": {"": [[13, 0, 0, "-", "receptor_utils"]], "receptor_utils": [[13, 0, 0, "-", "novel_allele_name"], [13, 0, 0, "-", "number_v"], [13, 0, 0, "-", "simple_bio_seq"]], "receptor_utils.novel_allele_name": [[13, 1, 1, "", "build_ambiguous_ref"], [13, 1, 1, "", "find_indels"], [13, 1, 1, "", "name_novel"], [13, 1, 1, "", "run_tests"]], "receptor_utils.number_v": [[13, 1, 1, "", "check_conserved_residues"], [13, 1, 1, "", "distribute_cdr"], [13, 1, 1, "", "gap_align"], [13, 1, 1, "", "gap_align_aa"], [13, 1, 1, "", "gap_align_aa_from_nt"], [13, 1, 1, "", "gap_nt_from_aa"], [13, 1, 1, "", "gap_sequence"], [13, 1, 1, "", "insert_space"], [13, 1, 1, "", "match_score"], [13, 1, 1, "", "nt_diff"], [13, 1, 1, "", "number_from_trp"], [13, 1, 1, "", "number_ighv"], [13, 1, 1, "", "pretty_gapped"], [13, 1, 1, "", "run_tests"]], "receptor_utils.simple_bio_seq": [[13, 1, 1, "", "chunks"], [13, 1, 1, "", "closest_ref"], [13, 1, 1, "", "dumb_consensus"], [13, 1, 1, "", "nt_diff"], [13, 1, 1, "", "read_csv"], [13, 1, 1, "", "read_fasta"], [13, 1, 1, "", "read_imgt_fasta"], [13, 1, 1, "", "read_single_fasta"], [13, 1, 1, "", "reverse_complement"], [13, 1, 1, "", "sample_fasta"], [13, 1, 1, "", "scored_consensus"], [13, 1, 1, "", "scored_gap_consensus"], [13, 1, 1, "", "toSeqRecords"], [13, 1, 1, "", "translate"], [13, 1, 1, "", "write_csv"], [13, 1, 1, "", "write_fasta"]]}, "objtypes": {"0": "py:module", "1": "py:function"}, "objnames": {"0": ["py", "module", "Python module"], "1": ["py", "function", "Python function"]}, "titleterms": {"annotate_j": 0, "posit": [0, 1, 3, 4, 5, 6, 9, 10, 12, 14], "argument": [0, 1, 3, 4, 5, 6, 9, 10, 12, 14], "at_coord": 1, "name": [1, 3, 8, 12, 15], "us": [2, 8], "custom": [2, 8], "databas": [2, 8], "igblast": [2, 8, 15], "extract_ref": 3, "fix_macaque_gap": 4, "gap_sequ": 5, "identical_seq": 6, "receptor": 7, "util": [7, 15], "code": 7, "work": [7, 15], "ig": 7, "tr": 7, "sequenc": [7, 8, 15], "content": [7, 13], "indic": 7, "tabl": 7, "introduct": 8, "instal": 8, "align": [8, 15], "v": [8, 15], "d": [8, 15], "j": [8, 15], "conveni": [8, 15], "tool": [8, 15], "simple_bio_seq": [8, 13], "api": [8, 11], "make_igblast_ndm": 9, "merge_fasta": 10, "name_allel": 12, "receptor_util": 13, "packag": 13, "submodul": 13, "novel_allele_nam": 13, "modul": 13, "number_ighv": [], "rev_comp": 14, "command": 15, "line": 15, "number_v": 13}, "envversion": {"sphinx.domains.c": 2, "sphinx.domains.changeset": 1, "sphinx.domains.citation": 1, "sphinx.domains.cpp": 8, "sphinx.domains.index": 1, "sphinx.domains.javascript": 2, "sphinx.domains.math": 2, "sphinx.domains.python": 3, "sphinx.domains.rst": 2, "sphinx.domains.std": 2, "sphinx.ext.viewcode": 1, "sphinx": 57}, "alltitles": {"annotate_j": [[0, "annotate-j"]], "Positional Arguments": [[0, "positional-arguments"], [1, "positional-arguments"], [3, "positional-arguments"], [4, "positional-arguments"], [5, "positional-arguments"], [6, "positional-arguments"], [9, "positional-arguments"], [10, "positional-arguments"], [12, "positional-arguments"], [14, "positional-arguments"]], "at_coords": [[1, "at-coords"]], "Named Arguments": [[1, "named-arguments"], [3, "named-arguments"], [12, "named-arguments"]], "Using custom databases with IgBlast": [[2, "using-custom-databases-with-igblast"], [8, "using-custom-databases-with-igblast"]], "extract_refs": [[3, "extract-refs"]], "fix_macaque_gaps": [[4, "fix-macaque-gaps"]], "gap_sequences": [[5, "gap-sequences"]], "identical_seqs": [[6, "identical-seqs-label"]], "Receptor-utils: code and utilities for working with IG/TR receptor sequences": [[7, "receptor-utils-code-and-utilities-for-working-with-ig-tr-receptor-sequences"]], "Contents:": [[7, null]], "Indices and tables": [[7, "indices-and-tables"]], "Introduction": [[8, "introduction"]], "Installation": [[8, "installation"]], "Naming and aligning V, D and J sequences": [[8, "naming-and-aligning-v-d-and-j-sequences"]], "Convenience Tools": [[8, "convenience-tools"], [15, "convenience-tools"]], "simple_bio_seq API": [[8, "simple-bio-seq-api"]], "make_igblast_ndm": [[9, "make-igblast-ndm"]], "merge_fasta": [[10, "merge-fasta"]], "API": [[11, "api"]], "name_allele": [[12, "name-allele"]], "rev_comp": [[14, "rev-comp"]], "Command-Line Utilities": [[15, "command-line-utilities"]], "Naming and Aligning V, D and J Sequences": [[15, "naming-and-aligning-v-d-and-j-sequences"]], "Working with IgBLAST": [[15, "working-with-igblast"]], "receptor_utils package": [[13, "receptor-utils-package"]], "Submodules": [[13, "submodules"]], "receptor_utils.novel_allele_name module": [[13, "module-receptor_utils.novel_allele_name"]], "receptor_utils.number_v module": [[13, "module-receptor_utils.number_v"]], "receptor_utils.simple_bio_seq module": [[13, "module-receptor_utils.simple_bio_seq"]], "Module contents": [[13, "module-receptor_utils"]]}, "indexentries": {"build_ambiguous_ref() (in module receptor_utils.novel_allele_name)": [[13, "receptor_utils.novel_allele_name.build_ambiguous_ref"]], "check_conserved_residues() (in module receptor_utils.number_v)": [[13, "receptor_utils.number_v.check_conserved_residues"]], "chunks() (in module receptor_utils.simple_bio_seq)": [[13, "receptor_utils.simple_bio_seq.chunks"]], "closest_ref() (in module receptor_utils.simple_bio_seq)": [[13, "receptor_utils.simple_bio_seq.closest_ref"]], "distribute_cdr() (in module receptor_utils.number_v)": [[13, "receptor_utils.number_v.distribute_cdr"]], "dumb_consensus() (in module receptor_utils.simple_bio_seq)": [[13, "receptor_utils.simple_bio_seq.dumb_consensus"]], "find_indels() (in module receptor_utils.novel_allele_name)": [[13, "receptor_utils.novel_allele_name.find_indels"]], "gap_align() (in module receptor_utils.number_v)": [[13, "receptor_utils.number_v.gap_align"]], "gap_align_aa() (in module receptor_utils.number_v)": [[13, "receptor_utils.number_v.gap_align_aa"]], "gap_align_aa_from_nt() (in module receptor_utils.number_v)": [[13, "receptor_utils.number_v.gap_align_aa_from_nt"]], "gap_nt_from_aa() (in module receptor_utils.number_v)": [[13, "receptor_utils.number_v.gap_nt_from_aa"]], "gap_sequence() (in module receptor_utils.number_v)": [[13, "receptor_utils.number_v.gap_sequence"]], "insert_space() (in module receptor_utils.number_v)": [[13, "receptor_utils.number_v.insert_space"]], "match_score() (in module receptor_utils.number_v)": [[13, "receptor_utils.number_v.match_score"]], "module": [[13, "module-receptor_utils"], [13, "module-receptor_utils.novel_allele_name"], [13, "module-receptor_utils.number_v"], [13, "module-receptor_utils.simple_bio_seq"]], "name_novel() (in module receptor_utils.novel_allele_name)": [[13, "receptor_utils.novel_allele_name.name_novel"]], "nt_diff() (in module receptor_utils.number_v)": [[13, "receptor_utils.number_v.nt_diff"]], "nt_diff() (in module receptor_utils.simple_bio_seq)": [[13, "receptor_utils.simple_bio_seq.nt_diff"]], "number_from_trp() (in module receptor_utils.number_v)": [[13, "receptor_utils.number_v.number_from_trp"]], "number_ighv() (in module receptor_utils.number_v)": [[13, "receptor_utils.number_v.number_ighv"]], "pretty_gapped() (in module receptor_utils.number_v)": [[13, "receptor_utils.number_v.pretty_gapped"]], "read_csv() (in module receptor_utils.simple_bio_seq)": [[13, "receptor_utils.simple_bio_seq.read_csv"]], "read_fasta() (in module receptor_utils.simple_bio_seq)": [[13, "receptor_utils.simple_bio_seq.read_fasta"]], "read_imgt_fasta() (in module receptor_utils.simple_bio_seq)": [[13, "receptor_utils.simple_bio_seq.read_imgt_fasta"]], "read_single_fasta() (in module receptor_utils.simple_bio_seq)": [[13, "receptor_utils.simple_bio_seq.read_single_fasta"]], "receptor_utils": [[13, "module-receptor_utils"]], "receptor_utils.novel_allele_name": [[13, "module-receptor_utils.novel_allele_name"]], "receptor_utils.number_v": [[13, "module-receptor_utils.number_v"]], "receptor_utils.simple_bio_seq": [[13, "module-receptor_utils.simple_bio_seq"]], "reverse_complement() (in module receptor_utils.simple_bio_seq)": [[13, "receptor_utils.simple_bio_seq.reverse_complement"]], "run_tests() (in module receptor_utils.novel_allele_name)": [[13, "receptor_utils.novel_allele_name.run_tests"]], "run_tests() (in module receptor_utils.number_v)": [[13, "receptor_utils.number_v.run_tests"]], "sample_fasta() (in module receptor_utils.simple_bio_seq)": [[13, "receptor_utils.simple_bio_seq.sample_fasta"]], "scored_consensus() (in module receptor_utils.simple_bio_seq)": [[13, "receptor_utils.simple_bio_seq.scored_consensus"]], "scored_gap_consensus() (in module receptor_utils.simple_bio_seq)": [[13, "receptor_utils.simple_bio_seq.scored_gap_consensus"]], "toseqrecords() (in module receptor_utils.simple_bio_seq)": [[13, "receptor_utils.simple_bio_seq.toSeqRecords"]], "translate() (in module receptor_utils.simple_bio_seq)": [[13, "receptor_utils.simple_bio_seq.translate"]], "write_csv() (in module receptor_utils.simple_bio_seq)": [[13, "receptor_utils.simple_bio_seq.write_csv"]], "write_fasta() (in module receptor_utils.simple_bio_seq)": [[13, "receptor_utils.simple_bio_seq.write_fasta"]]}})