Search.setIndex({"docnames": ["annotate_j", "at_coords", "custom_igblast", "download_germline_set", "extract_refs", "fix_macaque_gaps", "gap_sequences", "identical_seqs", "index", "introduction", "make_igblast_ndm", "merge_fasta", "modules", "name_allele", "receptor_utils", "rev_comp", "scripts"], "filenames": ["annotate_j.rst", "at_coords.rst", "custom_igblast.rst", "download_germline_set.rst", "extract_refs.rst", "fix_macaque_gaps.rst", "gap_sequences.rst", "identical_seqs.rst", "index.rst", "introduction.rst", "make_igblast_ndm.rst", "merge_fasta.rst", "modules.rst", "name_allele.rst", "receptor_utils.rst", "rev_comp.rst", "scripts.rst"], "titles": ["annotate_j", "at_coords", "Using custom databases with IgBlast", "download_germline_set", "extract_refs", "fix_macaque_gaps", "gap_sequences", "identical_seqs", "Receptor-utils: code and utilities for working with IG/TR receptor sequences", "Introduction", "make_igblast_ndm", "merge_fasta", "receptor_utils", "name_allele", "receptor_utils package", "rev_comp", "Command-Line Utilities"], "terms": {"annot": [0, 2, 3], "set": [0, 2, 3, 5, 6, 9, 10, 13, 14], "j": [0, 2, 3, 14], "sequenc": [0, 1, 2, 3, 5, 6, 7, 10, 11, 13, 14, 15], "find": [0, 2, 9, 14], "frame": [0, 2], "align": [0, 2, 5, 10, 14], "1": [0, 1, 2, 10, 11, 14], "base": [0, 1, 10], "index": [0, 8, 14], "first": [0, 11, 14], "nucleotid": [0, 9, 14, 15], "conserv": [0, 5, 9], "phe": 0, "trp": 0, "usag": [0, 1, 3, 4, 5, 6, 7, 10, 11, 13, 15], "h": [0, 1, 3, 4, 5, 6, 7, 10, 11, 13, 15], "seq_fil": 0, "out_fil": [0, 6], "fasta": [0, 1, 3, 4, 5, 6, 7, 9, 10, 11, 13, 14], "airr": [0, 3, 10], "c": [0, 3, 10], "json": [0, 3, 10], "format": [0, 2, 3, 10, 14], "output": [0, 2, 6, 10, 11, 14], "aux": [0, 2], "us": [0, 3, 5, 6, 8, 10, 13, 14], "igblast": [0, 3, 8, 10], "extract": [1, 4], "specifi": [1, 2, 3, 4, 10, 13, 14], "co": [1, 5, 9, 10], "ordin": [1, 5, 9, 10], "file": [1, 2, 3, 4, 5, 6, 7, 9, 10, 11, 14], "r": [1, 4, 13], "input_fil": [1, 7], "start": [1, 2, 10, 14], "end": [1, 6, 14], "contain": [1, 2, 3, 5, 6, 9, 10, 11, 13, 14], "refer": [1, 2, 3, 4, 6, 7, 9, 10, 13, 14], "ord": 1, "revers": [1, 9, 13, 14, 15], "complement": [1, 9, 13, 14, 15], "result": [1, 2], "default": [1, 3, 4, 10, 13, 14], "fals": [1, 4, 13, 14], "local": 2, "instal": 2, "can": [2, 3, 4, 6, 9, 10], "configur": [2, 3], "support": [2, 9], "organ": 2, "either": [2, 14], "an": [2, 3, 4, 10, 13, 14], "which": [2, 3, 9, 14], "ar": [2, 3, 6, 9, 10, 11, 14], "provid": [2, 3, 9, 14], "tool": [2, 3, 5, 6, 10], "modifi": [2, 14], "The": [2, 3, 4, 6, 9, 11, 14], "procedur": 2, "i": [2, 3, 5, 6, 9, 10, 14], "significantli": 2, "simplif": 2, "version": [2, 3], "20": 2, "upgrad": 2, "least": [2, 6], "thi": [2, 3, 4, 5, 9, 10, 11, 14], "encourag": 2, "For": [2, 9], "full": [2, 3, 14], "two": [2, 9, 11], "requir": [2, 9, 10, 14], "addit": [2, 3, 5, 9], "germlin": [2, 3, 9, 10], "auxiliary_data": 2, "describ": [2, 9], "detail": [2, 9], "gene": [2, 4, 13, 14], "orient": 2, "ndm": [2, 9, 10], "posit": [2, 14], "cdr": [2, 3, 5, 10, 14], "v": [2, 3, 10, 13, 14], "If": [2, 3, 4, 10, 14], "you": [2, 5, 10], "just": [2, 14], "wish": [2, 5], "add": [2, 14], "allel": [2, 9, 13, 14], "exist": 2, "d": [2, 3, 14], "all": [2, 3, 10, 11, 14], "necessari": [2, 13, 14], "creat": [2, 3, 9, 13, 14], "blast": 2, "link": 2, "abov": [2, 9], "case": [2, 9, 10, 14], "need": [2, 3, 9, 10, 14], "new": 2, "custom_internal_data": 2, "command": [2, 8], "line": [2, 8, 9], "paramet": [2, 14], "make_igblast_ndm": [2, 3, 9, 16], "make": [2, 10, 14], "from": [2, 3, 4, 5, 9, 10, 11, 13, 14], "imgt": [2, 4, 5, 7, 9, 10, 13, 14], "gap": [2, 3, 4, 5, 6, 7, 9, 10, 13, 14], "should": [2, 13, 14], "includ": [2, 3, 9, 14], "them": [2, 5], "gap_sequ": [2, 9, 12, 14, 16], "have": [2, 5, 10], "partial": 2, "match": [2, 11], "your": [2, 10], "close": [2, 14], "relat": 2, "annotate_j": [2, 3, 9, 16], "It": [2, 9], "veri": 2, "import": 2, "test": 2, "carefulli": 2, "when": 2, "issu": [2, 6], "particularli": 2, "mai": [2, 5, 10, 14], "fail": 2, "substanti": 2, "proport": 2, "submit": 2, "my": [2, 9], "own": 2, "approach": 2, "follow": [2, 10], "repertoir": [2, 5], "outfmt": 2, "7": [2, 14], "std": 2, "qseq": 2, "sseq": 2, "btop": 2, "convert": [2, 14], "changeo": 2, "makedb": 2, "option": [2, 3, 10, 14], "record": [2, 11, 14], "log": 2, "examin": 2, "count": [2, 14], "pass": 2, "typic": 2, "might": 2, "expect": [2, 5], "under": [2, 9], "10": 2, "possibli": 2, "mani": 2, "fewer": 2, "than": [2, 9, 11], "depend": [2, 10], "qualiti": 2, "read": [2, 9, 14], "A": [2, 5, 10, 14], "higher": 2, "warn": [2, 6, 9], "sign": 2, "open": [2, 3], "db_fail": 2, "were": [2, 14], "found": 2, "stop": 2, "codon": [2, 5, 9, 10, 14], "discount": 2, "how": 2, "other": [2, 5, 6, 9, 14], "what": 2, "do": [2, 10, 11, 14], "thei": [2, 6, 9], "repres": 2, "more": [2, 9, 14], "few": 2, "percent": 2, "obvious": 2, "short": 2, "e": [2, 3, 4, 13, 14], "length": [2, 14], "again": 2, "take": 2, "sampl": [2, 10, 14], "look": 2, "though": 2, "analys": [2, 10, 14], "quest": [2, 10], "interest": 2, "one": [2, 9, 10, 14], "doe": [2, 14], "non": [2, 6], "function": [2, 6, 9, 14], "number": [2, 5, 10, 14], "strongli": 2, "suspect": 2, "problem": [2, 5, 9], "error": 2, "check": [2, 5], "report": [2, 9], "ani": [2, 3, 6, 9, 11, 14], "analysi": [2, 10], "been": [2, 3], "mi": 2, "see": 2, "junction": 2, "assign": 2, "mean": [2, 3], "inde": 2, "part": 2, "miss": [2, 9], "There": 2, "simpli": 2, "too": 2, "cover": [2, 9, 14], "entir": [2, 14], "like": [2, 10], "incorrectli": 2, "empti": 2, "py": 2, "310": 2, "occur": 2, "pseudogen": 2, "corrupt": 2, "indic": 2, "anoth": [2, 3], "circumst": [2, 6], "recognis": 2, "call": [2, 14], "being": 2, "canon": 2, "someth": [2, 5], "along": 2, "ighv1": 2, "69": 2, "02": 2, "some": [2, 10], "latitud": 2, "allow": 2, "name": [2, 14], "depart": 2, "caus": [2, 5, 9], "In": [2, 6, 10], "particular": [2, 14], "without": [2, 3, 5], "suffix": 2, "onc": [2, 9], "identifi": [2, 4, 9], "step": 2, "resolv": 2, "excel": 2, "similar": 2, "reason": 2, "famili": 2, "arrang": 2, "investig": 2, "except": 2, "continu": 2, "exercis": 2, "caution": 2, "extend": [2, 6], "consciou": 2, "possibl": [2, 3], "hope": 2, "section": [2, 4, 9], "doesn": 2, "t": [2, 14], "put": 2, "off": 2, "But": 2, "serv": 2, "systemat": 2, "lookout": 2, "well": [2, 14], "flag": 2, "also": [2, 3], "understand": 2, "underli": 2, "data": [2, 9], "util": [3, 4, 5, 9, 11], "latest": [3, 4], "download": [3, 4, 9], "here": [4, 9, 14], "nomin": 4, "speci": [3, 4, 5, 9, 14], "ref": [4, 14], "l": [4, 14], "locu": [3, 4, 9, 13], "f": [4, 14, 16], "species_nam": 4, "quot": 4, "g": [3, 4, 13, 14], "homo": 4, "sapien": 4, "macaca": 4, "mulatta": 4, "current": 4, "http": [3, 4], "www": 4, "org": [3, 4], "db": 4, "imgtgenedb": 4, "referencesequ": 4, "nt": 4, "withgap": 4, "orf": 4, "inframep": 4, "igh": [3, 4, 5, 13], "igk": [3, 4, 5], "trb": [3, 4, 13], "trg": 4, "functional_onli": [4, 14], "": 5, "rhesu": [5, 10], "macaqu": [5, 9, 10], "ighv": [5, 14], "iglv": 5, "igkv": 5, "insert": [5, 9, 10], "rel": [5, 9], "most": [5, 6, 9], "am": [], "awar": [], "conduct": 10, "exhaust": [], "search": 8, "These": [3, 5, 9], "downstream": [5, 9, 10], "chang": 5, "o": 5, "featur": 5, "residu": [5, 9], "present": [5, 11], "specif": [3, 5, 9, 10], "restor": 5, "standard": [5, 9], "remov": [5, 9, 10], "small": [5, 6], "list": [5, 7, 9, 10, 14], "those": [5, 6, 10], "person": 5, "observ": 5, "befor": [5, 13], "produc": [5, 14], "unwant": 5, "ig": [5, 9], "infil": [5, 14], "outfil": [5, 11, 14], "chain": [5, 10, 14], "confirm": 5, "customari": 5, "igl": [3, 5], "closest": [6, 9, 14], "templat": [6, 9], "onli": [6, 13, 14], "complet": 6, "5": [6, 14], "second": [6, 11, 14], "cystein": 6, "neither": 6, "appar": 6, "given": [6, 9, 14], "fraction": 6, "total": 6, "ignor": 6, "inferred_fil": 6, "ref_fil": [6, 10], "ungap": [3, 6, 14], "infer": 6, "ident": 7, "sub": 7, "identical_sequ": 7, "introduct": 8, "custom": [3, 8], "databas": [3, 8], "receptor_util": 8, "modul": [8, 9, 12], "page": 8, "packag": [9, 12], "code": 9, "work": 9, "tr": 9, "receptor": [3, 9], "pip": 9, "biopython": [9, 14], "separ": [9, 10], "name_allel": [9, 16], "nearest": 9, "taken": 9, "suppli": 9, "incorpor": 9, "snp": [9, 13], "scheme": [9, 10], "accord": 9, "emploi": 9, "must": [9, 10, 14], "each": [9, 10], "fix_macaque_gap": [9, 10, 16], "ha": [3, 9], "revert": 9, "next": 9, "auxiliari": 9, "extract_ref": [9, 13, 16], "identical_seq": [9, 16], "where": [9, 14], "same": [9, 14], "differ": [9, 14], "id": [9, 11, 14], "rev_comp": [9, 13, 16], "at_coord": [9, 16], "within": 9, "merge_fasta": [9, 16], "merg": [9, 11], "duplic": 9, "simpl": [9, 14], "wrapper": [9, 14], "around": [9, 14], "commonli": 9, "suit": 9, "exampl": 9, "write": [9, 14], "comma": [9, 10], "tab": 9, "manag": 9, "string": [9, 14], "dict": [9, 14], "simplifi": 9, "compar": [9, 11, 13], "expens": 9, "flexibl": [3, 9], "rare": 9, "document": 9, "delin": [3, 10], "variabl": 10, "region": [10, 14], "cdr_coord": 10, "ndm_file": 10, "eg": 10, "vh": 10, "cdr1": 10, "cdr2": 10, "cdr3": 10, "79": 10, "114": 10, "166": 10, "195": 10, "313": 10, "file1": 11, "file2": 11, "themselv": 11, "rather": 11, "keep": 11, "2": 11, "file_1": 11, "file_2": 11, "submodul": 12, "novel_allele_nam": 12, "aligned_diff": [12, 14], "build_ambiguous_ref": [12, 14], "closest_aligned_ref": [12, 14], "find_indel": [12, 14], "name_novel": [12, 14], "run_test": [12, 14], "number_v": 12, "check_conserved_residu": [12, 14], "distribute_cdr": [12, 14], "gap_align": [12, 14], "gap_align_aa": [12, 14], "gap_align_aa_from_nt": [12, 14], "gap_nt_from_aa": [12, 14], "insert_spac": [12, 14], "match_scor": [12, 14], "nt_diff": [12, 14], "number_from_trp": [12, 14], "number_ighv": [12, 14], "pretty_gap": [12, 14], "simple_bio_seq": 12, "chunk": [12, 14], "closest_ref": [12, 14], "dumb_consensu": [12, 14], "read_csv": [12, 14], "read_fasta": [12, 14], "read_imgt_fasta": [12, 14], "read_single_fasta": [12, 14], "reverse_compl": [12, 14], "sample_fasta": [12, 14], "scored_consensu": [12, 14], "scored_gap_consensu": [12, 14], "toseqrecord": [12, 14], "translat": [12, 14], "write_csv": [12, 14], "write_fasta": [12, 14], "content": 12, "construct": 13, "novel": [13, 14], "suitabl": 13, "gene_nam": 13, "ref_set": [13, 14], "singl": [3, 13, 14], "input": [13, 14, 15], "consid": 13, "process": 13, "novel_seq": 14, "str": 14, "ref_seq": 14, "sourc": 14, "full_ref": 14, "seq": 14, "entri": 14, "determin": [3, 14], "global": 14, "return": [14, 15], "equal": 14, "type": 14, "closest_ref_seq": 14, "v_gene": 14, "bool": 14, "true": 14, "its": 14, "conform": 14, "descript": 14, "prime": 14, "we": [10, 14], "tupl": 14, "consist": 14, "three": 14, "novel_nam": 14, "note": [14, 16], "aa": 14, "aa_seq": 14, "nt_gap": 14, "nucleotide_seq": 14, "peptide_seq": 14, "gapped_ref": 14, "ungapped_ref": 14, "po": 14, "target": 14, "match_list": 14, "thresh": 14, "s1": 14, "s2": 14, "next_p": 14, "trp_po": 14, "bioseq": 14, "applic": [10, 14], "principl": [3, 14], "store": 14, "collectiopn": 14, "coerc": 14, "upper": 14, "iter": 14, "corec": 14, "eas": 14, "debug": 14, "basic": 14, "thing": 14, "don": 14, "n": [3, 14], "int": 14, "yield": 14, "success": 14, "size": 14, "threshold": 14, "float": 14, "0": 14, "dumb": 14, "consensu": 14, "bio": 14, "aligninfo": 14, "summaryinfo": 14, "valu": 14, "atom": 14, "shorter": 14, "delimit": 14, "none": [3, 14], "dictread": 14, "filenam": [3, 14], "pathnam": 14, "dictionari": 14, "ighd": 14, "ighj": 14, "ch": 14, "include_orphon": 14, "space": [3, 14], "replac": [3, 14], "underscor": 14, "mark": 14, "orphon": 14, "serquenc": 14, "nucelotid": 14, "random": 14, "resampl": 14, "valueerror": 14, "enough": 14, "gap_consensu": 14, "mimimum": 14, "score": 14, "achiev": 14, "min_scor": 14, "ambigu": 14, "x": 14, "require_multipl": 14, "adapt": 14, "seqrecord": 14, "truncat": 14, "ignore_partial_codon": 14, "amino": 14, "acid": 14, "so": 14, "termin": 14, "boundari": 14, "otherwis": [3, 14], "pad": 14, "ensur": [10, 14], "get": 14, "row": 14, "scan_al": 14, "header": 14, "kei": 14, "item": 14, "scan": 14, "written": 14, "avail": 10, "recommend": [3, 10], "correct": 10, "placement": 10, "crd": [], "coordin": 10, "ogrdb": [3, 10], "By": [3, 10], "assum": 10, "pleas": 10, "doubt": 10, "review": 10, "denot": 10, "b": 10, "etc": 10, "altern": 10, "affect": [], "offer": 10, "realign": 10, "provis": 10, "upon": 10, "commun": 3, "download_germline_set": 16, "airrc": 3, "fg": 3, "fu": 3, "multi": 3, "u": 3, "url": 3, "p": 3, "prefix": 3, "human": 3, "tra": 3, "trd": 3, "attempt": 3, "choic": 3, "api_v2": 3, "species_locu": 3, "_": 3, "substitut": 3, "design": 3, "manner": 3, "advantag": 3, "inform": 3, "delieat": 3, "load": 3, "respect": 3, "_gap": 3, "final": 3, "four": 3, "_v": 3, "_d": 3, "_j": 3, "_v_gap": 3, "multi_f": 3, "explan": 16, "argument": 16}, "objects": {"": [[14, 0, 0, "-", "receptor_utils"]], "receptor_utils": [[14, 0, 0, "-", "novel_allele_name"], [14, 0, 0, "-", "number_v"], [14, 0, 0, "-", "simple_bio_seq"]], "receptor_utils.novel_allele_name": [[14, 1, 1, "", "aligned_diff"], [14, 1, 1, "", "build_ambiguous_ref"], [14, 1, 1, "", "closest_aligned_ref"], [14, 1, 1, "", "find_indels"], [14, 1, 1, "", "name_novel"], [14, 1, 1, "", "run_tests"]], "receptor_utils.number_v": [[14, 1, 1, "", "check_conserved_residues"], [14, 1, 1, "", "distribute_cdr"], [14, 1, 1, "", "gap_align"], [14, 1, 1, "", "gap_align_aa"], [14, 1, 1, "", "gap_align_aa_from_nt"], [14, 1, 1, "", "gap_nt_from_aa"], [14, 1, 1, "", "gap_sequence"], [14, 1, 1, "", "insert_space"], [14, 1, 1, "", "match_score"], [14, 1, 1, "", "nt_diff"], [14, 1, 1, "", "number_from_trp"], [14, 1, 1, "", "number_ighv"], [14, 1, 1, "", "pretty_gapped"], [14, 1, 1, "", "run_tests"]], "receptor_utils.simple_bio_seq": [[14, 1, 1, "", "chunks"], [14, 1, 1, "", "closest_ref"], [14, 1, 1, "", "dumb_consensus"], [14, 1, 1, "", "nt_diff"], [14, 1, 1, "", "read_csv"], [14, 1, 1, "", "read_fasta"], [14, 1, 1, "", "read_imgt_fasta"], [14, 1, 1, "", "read_single_fasta"], [14, 1, 1, "", "reverse_complement"], [14, 1, 1, "", "sample_fasta"], [14, 1, 1, "", "scored_consensus"], [14, 1, 1, "", "scored_gap_consensus"], [14, 1, 1, "", "toSeqRecords"], [14, 1, 1, "", "translate"], [14, 1, 1, "", "write_csv"], [14, 1, 1, "", "write_fasta"]]}, "objtypes": {"0": "py:module", "1": "py:function"}, "objnames": {"0": ["py", "module", "Python module"], "1": ["py", "function", "Python function"]}, "titleterms": {"annotate_j": 0, "posit": [0, 1, 3, 4, 5, 6, 7, 10, 11, 13, 15], "argument": [0, 1, 3, 4, 5, 6, 7, 10, 11, 13, 15], "at_coord": 1, "name": [1, 3, 4, 9, 10, 13, 16], "us": [2, 9], "custom": [2, 9], "databas": [2, 9], "igblast": [2, 9, 16], "extract_ref": 4, "fix_macaque_gap": 5, "gap_sequ": 6, "identical_seq": 7, "receptor": 8, "util": [8, 16], "code": 8, "work": [8, 16], "ig": 8, "tr": 8, "sequenc": [8, 9, 16], "content": [8, 14], "indic": 8, "tabl": 8, "introduct": 9, "instal": 9, "align": [9, 16], "v": [9, 16], "d": [9, 16], "j": [9, 16], "conveni": [9, 16], "tool": [9, 16], "simple_bio_seq": [9, 14], "api": 9, "make_igblast_ndm": 10, "merge_fasta": 11, "receptor_util": [12, 14], "name_allel": 13, "packag": 14, "submodul": 14, "novel_allele_nam": 14, "modul": 14, "number_v": 14, "rev_comp": 15, "command": 16, "line": 16, "note": 10, "download_germline_set": 3, "explan": 3, "f": 3}, "envversion": {"sphinx.domains.c": 2, "sphinx.domains.changeset": 1, "sphinx.domains.citation": 1, "sphinx.domains.cpp": 8, "sphinx.domains.index": 1, "sphinx.domains.javascript": 2, "sphinx.domains.math": 2, "sphinx.domains.python": 3, "sphinx.domains.rst": 2, "sphinx.domains.std": 2, "sphinx.ext.viewcode": 1, "sphinx": 57}, "alltitles": {"annotate_j": [[0, "annotate-j"]], "Positional Arguments": [[0, "positional-arguments"], [1, "positional-arguments"], [4, "positional-arguments"], [6, "positional-arguments"], [7, "positional-arguments"], [11, "positional-arguments"], [13, "positional-arguments"], [15, "positional-arguments"], [5, "positional-arguments"], [10, "positional-arguments"], [3, "positional-arguments"]], "at_coords": [[1, "at-coords"]], "Named Arguments": [[1, "named-arguments"], [4, "named-arguments"], [13, "named-arguments"], [10, "named-arguments"], [3, "named-arguments"]], "Using custom databases with IgBlast": [[2, "using-custom-databases-with-igblast"], [9, "using-custom-databases-with-igblast"]], "extract_refs": [[4, "extract-refs"]], "gap_sequences": [[6, "gap-sequences"]], "identical_seqs": [[7, "identical-seqs-label"]], "Receptor-utils: code and utilities for working with IG/TR receptor sequences": [[8, "receptor-utils-code-and-utilities-for-working-with-ig-tr-receptor-sequences"]], "Contents:": [[8, null]], "Indices and tables": [[8, "indices-and-tables"]], "Introduction": [[9, "introduction"]], "Installation": [[9, "installation"]], "Naming and aligning V, D and J sequences": [[9, "naming-and-aligning-v-d-and-j-sequences"]], "Convenience Tools": [[9, "convenience-tools"], [16, "convenience-tools"]], "simple_bio_seq API": [[9, "simple-bio-seq-api"]], "merge_fasta": [[11, "merge-fasta"]], "receptor_utils": [[12, "receptor-utils"]], "name_allele": [[13, "name-allele"]], "rev_comp": [[15, "rev-comp"]], "fix_macaque_gaps": [[5, "fix-macaque-gaps"]], "make_igblast_ndm": [[10, "make-igblast-ndm"]], "Notes": [[10, "notes"]], "receptor_utils package": [[14, "receptor-utils-package"]], "Submodules": [[14, "submodules"]], "receptor_utils.novel_allele_name module": [[14, "module-receptor_utils.novel_allele_name"]], "receptor_utils.number_v module": [[14, "module-receptor_utils.number_v"]], "receptor_utils.simple_bio_seq module": [[14, "module-receptor_utils.simple_bio_seq"]], "Module contents": [[14, "module-receptor_utils"]], "Command-Line Utilities": [[16, "command-line-utilities"]], "Naming and Aligning V, D and J Sequences": [[16, "naming-and-aligning-v-d-and-j-sequences"]], "Working with IgBLAST": [[16, "working-with-igblast"]], "download_germline_set": [[3, "download-germline-set"]], "Explanation of the -f argument": [[3, "explanation-of-the-f-argument"]]}, "indexentries": {}})