Search.setIndex({"docnames": ["annotate_j", "at_coords", "custom_igblast", "extract_refs", "gap_sequences", "identical_seqs", "index", "introduction", "make_igblast_ndm", "modules", "name_allele", "receptor_utils", "reverse_complement", "scripts"], "filenames": ["annotate_j.rst", "at_coords.rst", "custom_igblast.rst", "extract_refs.rst", "gap_sequences.rst", "identical_seqs.rst", "index.rst", "introduction.rst", "make_igblast_ndm.rst", "modules.rst", "name_allele.rst", "receptor_utils.rst", "reverse_complement.rst", "scripts.rst"], "titles": ["annotate_j", "at_coords", "Using custom databases with IgBlast", "extract_refs", "gap_sequences", "identical_seqs", "Receptor-utils: code and utilities for working with IG/TR receptor sequences", "Introduction", "make_igblast_ndm", "API", "name_allele", "receptor_utils package", "rev_comp", "Command-Line Utilities"], "terms": {"annot": [0, 2], "set": [0, 2, 4, 7, 8, 10, 11], "j": [0, 2, 11], "sequenc": [0, 1, 2, 4, 5, 10, 11, 12], "find": [0, 2, 7, 11], "frame": [0, 2], "align": [0, 2, 11], "1": [0, 1, 2, 11], "base": [0, 1], "index": [0, 6, 11], "first": [0, 11], "nucleotid": [0, 11, 12], "conserv": [0, 7], "phe": 0, "trp": 0, "usag": [0, 1, 3, 4, 5, 8, 10, 12], "h": [0, 1, 3, 4, 5, 8, 10, 12], "seq_fil": 0, "out_fil": [0, 4], "fasta": [0, 1, 4, 5, 7, 8, 10, 11], "output": [0, 2, 4, 8, 11], "aux": [0, 2], "format": [0, 2, 11], "us": [0, 4, 6, 11], "igblast": [0, 6, 8], "extract": [1, 3, 7], "specifi": [1, 2, 10, 11], "co": 1, "ordin": 1, "file": [1, 2, 3, 4, 5, 7, 8, 11], "r": [1, 10], "input_fil": [1, 5], "start": [1, 2], "end": [1, 4, 11], "contain": [1, 2, 4, 7, 10, 11], "refer": [1, 2, 3, 4, 5, 7, 8, 10, 11], "ord": 1, "revers": [1, 7, 10, 11, 12], "complement": [1, 7, 10, 11, 12], "result": [1, 2], "default": [1, 3, 10, 11], "fals": [1, 3, 10, 11], "local": 2, "instal": 2, "can": [2, 3, 4], "configur": 2, "support": [2, 7], "organ": 2, "either": [2, 11], "an": [2, 3, 7, 8, 10, 11], "which": [2, 7, 11], "ar": [2, 4, 7, 11], "provid": [2, 7, 11], "tool": [2, 4], "modifi": [2, 11], "The": [2, 3, 4, 7, 11], "procedur": 2, "i": [2, 4, 7, 11], "significantli": 2, "simplif": 2, "version": 2, "20": 2, "upgrad": 2, "least": [2, 4], "thi": [2, 3, 7, 11], "encourag": 2, "For": [2, 7], "full": [2, 11], "two": 2, "requir": [2, 7, 8, 11], "addit": 2, "germlin": [2, 7], "auxiliary_data": 2, "describ": [2, 7], "detail": 2, "gene": [2, 7, 10, 11], "orient": 2, "ndm": [2, 8], "posit": [2, 11], "cdr": 2, "v": [2, 10, 11], "If": [2, 11], "you": 2, "just": [2, 11], "wish": 2, "add": [2, 11], "allel": [2, 7, 10, 11], "exist": 2, "d": [2, 11], "all": [2, 7, 11], "necessari": [2, 10, 11], "creat": [2, 11], "blast": 2, "link": 2, "abov": [2, 7], "case": [2, 7, 11], "ani": [2, 4, 7, 11], "need": [2, 7, 11], "new": 2, "custom_internal_data": 2, "command": [2, 6, 7], "line": [2, 6, 7], "paramet": [2, 11], "make_igblast_ndm": [2, 13], "make": [2, 8, 11], "from": [2, 3, 7, 8, 10, 11], "imgt": [2, 3, 5, 7, 8, 10, 11], "gap": [2, 3, 4, 5, 7, 8, 10, 11], "should": [2, 10, 11], "includ": [2, 7], "them": 2, "gap_sequ": [2, 7, 13], "have": [2, 7], "partial": 2, "match": 2, "your": 2, "close": [2, 11], "relat": 2, "annotate_j": [2, 13], "It": [2, 7], "veri": 2, "import": 2, "test": 2, "carefulli": 2, "when": [2, 7], "issu": [2, 4], "particularli": [2, 7], "mai": [2, 7, 11], "fail": 2, "substanti": 2, "proport": 2, "submit": 2, "my": [2, 7], "own": 2, "approach": 2, "follow": 2, "repertoir": 2, "outfmt": 2, "7": [2, 11], "std": 2, "qseq": 2, "sseq": 2, "btop": 2, "convert": [2, 11], "changeo": 2, "makedb": 2, "option": [2, 11], "record": [2, 11], "log": 2, "examin": 2, "count": [2, 11], "pass": 2, "typic": 2, "might": 2, "expect": 2, "under": [2, 7], "10": 2, "possibli": 2, "mani": 2, "fewer": 2, "than": 2, "depend": 2, "qualiti": 2, "read": [2, 7, 11], "A": [2, 11], "higher": 2, "warn": [2, 4, 7], "sign": 2, "open": 2, "db_fail": 2, "were": [2, 11], "found": [2, 7], "stop": 2, "codon": [2, 11], "discount": 2, "how": 2, "other": [2, 4, 7, 11], "what": [2, 7], "do": [2, 11], "thei": [2, 4, 7], "repres": 2, "more": [2, 11], "few": 2, "percent": 2, "obvious": 2, "short": 2, "e": [2, 3, 11], "length": [2, 11], "again": 2, "take": 2, "sampl": [2, 11], "look": 2, "though": 2, "analys": [2, 11], "quest": 2, "interest": 2, "one": [2, 7, 11], "doe": [2, 11], "non": [2, 4], "function": [2, 4, 7, 11], "number": [2, 11], "strongli": 2, "suspect": 2, "problem": 2, "error": 2, "check": 2, "report": 2, "analysi": 2, "been": 2, "mi": 2, "see": 2, "junction": 2, "assign": 2, "mean": 2, "inde": 2, "part": 2, "miss": [2, 7], "There": [2, 7], "simpli": 2, "too": 2, "cover": [2, 7, 11], "entir": [2, 11], "like": 2, "incorrectli": 2, "empti": 2, "py": 2, "310": 2, "occur": 2, "pseudogen": 2, "corrupt": 2, "indic": 2, "anoth": 2, "circumst": [2, 4], "recognis": 2, "call": [2, 11], "being": 2, "canon": 2, "someth": 2, "along": 2, "ighv1": 2, "69": 2, "02": 2, "some": [2, 7], "latitud": 2, "allow": 2, "name": [2, 11], "depart": 2, "caus": 2, "In": [2, 4], "particular": [2, 11], "without": 2, "suffix": 2, "onc": 2, "identifi": [2, 3, 7], "step": 2, "resolv": 2, "excel": 2, "similar": 2, "reason": 2, "famili": 2, "arrang": 2, "investig": 2, "except": 2, "continu": 2, "exercis": 2, "caution": 2, "extend": [2, 4], "consciou": 2, "possibl": 2, "hope": 2, "section": [2, 3, 7], "doesn": 2, "t": [2, 11], "put": 2, "off": 2, "But": 2, "serv": 2, "systemat": 2, "lookout": 2, "well": [2, 7, 11], "flag": 2, "also": 2, "understand": 2, "underli": 2, "data": 2, "util": [3, 7], "latest": 3, "download": 3, "here": [3, 7, 11], "nomin": 3, "speci": [3, 11], "l": [3, 11], "locu": 3, "f": [3, 11], "imgt_fil": 3, "species_nam": 3, "quot": 3, "g": [3, 10, 11], "homo": 3, "sapien": 3, "macaca": 3, "mulatta": 3, "igh": 3, "igk": 3, "trb": 3, "trg": 3, "functional_onli": [3, 11], "closest": [4, 7, 11], "templat": [4, 7], "onli": [4, 10, 11], "those": 4, "complet": 4, "5": [4, 11], "second": [4, 11], "cystein": 4, "neither": 4, "appar": 4, "given": [4, 7, 11], "most": 4, "small": 4, "fraction": 4, "total": 4, "ignor": 4, "inferred_fil": 4, "ref_fil": [4, 8], "ungap": [4, 11], "infer": [4, 7], "list": [5, 7, 11], "ident": 5, "sub": 5, "identical_sequ": 5, "introduct": 6, "custom": 6, "databas": 6, "api": 6, "modul": [6, 7, 9], "search": 6, "page": 6, "packag": [7, 9], "code": 7, "work": 7, "ig": 7, "tr": 7, "receptor": 7, "pip": 7, "biopython": [7, 11], "separ": 7, "deal": 7, "novel": [7, 10, 11], "context": 7, "express": 7, "genom": 7, "know": 7, "snp": [7, 10], "alli": 7, "task": 7, "name_allel": [7, 13], "nearest": 7, "taken": 7, "suppli": 7, "incorpor": 7, "scheme": 7, "accord": 7, "residu": 7, "emploi": 7, "must": [7, 11], "each": 7, "next": 7, "smaller": 7, "come": 7, "exampl": 7, "nt": 7, "noth": 7, "special": 7, "save": 7, "time": 7, "simpl": [7, 11], "wrapper": [7, 11], "around": [7, 11], "commonli": 7, "suit": 7, "write": [7, 11], "comma": 7, "tab": 7, "manag": 7, "string": [7, 11], "dict": [7, 11], "simplifi": 7, "compar": [7, 10], "expens": 7, "flexibl": 7, "rare": 7, "document": 7, "variabl": 8, "region": [8, 11], "chain": [8, 11], "ndm_file": 8, "eg": 8, "vh": 8, "receptor_util": 9, "submodul": 9, "novel_allele_nam": 9, "name_novel": [9, 11], "number_ighv": 9, "simple_bio_seq": 9, "chunk": [9, 11], "closest_ref": [9, 11], "dumb_consensu": [9, 11], "nt_diff": [9, 11], "read_csv": [9, 11], "read_fasta": [9, 11], "read_imgt_fasta": [9, 11], "read_single_fasta": [9, 11], "reverse_compl": [9, 11], "sample_fasta": [9, 11], "scored_consensu": [9, 11], "scored_gap_consensu": [9, 11], "toseqrecord": [9, 11], "translat": [9, 11], "write_csv": [9, 11], "write_fasta": [9, 11], "content": 9, "construct": 10, "gene_nam": 10, "ref_set": [10, 11], "input": [10, 11, 12], "consid": 10, "rev_comp": [10, 13], "befor": 10, "process": 10, "novel_seq": 11, "str": 11, "v_gene": 11, "bool": 11, "true": 11, "sourc": 11, "its": 11, "conform": 11, "descript": 11, "prime": 11, "return": [11, 12], "we": 11, "tupl": 11, "consist": 11, "three": 11, "novel_nam": 11, "note": 11, "type": 11, "bioseq": 11, "applic": 11, "principl": 11, "store": 11, "collectiopn": 11, "coerc": 11, "upper": 11, "iter": 11, "corec": 11, "eas": 11, "debug": 11, "basic": 11, "thing": 11, "where": 11, "don": 11, "n": 11, "int": 11, "yield": 11, "success": 11, "size": 11, "seq": 11, "ref": 11, "entri": 11, "determin": 11, "equal": 11, "threshold": 11, "float": 11, "0": 11, "dumb": 11, "consensu": 11, "bio": 11, "aligninfo": 11, "summaryinfo": 11, "same": 11, "valu": 11, "atom": 11, "s1": 11, "s2": 11, "differ": 11, "shorter": 11, "delimit": 11, "none": 11, "produc": 11, "dictread": 11, "filenam": 11, "infil": 11, "pathnam": 11, "dictionari": 11, "id": 11, "ighv": 11, "ighd": 11, "ighj": 11, "ch": 11, "space": 11, "replac": 11, "underscor": 11, "mark": 11, "serquenc": 11, "singl": 11, "nucelotid": 11, "random": 11, "resampl": 11, "valueerror": 11, "enough": 11, "gap_consensu": 11, "mimimum": 11, "score": 11, "achiev": 11, "min_scor": 11, "ambigu": 11, "x": 11, "require_multipl": 11, "adapt": 11, "seqrecord": 11, "truncat": 11, "ignore_partial_codon": 11, "amino": 11, "acid": 11, "so": 11, "termin": 11, "boundari": 11, "otherwis": 11, "pad": 11, "ensur": 11, "get": 11, "row": 11, "header": 11, "kei": 11, "item": 11, "outfil": 11, "written": 11, "extract_ref": 13, "identical_seq": 13, "at_coord": 13}, "objects": {"": [[11, 0, 0, "-", "receptor_utils"]], "receptor_utils": [[11, 0, 0, "-", "novel_allele_name"], [11, 0, 0, "-", "number_ighv"], [11, 0, 0, "-", "simple_bio_seq"]], "receptor_utils.novel_allele_name": [[11, 1, 1, "", "name_novel"]], "receptor_utils.simple_bio_seq": [[11, 1, 1, "", "chunks"], [11, 1, 1, "", "closest_ref"], [11, 1, 1, "", "dumb_consensus"], [11, 1, 1, "", "nt_diff"], [11, 1, 1, "", "read_csv"], [11, 1, 1, "", "read_fasta"], [11, 1, 1, "", "read_imgt_fasta"], [11, 1, 1, "", "read_single_fasta"], [11, 1, 1, "", "reverse_complement"], [11, 1, 1, "", "sample_fasta"], [11, 1, 1, "", "scored_consensus"], [11, 1, 1, "", "scored_gap_consensus"], [11, 1, 1, "", "toSeqRecords"], [11, 1, 1, "", "translate"], [11, 1, 1, "", "write_csv"], [11, 1, 1, "", "write_fasta"]]}, "objtypes": {"0": "py:module", "1": "py:function"}, "objnames": {"0": ["py", "module", "Python module"], "1": ["py", "function", "Python function"]}, "titleterms": {"annotate_j": 0, "posit": [0, 1, 3, 4, 5, 8, 10, 12], "argument": [0, 1, 3, 4, 5, 8, 10, 12], "at_coord": 1, "name": [1, 3, 7, 10], "us": [2, 7], "custom": [2, 7], "databas": [2, 7], "igblast": [2, 7, 13], "extract_ref": 3, "gap_sequ": 4, "identical_seq": 5, "receptor": 6, "util": [6, 13], "code": 6, "work": [6, 13], "ig": 6, "tr": 6, "sequenc": [6, 7], "content": [6, 11], "indic": 6, "tabl": 6, "introduct": 7, "instal": 7, "align": 7, "v": 7, "d": 7, "j": 7, "conveni": 7, "tool": 7, "simple_bio_seq": [7, 11], "api": [7, 9], "make_igblast_ndm": 8, "name_allel": 10, "receptor_util": 11, "packag": 11, "submodul": 11, "novel_allele_nam": 11, "modul": 11, "number_ighv": 11, "rev_comp": 12, "command": 13, "line": 13, "fasta": 13, "file": 13, "miscellan": 13}, "envversion": {"sphinx.domains.c": 2, "sphinx.domains.changeset": 1, "sphinx.domains.citation": 1, "sphinx.domains.cpp": 8, "sphinx.domains.index": 1, "sphinx.domains.javascript": 2, "sphinx.domains.math": 2, "sphinx.domains.python": 3, "sphinx.domains.rst": 2, "sphinx.domains.std": 2, "sphinx.ext.viewcode": 1, "sphinx": 57}, "alltitles": {"annotate_j": [[0, "annotate-j"]], "Positional Arguments": [[0, "positional-arguments"], [1, "positional-arguments"], [3, "positional-arguments"], [4, "positional-arguments"], [5, "positional-arguments"], [8, "positional-arguments"], [10, "positional-arguments"], [12, "positional-arguments"]], "at_coords": [[1, "at-coords"]], "Named Arguments": [[1, "named-arguments"], [3, "named-arguments"], [10, "named-arguments"]], "extract_refs": [[3, "extract-refs"]], "gap_sequences": [[4, "gap-sequences"]], "identical_seqs": [[5, "identical-seqs"]], "Receptor-utils: code and utilities for working with IG/TR receptor sequences": [[6, "receptor-utils-code-and-utilities-for-working-with-ig-tr-receptor-sequences"]], "Contents:": [[6, null]], "Indices and tables": [[6, "indices-and-tables"]], "Using custom databases with IgBlast": [[7, "using-custom-databases-with-igblast"], [2, "using-custom-databases-with-igblast"]], "Introduction": [[7, "introduction"]], "Installation": [[7, "installation"]], "Naming and aligning V, D and J sequences": [[7, "naming-and-aligning-v-d-and-j-sequences"]], "Convenience Tools": [[7, "convenience-tools"]], "simple_bio_seq API": [[7, "simple-bio-seq-api"]], "make_igblast_ndm": [[8, "make-igblast-ndm"]], "API": [[9, "api"]], "name_allele": [[10, "name-allele"]], "receptor_utils package": [[11, "receptor-utils-package"]], "Submodules": [[11, "submodules"]], "receptor_utils.novel_allele_name module": [[11, "module-receptor_utils.novel_allele_name"]], "receptor_utils.number_ighv module": [[11, "module-receptor_utils.number_ighv"]], "receptor_utils.simple_bio_seq module": [[11, "module-receptor_utils.simple_bio_seq"]], "Module contents": [[11, "module-receptor_utils"]], "rev_comp": [[12, "rev-comp"]], "Command-Line Utilities": [[13, "command-line-utilities"]], "Working with IgBLAST": [[13, "working-with-igblast"]], "Working with FASTA files": [[13, "working-with-fasta-files"]], "Miscellaneous": [[13, "miscellaneous"]]}, "indexentries": {}})