Search.setIndex({"docnames": ["annotate_j", "at_coords", "extract_refs", "gap_sequences", "identical_seqs", "index", "make_igblast_ndm", "modules", "name_allele", "receptor_utils", "reverse_complement", "scripts"], "filenames": ["annotate_j.rst", "at_coords.rst", "extract_refs.rst", "gap_sequences.rst", "identical_seqs.rst", "index.rst", "make_igblast_ndm.rst", "modules.rst", "name_allele.rst", "receptor_utils.rst", "reverse_complement.rst", "scripts.rst"], "titles": ["annotate_j", "at_coords", "extract_refs", "gap_sequences", "identical_seqs", "Receptor-utils: code and utilities for working with IG/TR receptor sequences", "make_igblast_ndm", "API", "name_allele", "receptor_utils package", "rev_comp", "Command-Line Utilities"], "terms": {"annot": 0, "set": [0, 3, 6, 8, 9], "j": [0, 9], "sequenc": [0, 1, 3, 4, 8, 9, 10], "find": [0, 9], "frame": 0, "align": [0, 9], "1": [0, 1, 9], "base": [0, 1], "index": [0, 5, 9], "first": [0, 9], "nucleotid": [0, 9, 10], "conserv": 0, "phe": 0, "trp": 0, "usag": [0, 1, 2, 3, 4, 6, 8, 10], "h": [0, 1, 2, 3, 4, 6, 8, 10], "seq_fil": 0, "out_fil": [0, 3], "fasta": [0, 1, 3, 4, 6, 8, 9], "output": [0, 3, 6, 9], "aux": 0, "format": [0, 9], "us": [0, 3, 9], "igblast": [0, 6], "extract": [1, 2], "specifi": [1, 8, 9], "co": 1, "ordin": 1, "file": [1, 2, 3, 4, 6, 9], "r": [1, 8], "input_fil": [1, 4], "start": 1, "end": [1, 9], "contain": [1, 3, 8, 9], "refer": [1, 2, 3, 4, 6, 8, 9], "ord": 1, "revers": [1, 8, 9, 10], "complement": [1, 8, 9, 10], "result": 1, "default": [1, 2, 8, 9], "fals": [1, 2, 8, 9], "thi": [2, 9], "util": 2, "section": 2, "from": [2, 6, 8, 9], "an": [2, 6, 8, 9], "imgt": [2, 4, 6, 8, 9], "gap": [2, 3, 4, 6, 8, 9], "The": [2, 9], "latest": 2, "can": 2, "download": 2, "here": [2, 9], "nomin": 2, "speci": [2, 9], "l": [2, 9], "locu": 2, "f": [2, 9], "imgt_fil": 2, "species_nam": 2, "quot": 2, "e": [2, 9], "g": [2, 8, 9], "homo": 2, "sapien": 2, "macaca": 2, "mulatta": 2, "identifi": 2, "igh": 2, "igk": 2, "trb": 2, "trg": 2, "functional_onli": [2, 9], "ina": 3, "closest": [3, 9], "templat": 3, "inferred_fil": 3, "ref_fil": [3, 6], "ungap": [3, 9], "infer": 3, "list": [4, 9], "ident": 4, "sub": 4, "identical_sequ": 4, "command": 5, "line": 5, "api": 5, "modul": [5, 7], "search": 5, "page": 5, "make": [6, 9], "ndm": 6, "variabl": 6, "region": [6, 9], "chain": [6, 9], "ndm_file": 6, "requir": [6, 9], "eg": 6, "vh": 6, "receptor_util": 7, "packag": 7, "submodul": 7, "novel_allele_nam": 7, "name_novel": [7, 9], "number_ighv": 7, "simple_bio_seq": 7, "chunk": [7, 9], "closest_ref": [7, 9], "dumb_consensu": [7, 9], "nt_diff": [7, 9], "read_csv": [7, 9], "read_fasta": [7, 9], "read_imgt_fasta": [7, 9], "read_single_fasta": [7, 9], "reverse_compl": [7, 9], "sample_fasta": [7, 9], "scored_consensu": [7, 9], "scored_gap_consensu": [7, 9], "toseqrecord": [7, 9], "translat": [7, 9], "write_csv": [7, 9], "write_fasta": [7, 9], "content": 7, "allel": [8, 9], "compar": 8, "construct": 8, "novel": [8, 9], "snp": 8, "necessari": [8, 9], "gene_nam": 8, "ref_set": [8, 9], "v": [8, 9], "gene": [8, 9], "should": [8, 9], "input": [8, 9, 10], "onli": [8, 9], "consid": 8, "rev_comp": [8, 11], "befor": 8, "process": 8, "novel_seq": 9, "str": 9, "dict": 9, "v_gene": 9, "bool": 9, "true": 9, "sourc": 9, "name": 9, "given": 9, "its": 9, "conform": 9, "descript": 9, "must": 9, "full": 9, "length": 9, "5": 9, "prime": 9, "paramet": 9, "which": 9, "mai": 9, "either": 9, "case": 9, "return": [9, 10], "we": 9, "ar": 9, "tupl": 9, "consist": 9, "three": 9, "string": 9, "novel_nam": 9, "note": 9, "i": 9, "type": 9, "wrapper": 9, "around": 9, "bioseq": 9, "function": 9, "simpl": 9, "applic": 9, "principl": 9, "store": 9, "collectiopn": 9, "all": 9, "coerc": 9, "upper": 9, "iter": 9, "corec": 9, "eas": 9, "debug": 9, "basic": 9, "just": 9, "thing": 9, "where": 9, "don": 9, "t": 9, "need": 9, "do": 9, "more": 9, "n": 9, "int": 9, "yield": 9, "success": 9, "size": 9, "seq": 9, "ref": 9, "entri": 9, "other": 9, "determin": 9, "were": 9, "equal": 9, "close": 9, "threshold": 9, "float": 9, "0": 9, "7": 9, "dumb": 9, "consensu": 9, "bio": 9, "aligninfo": 9, "summaryinfo": 9, "same": 9, "analys": 9, "valu": 9, "add": 9, "particular": 9, "atom": 9, "s1": 9, "s2": 9, "count": 9, "posit": 9, "differ": 9, "If": 9, "cover": 9, "shorter": 9, "second": 9, "number": 9, "delimit": 9, "option": 9, "none": 9, "read": 9, "produc": 9, "dictread": 9, "filenam": 9, "infil": 9, "pathnam": 9, "A": 9, "dictionari": 9, "id": 9, "ighv": 9, "ighd": 9, "ighj": 9, "ch": 9, "d": 9, "one": 9, "space": 9, "replac": 9, "underscor": 9, "mark": 9, "serquenc": 9, "singl": 9, "nucelotid": 9, "random": 9, "sampl": 9, "resampl": 9, "valueerror": 9, "doe": 9, "enough": 9, "gap_consensu": 9, "modifi": 9, "provid": 9, "well": 9, "mimimum": 9, "score": 9, "achiev": 9, "ani": 9, "min_scor": 9, "ambigu": 9, "x": 9, "require_multipl": 9, "adapt": 9, "biopython": 9, "call": 9, "convert": 9, "seqrecord": 9, "truncat": 9, "ignore_partial_codon": 9, "amino": 9, "acid": 9, "so": 9, "termin": 9, "codon": 9, "boundari": 9, "otherwis": 9, "pad": 9, "entir": 9, "ensur": 9, "get": 9, "row": 9, "write": 9, "header": 9, "kei": 9, "item": 9, "creat": 9, "outfil": 9, "written": 9, "record": 9, "annotate_j": 11, "make_igblast_ndm": 11, "extract_ref": 11, "gap_sequ": 11, "identical_seq": 11, "at_coord": 11, "name_allel": 11}, "objects": {"": [[9, 0, 0, "-", "receptor_utils"]], "receptor_utils": [[9, 0, 0, "-", "novel_allele_name"], [9, 0, 0, "-", "number_ighv"], [9, 0, 0, "-", "simple_bio_seq"]], "receptor_utils.novel_allele_name": [[9, 1, 1, "", "name_novel"]], "receptor_utils.simple_bio_seq": [[9, 1, 1, "", "chunks"], [9, 1, 1, "", "closest_ref"], [9, 1, 1, "", "dumb_consensus"], [9, 1, 1, "", "nt_diff"], [9, 1, 1, "", "read_csv"], [9, 1, 1, "", "read_fasta"], [9, 1, 1, "", "read_imgt_fasta"], [9, 1, 1, "", "read_single_fasta"], [9, 1, 1, "", "reverse_complement"], [9, 1, 1, "", "sample_fasta"], [9, 1, 1, "", "scored_consensus"], [9, 1, 1, "", "scored_gap_consensus"], [9, 1, 1, "", "toSeqRecords"], [9, 1, 1, "", "translate"], [9, 1, 1, "", "write_csv"], [9, 1, 1, "", "write_fasta"]]}, "objtypes": {"0": "py:module", "1": "py:function"}, "objnames": {"0": ["py", "module", "Python module"], "1": ["py", "function", "Python function"]}, "titleterms": {"annotate_j": 0, "posit": [0, 1, 2, 3, 4, 6, 8, 10], "argument": [0, 1, 2, 3, 4, 6, 8, 10], "at_coord": 1, "name": [1, 2, 8], "extract_ref": 2, "gap_sequ": 3, "identical_seq": 4, "receptor": 5, "util": [5, 11], "code": 5, "work": [5, 11], "ig": 5, "tr": 5, "sequenc": 5, "content": [5, 9], "indic": 5, "tabl": 5, "make_igblast_ndm": 6, "api": 7, "name_allel": 8, "receptor_util": 9, "packag": 9, "submodul": 9, "novel_allele_nam": 9, "modul": 9, "number_ighv": 9, "simple_bio_seq": 9, "rev_comp": 10, "command": 11, "line": 11, "igblast": 11, "fasta": 11, "file": 11, "miscellan": 11}, "envversion": {"sphinx.domains.c": 2, "sphinx.domains.changeset": 1, "sphinx.domains.citation": 1, "sphinx.domains.cpp": 8, "sphinx.domains.index": 1, "sphinx.domains.javascript": 2, "sphinx.domains.math": 2, "sphinx.domains.python": 3, "sphinx.domains.rst": 2, "sphinx.domains.std": 2, "sphinx.ext.viewcode": 1, "sphinx": 57}, "alltitles": {"annotate_j": [[0, "annotate-j"]], "Positional Arguments": [[0, "positional-arguments"], [1, "positional-arguments"], [2, "positional-arguments"], [3, "positional-arguments"], [4, "positional-arguments"], [6, "positional-arguments"], [8, "positional-arguments"], [10, "positional-arguments"]], "at_coords": [[1, "at-coords"]], "Named Arguments": [[1, "named-arguments"], [2, "named-arguments"], [8, "named-arguments"]], "extract_refs": [[2, "extract-refs"]], "gap_sequences": [[3, "gap-sequences"]], "identical_seqs": [[4, "identical-seqs"]], "Receptor-utils: code and utilities for working with IG/TR receptor sequences": [[5, "receptor-utils-code-and-utilities-for-working-with-ig-tr-receptor-sequences"]], "Contents:": [[5, null]], "Indices and tables": [[5, "indices-and-tables"]], "make_igblast_ndm": [[6, "make-igblast-ndm"]], "API": [[7, "api"]], "name_allele": [[8, "name-allele"]], "receptor_utils package": [[9, "receptor-utils-package"]], "Submodules": [[9, "submodules"]], "receptor_utils.novel_allele_name module": [[9, "module-receptor_utils.novel_allele_name"]], "receptor_utils.number_ighv module": [[9, "module-receptor_utils.number_ighv"]], "receptor_utils.simple_bio_seq module": [[9, "module-receptor_utils.simple_bio_seq"]], "Module contents": [[9, "module-receptor_utils"]], "rev_comp": [[10, "rev-comp"]], "Command-Line Utilities": [[11, "command-line-utilities"]], "Working with IgBLAST": [[11, "working-with-igblast"]], "Working with FASTA files": [[11, "working-with-fasta-files"]], "Miscellaneous": [[11, "miscellaneous"]]}, "indexentries": {"chunks() (in module receptor_utils.simple_bio_seq)": [[9, "receptor_utils.simple_bio_seq.chunks"]], "closest_ref() (in module receptor_utils.simple_bio_seq)": [[9, "receptor_utils.simple_bio_seq.closest_ref"]], "dumb_consensus() (in module receptor_utils.simple_bio_seq)": [[9, "receptor_utils.simple_bio_seq.dumb_consensus"]], "module": [[9, "module-receptor_utils"], [9, "module-receptor_utils.novel_allele_name"], [9, "module-receptor_utils.number_ighv"], [9, "module-receptor_utils.simple_bio_seq"]], "name_novel() (in module receptor_utils.novel_allele_name)": [[9, "receptor_utils.novel_allele_name.name_novel"]], "nt_diff() (in module receptor_utils.simple_bio_seq)": [[9, "receptor_utils.simple_bio_seq.nt_diff"]], "read_csv() (in module receptor_utils.simple_bio_seq)": [[9, "receptor_utils.simple_bio_seq.read_csv"]], "read_fasta() (in module receptor_utils.simple_bio_seq)": [[9, "receptor_utils.simple_bio_seq.read_fasta"]], "read_imgt_fasta() (in module receptor_utils.simple_bio_seq)": [[9, "receptor_utils.simple_bio_seq.read_imgt_fasta"]], "read_single_fasta() (in module receptor_utils.simple_bio_seq)": [[9, "receptor_utils.simple_bio_seq.read_single_fasta"]], "receptor_utils": [[9, "module-receptor_utils"]], "receptor_utils.novel_allele_name": [[9, "module-receptor_utils.novel_allele_name"]], "receptor_utils.number_ighv": [[9, "module-receptor_utils.number_ighv"]], "receptor_utils.simple_bio_seq": [[9, "module-receptor_utils.simple_bio_seq"]], "reverse_complement() (in module receptor_utils.simple_bio_seq)": [[9, "receptor_utils.simple_bio_seq.reverse_complement"]], "sample_fasta() (in module receptor_utils.simple_bio_seq)": [[9, "receptor_utils.simple_bio_seq.sample_fasta"]], "scored_consensus() (in module receptor_utils.simple_bio_seq)": [[9, "receptor_utils.simple_bio_seq.scored_consensus"]], "scored_gap_consensus() (in module receptor_utils.simple_bio_seq)": [[9, "receptor_utils.simple_bio_seq.scored_gap_consensus"]], "toseqrecords() (in module receptor_utils.simple_bio_seq)": [[9, "receptor_utils.simple_bio_seq.toSeqRecords"]], "translate() (in module receptor_utils.simple_bio_seq)": [[9, "receptor_utils.simple_bio_seq.translate"]], "write_csv() (in module receptor_utils.simple_bio_seq)": [[9, "receptor_utils.simple_bio_seq.write_csv"]], "write_fasta() (in module receptor_utils.simple_bio_seq)": [[9, "receptor_utils.simple_bio_seq.write_fasta"]]}})