Search.setIndex({"docnames": ["airrc_sets_with_igblast", "airrc_sets_with_mixcr", "annotate_j", "at_coords", "custom_igblast", "download_germline_set", "extract_refs", "fix_macaque_gaps", "gap_sequences", "identical_seqs", "index", "introduction", "make_igblast_ndm", "merge_fasta", "modules", "name_allele", "receptor_utils", "rev_comp", "scripts"], "filenames": ["airrc_sets_with_igblast.rst", "airrc_sets_with_mixcr.rst", "annotate_j.rst", "at_coords.rst", "custom_igblast.rst", "download_germline_set.rst", "extract_refs.rst", "fix_macaque_gaps.rst", "gap_sequences.rst", "identical_seqs.rst", "index.rst", "introduction.rst", "make_igblast_ndm.rst", "merge_fasta.rst", "modules.rst", "name_allele.rst", "receptor_utils.rst", "rev_comp.rst", "scripts.rst"], "titles": ["Using AIRR Community Reference Sets with IgBLAST", "Using AIRR Community Reference Sets with MiXCR", "annotate_j", "at_coords", "Using custom databases with IgBlast", "download_germline_set", "extract_refs", "fix_macaque_gaps", "gap_sequences", "identical_seqs", "Receptor-utils: code and utilities for working with IG/TR receptor sequences", "Introduction", "make_igblast_ndm", "merge_fasta", "receptor_utils", "name_allele", "receptor_utils package", "rev_comp", "Command-Line Utilities"], "terms": {"The": [0, 1, 4, 5, 6, 8, 11, 13, 16], "download_germline_set": [0, 1, 11, 18], "util": [0, 1, 5, 6, 7, 11, 13], "i": [0, 1, 3, 4, 5, 7, 8, 11, 12, 16], "design": 0, "make": [0, 4, 12, 16], "easi": 0, "possibl": [0, 4, 5], "from": [0, 1, 3, 4, 5, 6, 7, 11, 12, 13, 15, 16], "open": [0, 4, 5, 11], "receptor": [0, 1, 5, 11], "ogrdb": [0, 1, 5, 11, 12], "compat": 0, "creat": [0, 1, 4, 5, 11, 15, 16], "necessari": [0, 4, 15, 16], "auxiliari": [0, 11], "In": [0, 1, 4, 8, 12], "thi": [0, 1, 4, 5, 6, 7, 11, 12, 13, 16], "section": [0, 4, 6, 11], "describ": [0, 4, 11], "how": [0, 4], "small": [0, 7, 8], "exampl": [0, 11], "dataset": 0, "accompani": 0, "For": [0, 1, 4, 5, 11], "simplic": [0, 1], "docker": [0, 1], "singular": [0, 1], "togeth": 0, "other": [0, 1, 4, 7, 8, 11, 16], "If": [0, 1, 4, 5, 6, 12, 16], "you": [0, 1, 4, 7, 12], "ar": [0, 1, 4, 5, 8, 11, 12, 13, 16], "alreadi": 0, "familiar": [0, 1], "read": [0, 1, 4, 11, 16], "about": [0, 1], "document": [0, 1, 11], "altern": [0, 1, 12], "your": [0, 1, 4, 12], "local": [0, 1, 4], "machin": [0, 1], "skip": [0, 1], "step": [0, 1, 4], "below": 0, "To": 0, "instruct": [0, 1], "provid": [0, 1, 4, 5, 11, 16], "websit": [0, 1], "pleas": [0, 5, 12], "sure": 0, "recent": 0, "version": [0, 4, 5], "procedur": [0, 4], "custom": [0, 5, 10], "ha": [0, 1, 11], "been": [0, 1, 4], "significantli": [0, 4], "simplifi": [0, 11], "20": [0, 4], "onward": 0, "have": [0, 1, 4, 7, 12], "preprocess": [0, 1], "qualiti": [0, 1, 4], "filter": [0, 1], "illumina": [0, 1], "pair": [0, 1], "end": [0, 1, 3, 8, 16], "present": [0, 1, 7, 13], "fasta": [0, 1, 2, 3, 5, 6, 7, 8, 9, 11, 12, 13, 15, 16], "call": [0, 1, 4, 16], "hd13m": [0, 1], "process": [0, 1, 15], "requir": [0, 1, 4, 11, 12, 16], "depend": [0, 1, 4, 12], "protocol": [0, 1], "consult": [0, 1], "presto": 0, "sourc": [0, 1, 16], "determin": [0, 1, 5, 16], "suitabl": [0, 1, 15], "approach": [0, 1, 4], "data": [0, 1, 4, 11], "befor": [0, 1, 7, 15], "start": [0, 1, 3, 4, 12, 16], "need": [0, 1, 4, 5, 11, 12, 16], "follow": [0, 1, 4, 12], "python": [0, 1], "9": [0, 1], "abov": [0, 1, 4, 5, 11], "packag": [0, 1, 11, 14], "see": [0, 1, 4, 5], "introduct": [0, 1, 10], "extract": [0, 1, 3, 6], "tarbal": [0, 1], "which": [0, 1, 4, 5, 11, 16], "here": [0, 1, 6, 11, 16], "shall": [0, 1], "directori": [0, 1], "test": [0, 1, 4], "homo": [0, 1, 5, 6], "sapien": [0, 1, 5, 6], "igh": [0, 1, 5, 6, 7, 15], "f": [0, 1, 5, 6, 16], "multi": [0, 1, 5], "http": [0, 1, 5, 6], "org": [0, 1, 5, 6], "api_v2": [0, 1, 5], "speci": [0, 1, 5, 6, 7, 11, 16], "9606": [0, 1], "igh_vdj": [0, 1], "save": [0, 1], "homo_sapiens_igh_v": [0, 1], "homo_sapiens_igh_d": [0, 1], "homo_sapiens_igh_j": [0, 1], "homo_sapiens_igh_v_gap": [0, 1], "ndm": [0, 4, 5, 11, 12], "homo_sapiens_igh": 0, "aux": [0, 2, 4, 5], "final": [0, 1, 5], "copi": [0, 1], "mount": [0, 1], "current": [0, 1, 6], "linux": [0, 1], "run": [0, 1], "v": [0, 1, 2, 4, 5, 12, 15, 16], "pwd": [0, 1, 5], "z": 0, "suit": [0, 11], "0": [0, 16], "bash": [0, 1], "window": [0, 1], "cd": [0, 1], "further": [0, 5], "option": [0, 4, 12, 16], "e": [0, 4, 5, 6, 15, 16], "g": [0, 5, 6, 15, 16], "onc": [0, 1, 4, 11], "check": [0, 1, 4, 7], "root": 0, "b426e4d7c0a": 0, "l": [0, 1, 6, 16], "": [0, 1, 7], "makeblastdb": 0, "parse_seqid": 0, "dbtype": 0, "nucl": 0, "out": 0, "after": [0, 1], "command": [0, 1, 4, 5, 10], "mani": [0, 4], "more": [0, 3, 4, 11, 16], "nhr": 0, "nj": 0, "nos": 0, "nsq": 0, "nto": 0, "ndb": 0, "nin": 0, "nog": 0, "ntf": 0, "verbos": [0, 2], "7": [0, 4, 16], "std": [0, 4], "qseq": [0, 4], "sseq": [0, 4], "btop": [0, 4], "igdata": 0, "environ": 0, "variabl": [0, 12], "fact": 0, "overrid": 0, "them": [0, 4, 7], "line": [0, 4, 10, 11], "order": 0, "export": [0, 1], "usr": 0, "share": 0, "igblastn": 0, "germline_db_v": 0, "germline_db_d": 0, "germline_db_j": 0, "auxiliary_data": [0, 4], "custom_internal_data": [0, 4], "domain_system": 0, "imgt": [0, 4, 6, 7, 9, 11, 12, 15, 16], "outfmt": [0, 4], "num_thread": 0, "queri": 0, "fmt7": 0, "examin": [0, 4], "witn": 0, "anoth": [0, 4, 5], "confirm": [0, 7], "makedb": [0, 4], "tsv": [0, 1], "note": [0, 5, 16], "fail": [0, 4], "captur": 0, "detail": [0, 4, 5, 11], "ani": [0, 4, 5, 8, 11, 13, 16], "import": [0, 4], "some": [0, 1, 4, 12], "guidanc": 0, "error": [0, 4, 5], "py": [0, 4], "r": [0, 3, 6, 15], "extend": [0, 4, 8], "should": [0, 4, 15, 16], "similar": [0, 4], "hd13m_db": 0, "pass": [0, 4], "7887": 0, "1402": 0, "review": [0, 12], "result": [0, 1, 3, 4], "case": [1, 4, 11, 12, 16], "just": [1, 4, 16], "walkthrough": 1, "immcant": 1, "file": [1, 3, 4, 5, 6, 7, 8, 9, 11, 12, 13, 16], "addit": [4, 5, 7, 11], "igblast": [2, 5, 10, 12], "format": [1, 2, 4, 12, 16], "work": [1, 11], "ghcr": 1, "io": 1, "milaboratori": 1, "latest": [1, 5, 6], "4": [], "licens": 1, "match": [1, 4, 13], "kei": [1, 16], "mi_licens": 1, "licenc": 1, "buildlibrari": 1, "debug": [1, 16], "featur": [1, 7], "vregion": 1, "j": [1, 2, 4, 5, 16], "d": [1, 4, 5, 16], "chain": [1, 7, 12, 16], "taxon": 1, "id": [1, 11, 13, 16], "json": [1, 2, 5, 12], "gz": 1, "mai": [1, 4, 7, 12, 16], "warn": [1, 4, 8, 11], "dure": 1, "stop": [1, 4], "codon": [1, 4, 7, 11, 12, 16], "were": [1, 4, 16], "found": [1, 4, 5], "expect": [1, 4, 7], "pseudogen": [1, 4], "includ": [1, 4, 5, 11, 16], "c": [1, 2, 5, 12], "No": 1, "action": 1, "As": 1, "plain": 1, "gener": 1, "pacbio": 1, "templat": [1, 8, 11], "analyz": 1, "librari": 1, "assembl": 1, "clonotyp": 1, "fr1": 1, "cdr1": [1, 12], "fr2": 1, "cdr2": [1, 12], "fr3": 1, "cdr3": [1, 12], "fr4": 1, "view": 1, "clones_igh": 1, "annot": [2, 4, 5], "set": [2, 4, 5, 7, 8, 10, 11, 12, 15, 16], "sequenc": [2, 3, 4, 5, 7, 8, 9, 12, 13, 15, 16, 17], "find": [2, 4, 11, 16], "frame": [2, 4], "align": [2, 4, 7, 12, 16], "1": [2, 3, 4, 12, 13, 16], "base": [2, 3, 12], "index": [2, 10, 16], "first": [2, 13, 16], "nucleotid": [2, 11, 16, 17], "conserv": [2, 7, 11], "phe": 2, "trp": 2, "usag": [2, 3, 5, 6, 7, 8, 9, 12, 13, 15, 17], "h": [2, 3, 5, 6, 7, 8, 9, 12, 13, 15, 17], "seq_fil": 2, "out_fil": [2, 8], "airr": [2, 5, 10, 12], "output": [2, 4, 8, 12, 13, 16], "us": [2, 5, 7, 8, 10, 12, 15, 16], "default": [2, 3, 5, 6, 12, 15, 16], "fals": [2, 3, 6, 15, 16], "specifi": [3, 4, 5, 6, 12, 15, 16], "co": [3, 7, 11, 12], "ordin": [3, 7, 11, 12], "n": [3, 5, 16], "xr": 3, "input_fil": [3, 9], "contain": [3, 4, 5, 7, 8, 11, 12, 13, 15, 16], "refer": [3, 4, 5, 6, 8, 9, 10, 11, 12, 15, 16], "coord": 3, "label": 3, "must": [3, 11, 12, 16], "than": [3, 4, 11, 13], "one": [3, 4, 11, 12, 16], "revers": [3, 11, 15, 16, 17], "complement": [3, 11, 15, 16, 17], "instal": 4, "can": [4, 5, 6, 8, 11, 12], "configur": [4, 5], "support": [4, 11], "organ": 4, "either": [4, 16], "an": [4, 5, 6, 12, 15, 16], "tool": [4, 5, 7, 8, 12], "modifi": [4, 16], "simplif": 4, "upgrad": 4, "least": [4, 8], "encourag": 4, "full": [4, 5, 16], "two": [4, 11, 13], "germlin": [4, 5, 11, 12], "gene": [4, 6, 15, 16], "orient": 4, "posit": [4, 16], "cdr": [4, 5, 7, 12, 16], "wish": [4, 7], "add": [4, 16], "allel": [4, 11, 15, 16], "exist": 4, "all": [4, 5, 12, 13, 16], "blast": 4, "link": 4, "new": 4, "paramet": [4, 16], "make_igblast_ndm": [4, 5, 11, 18], "gap": [4, 5, 6, 7, 8, 9, 11, 12, 15, 16], "gap_sequ": [4, 11, 14, 16, 18], "partial": 4, "close": [4, 16], "relat": 4, "annotate_j": [4, 5, 11, 18], "It": [4, 11], "veri": 4, "carefulli": 4, "when": 4, "issu": [4, 8], "particularli": 4, "substanti": 4, "proport": 4, "submit": 4, "my": [4, 11], "own": 4, "repertoir": [4, 7], "convert": [4, 16], "changeo": 4, "record": [4, 13, 16], "log": 4, "count": [4, 16], "typic": 4, "might": 4, "under": [4, 11], "10": 4, "possibli": 4, "fewer": 4, "A": [4, 7, 12, 16], "higher": 4, "sign": 4, "db_fail": 4, "discount": 4, "what": 4, "do": [4, 12, 13, 16], "thei": [4, 8, 11], "repres": 4, "few": 4, "percent": 4, "obvious": 4, "short": 4, "length": [4, 16], "again": 4, "take": 4, "sampl": [4, 12, 16], "look": 4, "though": 4, "analys": [4, 12, 16], "quest": [4, 12], "interest": 4, "doe": [4, 16], "non": [4, 8], "function": [4, 8, 11, 16], "number": [4, 7, 12, 16], "strongli": 4, "suspect": 4, "problem": [4, 7, 11], "report": [4, 11], "analysi": [4, 12], "mi": 4, "junction": 4, "assign": 4, "mean": [4, 5], "inde": 4, "part": 4, "miss": [4, 11], "There": 4, "simpli": 4, "too": 4, "cover": [4, 11, 16], "entir": [4, 16], "like": [4, 12], "incorrectli": 4, "empti": 4, "310": 4, "occur": 4, "corrupt": 4, "indic": 4, "circumst": [4, 8], "recognis": 4, "being": 4, "canon": 4, "someth": [4, 7], "along": 4, "ighv1": 4, "69": 4, "02": 4, "latitud": 4, "allow": 4, "name": [4, 16], "depart": 4, "caus": [4, 7, 11], "particular": [4, 16], "without": [4, 5, 7], "suffix": 4, "identifi": [4, 6, 11], "resolv": 4, "excel": 4, "reason": 4, "famili": 4, "arrang": 4, "investig": 4, "except": 4, "continu": 4, "exercis": 4, "caution": 4, "consciou": 4, "hope": 4, "doesn": 4, "t": [4, 16], "put": 4, "off": 4, "But": 4, "serv": 4, "systemat": 4, "lookout": 4, "well": [4, 16], "flag": 4, "also": [4, 5, 6], "understand": 4, "underli": 4, "download": [5, 6, 11], "databas": [5, 10], "airrc": 5, "singl": [5, 15, 16], "fg": 5, "fu": 5, "u": 5, "url": 5, "p": 5, "prefix": 5, "locu": [5, 6, 15], "igk": [5, 6, 7], "igl": [5, 7], "tra": 5, "trb": [5, 6, 15], "trd": 5, "attempt": 5, "none": [5, 16], "specif": [5, 7, 11, 12], "otherwis": [5, 16], "choic": 5, "commun": [5, 10], "filenam": [5, 16], "species_locu": 5, "_": 5, "substitut": 5, "space": [5, 16], "multipl": 5, "separ": [5, 11, 12], "plu": 5, "By": [5, 12], "advantag": 5, "inform": 5, "delin": [5, 12], "delieat": 5, "principl": [5, 16], "load": 5, "recommend": [5, 12], "ungap": [5, 8, 16], "respect": 5, "_gap": 5, "replac": [5, 16], "four": 5, "These": [5, 7, 11], "_v": 5, "_d": 5, "_j": 5, "_v_gap": 5, "multi_f": 5, "human": 5, "mous": 5, "c57bl": 5, "6": 5, "mu": 5, "musculu": 5, "try": 5, "help": 5, "event": 5, "10090": 5, "avail": [5, 12], "balb": 5, "ighv": [5, 7, 16], "cast": 5, "eij": 5, "lew": 5, "msm": 5, "msj": 5, "nod": 5, "shiltj": 5, "phj": 5, "airrc_igh": 5, "nomin": 6, "ref": [6, 16], "species_nam": 6, "quot": 6, "macaca": 6, "mulatta": 6, "www": 6, "db": 6, "imgtgenedb": 6, "referencesequ": 6, "nt": 6, "withgap": 6, "orf": 6, "inframep": 6, "trg": 6, "functional_onli": [6, 16], "rhesu": [7, 12], "macaqu": [7, 11, 12], "iglv": 7, "igkv": 7, "insert": [7, 11, 12], "rel": [7, 11], "most": [7, 8, 11], "downstream": [7, 11, 12], "chang": 7, "o": 7, "residu": [7, 11], "restor": 7, "standard": [7, 11], "remov": [7, 11, 12], "list": [7, 9, 11, 12, 16], "those": [7, 8, 12], "person": 7, "observ": 7, "produc": [7, 16], "unwant": 7, "ig": [7, 11], "infil": [7, 16], "outfil": [7, 13, 16], "customari": 7, "closest": [8, 11, 16], "onli": [8, 15, 16], "complet": 8, "5": [8, 16], "second": [8, 13, 16], "cystein": 8, "neither": 8, "appar": 8, "given": [8, 11, 16], "fraction": 8, "total": 8, "ignor": 8, "inferred_fil": 8, "ref_fil": [8, 12], "infer": 8, "ident": 9, "sub": 9, "identical_sequ": 9, "mixcr": 10, "receptor_util": 10, "modul": [10, 11, 14], "search": 10, "page": 10, "code": 11, "tr": 11, "pip": 11, "biopython": [11, 16], "name_allel": [11, 18], "nearest": 11, "taken": 11, "suppli": 11, "incorpor": 11, "snp": [11, 15], "scheme": [11, 12], "accord": 11, "emploi": 11, "each": [11, 12], "fix_macaque_gap": [11, 12, 18], "revert": 11, "next": 11, "extract_ref": [11, 15, 18], "identical_seq": [11, 18], "where": [11, 16], "same": [11, 16], "differ": [11, 16], "rev_comp": [11, 15, 18], "at_coord": [11, 18], "within": 11, "merge_fasta": [11, 18], "merg": [11, 13], "duplic": 11, "simpl": [11, 16], "wrapper": [11, 16], "around": [11, 16], "commonli": 11, "write": [11, 16], "comma": [11, 12], "tab": 11, "manag": 11, "string": [11, 16], "dict": [11, 16], "compar": [11, 13, 15], "expens": 11, "flexibl": 11, "rare": 11, "region": [12, 16], "cdr_coord": 12, "ndm_file": 12, "eg": 12, "vh": 12, "79": 12, "114": 12, "166": 12, "195": 12, "313": 12, "ensur": [12, 16], "correct": 12, "placement": 12, "coordin": 12, "assum": 12, "doubt": 12, "denot": 12, "b": 12, "etc": 12, "upon": 12, "applic": [12, 16], "we": [12, 16], "offer": 12, "conduct": 12, "realign": 12, "provis": 12, "file1": 13, "file2": 13, "themselv": 13, "rather": 13, "keep": 13, "2": [1, 13], "file_1": 13, "file_2": 13, "submodul": 14, "novel_allele_nam": 14, "aligned_diff": [14, 16], "build_ambiguous_ref": [14, 16], "closest_aligned_ref": [14, 16], "find_indel": [14, 16], "name_novel": [14, 16], "run_test": [14, 16], "number_v": 14, "check_conserved_residu": [14, 16], "distribute_cdr": [14, 16], "gap_align": [14, 16], "gap_align_aa": [14, 16], "gap_align_aa_from_nt": [14, 16], "gap_nt_from_aa": [14, 16], "insert_spac": [14, 16], "match_scor": [14, 16], "nt_diff": [14, 16], "number_from_trp": [14, 16], "number_ighv": [14, 16], "pretty_gap": [14, 16], "simple_bio_seq": 14, "chunk": [14, 16], "closest_ref": [14, 16], "dumb_consensu": [14, 16], "read_csv": [14, 16], "read_fasta": [14, 16], "read_imgt_fasta": [14, 16], "read_single_fasta": [14, 16], "reverse_compl": [14, 16], "sample_fasta": [14, 16], "scored_consensu": [14, 16], "scored_gap_consensu": [14, 16], "toseqrecord": [14, 16], "translat": [14, 16], "write_csv": [14, 16], "write_fasta": [14, 16], "content": 14, "construct": 15, "novel": [15, 16], "gene_nam": 15, "ref_set": [15, 16], "input": [15, 16, 17], "consid": 15, "novel_seq": 16, "str": 16, "ref_seq": 16, "full_ref": 16, "seq": 16, "entri": 16, "global": 16, "return": [16, 17], "equal": 16, "type": 16, "closest_ref_seq": 16, "v_gene": 16, "bool": 16, "true": 16, "its": 16, "conform": 16, "descript": 16, "prime": 16, "tupl": 16, "consist": 16, "three": 16, "novel_nam": 16, "aa": 16, "aa_seq": 16, "nt_gap": 16, "nucleotide_seq": 16, "peptide_seq": 16, "gapped_ref": 16, "ungapped_ref": 16, "po": 16, "target": 16, "match_list": 16, "thresh": 16, "s1": 16, "s2": 16, "next_p": 16, "trp_po": 16, "bioseq": 16, "store": 16, "collectiopn": 16, "coerc": 16, "upper": 16, "iter": 16, "corec": 16, "eas": 16, "basic": 16, "thing": 16, "don": 16, "int": 16, "yield": 16, "success": 16, "size": 16, "threshold": 16, "float": 16, "dumb": 16, "consensu": 16, "bio": 16, "aligninfo": 16, "summaryinfo": 16, "valu": 16, "atom": 16, "shorter": 16, "delimit": 16, "dictread": 16, "pathnam": 16, "dictionari": 16, "ighd": 16, "ighj": 16, "ch": 16, "include_orphon": 16, "underscor": 16, "mark": 16, "orphon": 16, "serquenc": 16, "nucelotid": 16, "random": 16, "resampl": 16, "valueerror": 16, "enough": 16, "gap_consensu": 16, "mimimum": 16, "score": 16, "achiev": 16, "min_scor": 16, "ambigu": 16, "x": 16, "require_multipl": 16, "adapt": 16, "seqrecord": 16, "truncat": 16, "ignore_partial_codon": 16, "amino": 16, "acid": 16, "so": 16, "termin": 16, "boundari": 16, "pad": 16, "get": 16, "row": 16, "scan_al": 16, "header": 16, "item": 16, "scan": 16, "written": 16}, "objects": {"": [[16, 0, 0, "-", "receptor_utils"]], "receptor_utils": [[16, 0, 0, "-", "novel_allele_name"], [16, 0, 0, "-", "number_v"], [16, 0, 0, "-", "simple_bio_seq"]], "receptor_utils.novel_allele_name": [[16, 1, 1, "", "aligned_diff"], [16, 1, 1, "", "build_ambiguous_ref"], [16, 1, 1, "", "closest_aligned_ref"], [16, 1, 1, "", "find_indels"], [16, 1, 1, "", "name_novel"], [16, 1, 1, "", "run_tests"]], "receptor_utils.number_v": [[16, 1, 1, "", "check_conserved_residues"], [16, 1, 1, "", "distribute_cdr"], [16, 1, 1, "", "gap_align"], [16, 1, 1, "", "gap_align_aa"], [16, 1, 1, "", "gap_align_aa_from_nt"], [16, 1, 1, "", "gap_nt_from_aa"], [16, 1, 1, "", "gap_sequence"], [16, 1, 1, "", "insert_space"], [16, 1, 1, "", "match_score"], [16, 1, 1, "", "nt_diff"], [16, 1, 1, "", "number_from_trp"], [16, 1, 1, "", "number_ighv"], [16, 1, 1, "", "pretty_gapped"], [16, 1, 1, "", "run_tests"]], "receptor_utils.simple_bio_seq": [[16, 1, 1, "", "chunks"], [16, 1, 1, "", "closest_ref"], [16, 1, 1, "", "dumb_consensus"], [16, 1, 1, "", "nt_diff"], [16, 1, 1, "", "read_csv"], [16, 1, 1, "", "read_fasta"], [16, 1, 1, "", "read_imgt_fasta"], [16, 1, 1, "", "read_single_fasta"], [16, 1, 1, "", "reverse_complement"], [16, 1, 1, "", "sample_fasta"], [16, 1, 1, "", "scored_consensus"], [16, 1, 1, "", "scored_gap_consensus"], [16, 1, 1, "", "toSeqRecords"], [16, 1, 1, "", "translate"], [16, 1, 1, "", "write_csv"], [16, 1, 1, "", "write_fasta"]]}, "objtypes": {"0": "py:module", "1": "py:function"}, "objnames": {"0": ["py", "module", "Python module"], "1": ["py", "function", "Python function"]}, "titleterms": {"us": [0, 1, 4, 11], "airr": [0, 1], "commun": [0, 1], "refer": [0, 1], "set": [0, 1], "igblast": [0, 4, 11, 18], "prerequisit": [0, 1], "1": [0, 1], "download": [0, 1], "germlin": [0, 1], "human": [0, 1], "ighv": [0, 1], "gene": [0, 1], "addit": 0, "file": 0, "2": 0, "log": [0, 1], "immcant": 0, "contain": [0, 1], "so": [0, 1], "we": [0, 1], "can": [0, 1], "its": [0, 1], "instal": [0, 1, 11], "tool": [0, 1, 11, 18], "3": [0, 1], "build": [0, 1], "databas": [0, 1, 4, 11], "4": [0, 1], "annot": [0, 1], "sampl": [0, 1], "sequenc": [0, 1, 10, 11, 18], "5": 0, "convert": 0, "output": 0, "chang": 0, "o": 0, "format": [0, 5], "mixcr": 1, "annotate_j": 2, "posit": [2, 3, 5, 6, 7, 8, 9, 12, 13, 15, 17], "argument": [2, 3, 5, 6, 7, 8, 9, 12, 13, 15, 17], "name": [2, 3, 5, 6, 11, 12, 15, 18], "at_coord": 3, "custom": [4, 11], "download_germline_set": 5, "option": 5, "exampl": 5, "extract_ref": 6, "fix_macaque_gap": 7, "gap_sequ": 8, "identical_seq": 9, "receptor": 10, "util": [10, 18], "code": 10, "work": [10, 18], "ig": 10, "tr": 10, "content": [10, 16], "indic": 10, "tabl": 10, "introduct": 11, "align": [11, 18], "v": [11, 18], "d": [11, 18], "j": [11, 18], "conveni": [11, 18], "simple_bio_seq": [11, 16], "api": 11, "make_igblast_ndm": 12, "note": 12, "merge_fasta": 13, "receptor_util": [14, 16], "name_allel": 15, "packag": 16, "submodul": 16, "novel_allele_nam": 16, "modul": 16, "number_v": 16, "rev_comp": 17, "command": 18, "line": 18}, "envversion": {"sphinx.domains.c": 2, "sphinx.domains.changeset": 1, "sphinx.domains.citation": 1, "sphinx.domains.cpp": 8, "sphinx.domains.index": 1, "sphinx.domains.javascript": 2, "sphinx.domains.math": 2, "sphinx.domains.python": 3, "sphinx.domains.rst": 2, "sphinx.domains.std": 2, "sphinx.ext.viewcode": 1, "sphinx": 57}, "alltitles": {"annotate_j": [[2, "annotate-j"]], "Positional Arguments": [[2, "positional-arguments"], [3, "positional-arguments"], [5, "positional-arguments"], [6, "positional-arguments"], [7, "positional-arguments"], [8, "positional-arguments"], [9, "positional-arguments"], [12, "positional-arguments"], [13, "positional-arguments"], [15, "positional-arguments"], [17, "positional-arguments"]], "Named Arguments": [[2, "named-arguments"], [3, "named-arguments"], [5, "named-arguments"], [6, "named-arguments"], [12, "named-arguments"], [15, "named-arguments"]], "at_coords": [[3, "at-coords"]], "Using custom databases with IgBlast": [[4, "using-custom-databases-with-igblast"], [11, "using-custom-databases-with-igblast"]], "download_germline_set": [[5, "download-germline-set"]], "Format Options": [[5, "format-options"]], "Examples": [[5, "examples"]], "extract_refs": [[6, "extract-refs"]], "fix_macaque_gaps": [[7, "fix-macaque-gaps"]], "gap_sequences": [[8, "gap-sequences"]], "identical_seqs": [[9, "identical-seqs-label"]], "Receptor-utils: code and utilities for working with IG/TR receptor sequences": [[10, "receptor-utils-code-and-utilities-for-working-with-ig-tr-receptor-sequences"]], "Contents:": [[10, null]], "Indices and tables": [[10, "indices-and-tables"]], "Introduction": [[11, "introduction"]], "Installation": [[11, "installation"]], "Naming and aligning V, D and J sequences": [[11, "naming-and-aligning-v-d-and-j-sequences"]], "Convenience Tools": [[11, "convenience-tools"], [18, "convenience-tools"]], "simple_bio_seq API": [[11, "simple-bio-seq-api"]], "make_igblast_ndm": [[12, "make-igblast-ndm"]], "Notes": [[12, "notes"]], "merge_fasta": [[13, "merge-fasta"]], "receptor_utils": [[14, "receptor-utils"]], "name_allele": [[15, "name-allele"]], "receptor_utils package": [[16, "receptor-utils-package"]], "Submodules": [[16, "submodules"]], "receptor_utils.novel_allele_name module": [[16, "module-receptor_utils.novel_allele_name"]], "receptor_utils.number_v module": [[16, "module-receptor_utils.number_v"]], "receptor_utils.simple_bio_seq module": [[16, "module-receptor_utils.simple_bio_seq"]], "Module contents": [[16, "module-receptor_utils"]], "rev_comp": [[17, "rev-comp"]], "Command-Line Utilities": [[18, "command-line-utilities"]], "Naming and Aligning V, D and J Sequences": [[18, "naming-and-aligning-v-d-and-j-sequences"]], "Working with IgBLAST": [[18, "working-with-igblast"]], "Using AIRR Community Reference Sets with IgBLAST": [[0, "using-airr-community-reference-sets-with-igblast"]], "Prerequisites": [[0, "prerequisites"], [1, "prerequisites"]], "1. Download the germline set for human IGHV genes and additional IgBLAST files": [[0, "download-the-germline-set-for-human-ighv-genes-and-additional-igblast-files"]], "2. Log in to the Immcantation container so that we can use its installed tools": [[0, "log-in-to-the-immcantation-container-so-that-we-can-use-its-installed-tools"]], "3. Build IgBLAST databases for the germline set": [[0, "build-igblast-databases-for-the-germline-set"]], "4. Annotate the sample sequences using IgBLAST": [[0, "annotate-the-sample-sequences-using-igblast"]], "5. Convert the IgBLAST output to Change-O format": [[0, "convert-the-igblast-output-to-change-o-format"]], "Using AIRR Community Reference Sets with MiXCR": [[1, "using-airr-community-reference-sets-with-mixcr"]], "1. Download the germline set for human IGHV genes": [[1, "download-the-germline-set-for-human-ighv-genes"]], "1. Log in to the MiXCR container so that we can use its installed tools": [[1, "log-in-to-the-mixcr-container-so-that-we-can-use-its-installed-tools"]], "3. Build MiXCR database for the germline set": [[1, "build-mixcr-database-for-the-germline-set"]], "4. Annotate the sample sequences using MiXCR": [[1, "annotate-the-sample-sequences-using-mixcr"]]}, "indexentries": {}})